

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>The liquorice python package &mdash; LIQUORICE 0.5.0 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="The LIQUORICE command-line-tool" href="liquorice_commandline_tool.html" />
    <link rel="prev" title="Introduction" href="intro.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html">
          

          
            
            <img src="_static/liquorice_logo_fitted_transparent.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="intro.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="intro.html#installation">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="intro.html#citation">Citation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">The <cite>liquorice</cite> python package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#module-liquorice.LIQUORICE">liquorice.LIQUORICE module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-liquorice.LIQUORICE_summary">liquorice.LIQUORICE_summary module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#subpackage-liquorice-utils">Subpackage: liquorice.utils</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#module-liquorice.utils.AggregateAcrossRegions">liquorice.utils.AggregateAcrossRegions module</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-liquorice.utils.BiasFactorWeights">liquorice.utils.BiasFactorWeights module</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-liquorice.utils.BiasModel">liquorice.utils.BiasModel module</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-liquorice.utils.BinTableBiasFactors">liquorice.utils.BinTableBiasFactors module</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-liquorice.utils.CorrectForCNAs">liquorice.utils.CorrectForCNAs module</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-liquorice.utils.CoverageAndSequenceTablePreparation">liquorice.utils.CoverageAndSequenceTablePreparation module</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-liquorice.utils.FitGaussians">liquorice.utils.FitGaussians module</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-liquorice.utils.GlobalFragmentSize">liquorice.utils.GlobalFragmentSize module</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-liquorice.utils.MeanSequencingDepth">liquorice.utils.MeanSequencingDepth module</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-liquorice.utils.Plotting">liquorice.utils.Plotting module</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-liquorice.utils.RegionFilteringAndBinning">liquorice.utils.RegionFilteringAndBinning module</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-liquorice.utils.Workflows">liquorice.utils.Workflows module</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="liquorice_commandline_tool.html">The LIQUORICE command-line-tool</a></li>
<li class="toctree-l1"><a class="reference internal" href="liquorice_commandline_tool.html#liquorice-s-summary-tool">LIQUORICE’s summary tool</a></li>
<li class="toctree-l1"><a class="reference internal" href="liquorice_commandline_tool.html#usage-parameters-and-examples">Usage, Parameters, and Examples</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">LIQUORICE</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>The <cite>liquorice</cite> python package</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="_sources/liquorice.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="the-liquorice-python-package">
<h1>The <cite>liquorice</cite> python package<a class="headerlink" href="#the-liquorice-python-package" title="Permalink to this headline">¶</a></h1>
<div class="section" id="module-liquorice.LIQUORICE">
<span id="liquorice-liquorice-module"></span><span id="liquorice-liquorice-anchor"></span><h2>liquorice.LIQUORICE module<a class="headerlink" href="#module-liquorice.LIQUORICE" title="Permalink to this headline">¶</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="liquorice.LIQUORICE.FullPaths">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">liquorice.LIQUORICE.</span></span><span class="sig-name descname"><span class="pre">FullPaths</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">option_strings</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dest</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nargs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">const</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">default</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">choices</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">required</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">help</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metavar</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/LIQUORICE.html#FullPaths"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.LIQUORICE.FullPaths" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">argparse.Action</span></code></p>
<p>Expand user- and relative-paths. From: <a class="reference external" href="https://gist.github.com/brantfaircloth/1252339">https://gist.github.com/brantfaircloth/1252339</a>. Does not convert the path
to the blacklist if the value is ‘hg38’ or ‘10k_random, which is interpreted as an instruction to use the shipped
list instead.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="liquorice.LIQUORICE.main">
<span class="sig-prename descclassname"><span class="pre">liquorice.LIQUORICE.</span></span><span class="sig-name descname"><span class="pre">main</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/LIQUORICE.html#main"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.LIQUORICE.main" title="Permalink to this definition">¶</a></dt>
<dd><p>Main function for the LIQUORICE command line tool. Performs the following steps: Sanity checks for user input,
setup of environment variables and general setup-steps, handles file paths, selects which steps need to be
performed based on user input, logs the progress and performs the actual LIQUORICE analysis by calling
<a class="reference internal" href="#liquorice.utils.GlobalFragmentSize.get_list_of_fragment_lengths_and_avg_readlength" title="liquorice.utils.GlobalFragmentSize.get_list_of_fragment_lengths_and_avg_readlength"><code class="xref py py-func docutils literal notranslate"><span class="pre">liquorice.utils.GlobalFragmentSize.get_list_of_fragment_lengths_and_avg_readlength()</span></code></a>,
<a class="reference internal" href="#liquorice.utils.MeanSequencingDepth.sample_bam_to_get_sequencing_depth" title="liquorice.utils.MeanSequencingDepth.sample_bam_to_get_sequencing_depth"><code class="xref py py-func docutils literal notranslate"><span class="pre">liquorice.utils.MeanSequencingDepth.sample_bam_to_get_sequencing_depth()</span></code></a>,
<a class="reference internal" href="#liquorice.utils.Workflows.train_biasmodel_for_sample" title="liquorice.utils.Workflows.train_biasmodel_for_sample"><code class="xref py py-func docutils literal notranslate"><span class="pre">liquorice.utils.Workflows.train_biasmodel_for_sample()</span></code></a>,
<a class="reference internal" href="#liquorice.utils.Workflows.run_liquorice_on_regionset_with_pretrained_biasmodel" title="liquorice.utils.Workflows.run_liquorice_on_regionset_with_pretrained_biasmodel"><code class="xref py py-func docutils literal notranslate"><span class="pre">liquorice.utils.Workflows.run_liquorice_on_regionset_with_pretrained_biasmodel()</span></code></a>, and/or
<a class="reference internal" href="#liquorice.utils.Workflows.run_liquorice_train_biasmodel_on_same_regions" title="liquorice.utils.Workflows.run_liquorice_train_biasmodel_on_same_regions"><code class="xref py py-func docutils literal notranslate"><span class="pre">liquorice.utils.Workflows.run_liquorice_train_biasmodel_on_same_regions()</span></code></a> after generating the
corresponding objects.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="liquorice.LIQUORICE.parse_args">
<span class="sig-prename descclassname"><span class="pre">liquorice.LIQUORICE.</span></span><span class="sig-name descname"><span class="pre">parse_args</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/LIQUORICE.html#parse_args"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.LIQUORICE.parse_args" title="Permalink to this definition">¶</a></dt>
<dd><p>Parses the arguments from the command line. For a full list of arguments,
see the documentation of the LIQUORICE command line tool.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>An <cite>argparse.ArgumentParser</cite> object storing the arguments.</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-liquorice.LIQUORICE_summary">
<span id="liquorice-liquorice-summary-module"></span><h2>liquorice.LIQUORICE_summary module<a class="headerlink" href="#module-liquorice.LIQUORICE_summary" title="Permalink to this headline">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="liquorice.LIQUORICE_summary.boxplot_score_summary">
<span class="sig-prename descclassname"><span class="pre">liquorice.LIQUORICE_summary.</span></span><span class="sig-name descname"><span class="pre">boxplot_score_summary</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">summary_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">comparison_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_uncorrected_coverage</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/LIQUORICE_summary.html#boxplot_score_summary"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.LIQUORICE_summary.boxplot_score_summary" title="Permalink to this definition">¶</a></dt>
<dd><p>Summarize scores via boxplots, with one box per group (case/control) - region-set combination.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>summary_df</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code>) – The output of the function <a class="reference internal" href="#liquorice.LIQUORICE_summary.create_summary_table_LIQUORICE" title="liquorice.LIQUORICE_summary.create_summary_table_LIQUORICE"><code class="xref py py-func docutils literal notranslate"><span class="pre">create_summary_table_LIQUORICE()</span></code></a>.</p></li>
<li><p><strong>comparison_col</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Use this column for the y axis of the boxplots.</p></li>
<li><p><strong>use_uncorrected_coverage</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – If True, indicate in the output filename that the uncorrected coverage scores are
shown.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="liquorice.LIQUORICE_summary.check_difference_to_control">
<span class="sig-prename descclassname"><span class="pre">liquorice.LIQUORICE_summary.</span></span><span class="sig-name descname"><span class="pre">check_difference_to_control</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">row</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">control_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">negative_is_strong</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">zscore_treshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/LIQUORICE_summary.html#check_difference_to_control"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.LIQUORICE_summary.check_difference_to_control" title="Permalink to this definition">¶</a></dt>
<dd><p>For a row in a <cite>pandas.DataFrame</cite>, find control values for the same regionset and return if sample is significantly
different compared to the controls, as measured by metric <strong>col</strong>. Assumes controls are normally distributed.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>row</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">Series</span></code>) – A <cite>pandas.Series</cite> with the columns ‘region-set’, ‘is control’, and <strong>col</strong>, corresponding to a single
sample.</p></li>
<li><p><strong>control_df</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code>) – A <cite>pandas.DataFrame</cite> with at least the columns ‘region-set’, and <strong>col</strong>. Should only
contain data of control samples.</p></li>
<li><p><strong>col</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Use this column as a metric to determine differences.</p></li>
<li><p><strong>negative_is_strong</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – Set to <cite>True</cite> if a very negative value of the metric is associated with a strong dip
signal, such as for the dip area.</p></li>
<li><p><strong>zscore_treshold</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Use this number as a treshold for significance: Significant samples must have a signal that
is outside &lt;mean of controls&gt; +- &lt;zscore_treshold&gt;*&lt;standard deviation of controls&gt;. The default of 2 roughly
corresponds to alpha=0.05 if the signal in the controls is normally distributed.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A string indicating the result of the comparison, or np.nan if row[col] is NaN.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="liquorice.LIQUORICE_summary.create_summary_table_LIQUORICE">
<span class="sig-prename descclassname"><span class="pre">liquorice.LIQUORICE_summary.</span></span><span class="sig-name descname"><span class="pre">create_summary_table_LIQUORICE</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dirname</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">control_name_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">these_regionsets_only</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_uncorrected_coverage</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">zscore_treshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/LIQUORICE_summary.html#create_summary_table_LIQUORICE"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.LIQUORICE_summary.create_summary_table_LIQUORICE" title="Permalink to this definition">¶</a></dt>
<dd><p>For a LIQUORICE result directory, creates and writes to csv a pd.DataFrame summarizing the coverage drop metrics
for all samples and
region-sets. If <strong>control_name_list</strong> is given, compares the case samples to control samples, separately for each
region-set.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>use_uncorrected_coverage</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – If True, summarize the coverage profile that is not corrected for biases by
LIQUORICE instead of the corrected coverage.</p></li>
<li><p><strong>these_regionsets_only</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]) – Summarize only data for these regionsets.</p></li>
<li><p><strong>dirname</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Output directory of LIQUORICE in which to search for fitted_gaussians_parameter_summary.csv
(or uncorrected_fitted_gaussians_parameter_summary.csv) files</p></li>
<li><p><strong>control_name_list</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]) – Sample names of the controls, which will be used as reference for generating z-scores.</p></li>
<li><p><strong>zscore_treshold</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Use this number as a treshold for significance: Significant samples must have a signal that
is outside &lt;mean of controls&gt; +- &lt;zscore_treshold&gt;*&lt;standard deviation of controls&gt;. The default of 2 roughly
corresponds to alpha=0.05 if the signal in the controls is normally distributed.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code></p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A <cite>pandas.DataFrame</cite> in which all parameters saved in the ‘fitted_gaussians_parameter_summary.csv’
(or uncorrected_fitted_gaussians_parameter_summary.csv) files
are summarized over all samples and region-sets. If control_name_list is not empty, additional columns of the
DataFame contain metric comparisons to the the control samples in the form of z-scores of dip area and depth.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="liquorice.LIQUORICE_summary.get_binsize_and_extendto_from_saved_settings">
<span class="sig-prename descclassname"><span class="pre">liquorice.LIQUORICE_summary.</span></span><span class="sig-name descname"><span class="pre">get_binsize_and_extendto_from_saved_settings</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dirname</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">regionset</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/LIQUORICE_summary.html#get_binsize_and_extendto_from_saved_settings"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.LIQUORICE_summary.get_binsize_and_extendto_from_saved_settings" title="Permalink to this definition">¶</a></dt>
<dd><p>Calls <a class="reference internal" href="#liquorice.LIQUORICE_summary.verify_consistant_binning_settings" title="liquorice.LIQUORICE_summary.verify_consistant_binning_settings"><code class="xref py py-func docutils literal notranslate"><span class="pre">verify_consistant_binning_settings()</span></code></a> and returns the binsize and extend_to settings used by LIQUORICE
for a given regionset, if no error is raised.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dirname</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Path to LIQUORICE’s working directory.</p></li>
<li><p><strong>regionset</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Name of the regionset of interest.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>]</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A tuple with two integers: binsize and extend_to.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="liquorice.LIQUORICE_summary.get_list_of_coveragefiles">
<span class="sig-prename descclassname"><span class="pre">liquorice.LIQUORICE_summary.</span></span><span class="sig-name descname"><span class="pre">get_list_of_coveragefiles</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dirname</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">regionset</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_uncorrected_coverage</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/LIQUORICE_summary.html#get_list_of_coveragefiles"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.LIQUORICE_summary.get_list_of_coveragefiles" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a list of ‘corrected_coverage_mean_per_bin.csv’ or ‘uncorrected_coverage_mean_per_bin.csv’ files, or
return an error if no such files could be found.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dirname</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – path to LIQUORICE output directory</p></li>
<li><p><strong>regionset</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Name of the regionset of interest</p></li>
<li><p><strong>use_uncorrected_coverage</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – If True, load ‘uncorrected_coverage_mean_per_bin.csv’ files, else, load
‘corrected_coverage_mean_per_bin.csv files</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A list of ‘corrected_coverage_mean_per_bin.csv’ or ‘uncorrected_coverage_mean_per_bin.csv’ file paths</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="liquorice.LIQUORICE_summary.get_ymin_ymax_over_all_samples">
<span class="sig-prename descclassname"><span class="pre">liquorice.LIQUORICE_summary.</span></span><span class="sig-name descname"><span class="pre">get_ymin_ymax_over_all_samples</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">coverage_filelist</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">normalize_by_intercept</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">regionset</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">summary_df</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/LIQUORICE_summary.html#get_ymin_ymax_over_all_samples"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.LIQUORICE_summary.get_ymin_ymax_over_all_samples" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the the upper and lower limits for the y-axis for a given regionset, such that the same scale is used for all
samples.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>coverage_filelist</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]) – List of ‘corrected_coverage_mean_per_bin.csv’ (or ‘uncorrected_coverage_mean_per_bin.csv’)
files</p></li>
<li><p><strong>normalize_by_intercept</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – f True, extract the intercept from the fitted model to normalize the dips between
samples (intercept of each sample is positioned at y=0). Otherwise, the mean coverage of each sample is
positioned at y=0.</p></li>
<li><p><strong>regionset</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Name of the regionset of intest as given in the summary_df</p></li>
<li><p><strong>summary_df</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code>) – pd.DataFrame with columns ‘sample’, ‘region-set’, and ‘Intercept’.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A tuple of floats: y_min,y_max</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="liquorice.LIQUORICE_summary.main">
<span class="sig-prename descclassname"><span class="pre">liquorice.LIQUORICE_summary.</span></span><span class="sig-name descname"><span class="pre">main</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/LIQUORICE_summary.html#main"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.LIQUORICE_summary.main" title="Permalink to this definition">¶</a></dt>
<dd><p>Main function for the LIQUOIRCE_summary tool. Calls the argument parser, checks user input, and calls the functions
<a class="reference internal" href="#liquorice.LIQUORICE_summary.verify_consistant_binning_settings" title="liquorice.LIQUORICE_summary.verify_consistant_binning_settings"><code class="xref py py-func docutils literal notranslate"><span class="pre">verify_consistant_binning_settings()</span></code></a>
<a class="reference internal" href="#liquorice.LIQUORICE_summary.create_summary_table_LIQUORICE" title="liquorice.LIQUORICE_summary.create_summary_table_LIQUORICE"><code class="xref py py-func docutils literal notranslate"><span class="pre">create_summary_table_LIQUORICE()</span></code></a> (saving output to .csv),
<a class="reference internal" href="#liquorice.LIQUORICE_summary.plot_overlay" title="liquorice.LIQUORICE_summary.plot_overlay"><code class="xref py py-func docutils literal notranslate"><span class="pre">plot_overlay()</span></code></a>, <a class="reference internal" href="#liquorice.LIQUORICE_summary.plot_as_subplots" title="liquorice.LIQUORICE_summary.plot_as_subplots"><code class="xref py py-func docutils literal notranslate"><span class="pre">plot_as_subplots()</span></code></a>, and <a class="reference internal" href="#liquorice.LIQUORICE_summary.boxplot_score_summary" title="liquorice.LIQUORICE_summary.boxplot_score_summary"><code class="xref py py-func docutils literal notranslate"><span class="pre">boxplot_score_summary()</span></code></a>.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="liquorice.LIQUORICE_summary.parse_args">
<span class="sig-prename descclassname"><span class="pre">liquorice.LIQUORICE_summary.</span></span><span class="sig-name descname"><span class="pre">parse_args</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/LIQUORICE_summary.html#parse_args"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.LIQUORICE_summary.parse_args" title="Permalink to this definition">¶</a></dt>
<dd><p>Parses the arguments from the command line. For a full list of arguments,
see the documentation of the LIQUORICE_summary command line tool.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>An <cite>argparse.ArgumentParser</cite> object storing the arguments.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="liquorice.LIQUORICE_summary.plot_as_subplots">
<span class="sig-prename descclassname"><span class="pre">liquorice.LIQUORICE_summary.</span></span><span class="sig-name descname"><span class="pre">plot_as_subplots</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dirname</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">summary_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">control_name_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extend_to</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">binsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">normalize_by_intercept</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">smooth_sigma</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">significance_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Dip</span> <span class="pre">area:</span> <span class="pre">interpretation</span> <span class="pre">vs</span> <span class="pre">controls</span> <span class="pre">in</span> <span class="pre">same</span> <span class="pre">region</span> <span class="pre">set'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_uncorrected_coverage</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y_min_fixed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y_max_fixed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/LIQUORICE_summary.html#plot_as_subplots"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.LIQUORICE_summary.plot_as_subplots" title="Permalink to this definition">¶</a></dt>
<dd><p>Summarize plots, create one set of plots per regionset. Control samples are compared to case samples, significant
changes are marked by color. <strong>smooth_sigma</strong> can be used to make the plots smoother and easier to compare, this is
a visual effect only.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dirname</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Path to LIQUORICE’s working directory.</p></li>
<li><p><strong>summary_df</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code>) – The output of the function <a class="reference internal" href="#liquorice.LIQUORICE_summary.create_summary_table_LIQUORICE" title="liquorice.LIQUORICE_summary.create_summary_table_LIQUORICE"><code class="xref py py-func docutils literal notranslate"><span class="pre">create_summary_table_LIQUORICE()</span></code></a>.</p></li>
<li><p><strong>control_name_list</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]) – List of names of samples that should be plotted as controls.  Can be empty.</p></li>
<li><p><strong>extend_to</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>]) – <em>extend_to</em> setting that was used when running LIQUORICE. If None, infer from the
binning_settings.json files that are saved by LIQUORICE by default.</p></li>
<li><p><strong>binsize</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>]) – <em>binsize</em> setting that was used when running LIQUORICE. If None, infer from the
binning_settings.json files that are saved by LIQUORICE by default.</p></li>
<li><p><strong>normalize_by_intercept</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – If True, extract the intercept from the fitted model to normalize the dips between
samples (intercept of each sample is positioned at y=0). Otherwise, the mean coverage of each sample is
positioned at y=0.</p></li>
<li><p><strong>smooth_sigma</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – Visually smooth the coverage signals, using this strength of smoothing. Higher values indicate
stronger smoothing. Set to 0 for no smoothing.</p></li>
<li><p><strong>significance_col</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Use this columns as an indicator for significant differences between case and control
samples. Can contain the following values: “Significantly stronger coverage drop”,
“Significantly weaker coverage drop”, “n.s.”, and “N/A”.</p></li>
<li><p><strong>use_uncorrected_coverage</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – If True, plot the coverage profile that is not corrected for biases by LIQUORICE.</p></li>
<li><p><strong>y_min_fixed</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]) – If specified, use this value as the minimum value for the y axis.</p></li>
<li><p><strong>y_min_fixed</strong> – If specified, use this value as the maximum value for the y axis.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="liquorice.LIQUORICE_summary.plot_overlay">
<span class="sig-prename descclassname"><span class="pre">liquorice.LIQUORICE_summary.</span></span><span class="sig-name descname"><span class="pre">plot_overlay</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dirname</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">summary_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">control_name_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extend_to</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">binsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">normalize_by_intercept</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">smooth_sigma</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">significance_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Dip</span> <span class="pre">area:</span> <span class="pre">interpretation</span> <span class="pre">vs</span> <span class="pre">controls</span> <span class="pre">in</span> <span class="pre">same</span> <span class="pre">region</span> <span class="pre">set'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_uncorrected_coverage</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">linewidth</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/LIQUORICE_summary.html#plot_overlay"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.LIQUORICE_summary.plot_overlay" title="Permalink to this definition">¶</a></dt>
<dd><p>Summarize plots, create one plot per regionset. Control samples are compared to case samples, significant changes
(based on z-scores; assumes normal distribution of controls) are marked by color.
<strong>smooth_sigma</strong> can be used to make the plots smoother and easier to compare - this is a visual
effect only.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dirname</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Path to LIQUORICE’s working directory.</p></li>
<li><p><strong>summary_df</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code>) – The output of the function <a class="reference internal" href="#liquorice.LIQUORICE_summary.create_summary_table_LIQUORICE" title="liquorice.LIQUORICE_summary.create_summary_table_LIQUORICE"><code class="xref py py-func docutils literal notranslate"><span class="pre">create_summary_table_LIQUORICE()</span></code></a>.</p></li>
<li><p><strong>control_name_list</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]) – List of names of samples that should be plotted as controls.  Can be empty.</p></li>
<li><p><strong>extend_to</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>]) – <em>extend_to</em> setting that was used when running LIQUORICE. If None, infer from the
binning_settings.json files that are saved by LIQUORICE by default.</p></li>
<li><p><strong>binsize</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>]) – <em>binsize</em> setting that was used when running LIQUORICE. If None, infer from the
binning_settings.json files that are saved by LIQUORICE by default.</p></li>
<li><p><strong>normalize_by_intercept</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – If True, extract the intercept from the fitted model to normalize the dips between
samples (intercept of each sample is positioned at y=0). Otherwise, the mean coverage of each sample is
positioned at y=0.</p></li>
<li><p><strong>smooth_sigma</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – Visually smooth the coverage signals, using this strength of smoothing. Higher values indicate
stronger smoothing. Set to 0 for no smoothing.</p></li>
<li><p><strong>significance_col</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Use this columns as an indicator for significant differences between case and control
samples. Can contain the following values: “Significantly stronger coverage drop”,
“Significantly weaker coverage drop”, “n.s.”, and “N/A”.</p></li>
<li><p><strong>use_uncorrected_coverage</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – If True, plot the coverage profile that is not corrected for biases by LIQUORICE.</p></li>
<li><p><strong>alpha</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – Alpha (transparency) parameter for plotting.</p></li>
<li><p><strong>linewidth</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – Linewidth parameter for plotting</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="liquorice.LIQUORICE_summary.verify_consistant_binning_settings">
<span class="sig-prename descclassname"><span class="pre">liquorice.LIQUORICE_summary.</span></span><span class="sig-name descname"><span class="pre">verify_consistant_binning_settings</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dirname</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">regionset</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/LIQUORICE_summary.html#verify_consistant_binning_settings"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.LIQUORICE_summary.verify_consistant_binning_settings" title="Permalink to this definition">¶</a></dt>
<dd><p>Asserts that, for the given regionset, all samples have the same binning settings.
Calls <cite>sys.exit()</cite> with an error message if settings are inconsistent or cannot be found.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>regionset_list</strong> – A list of region-sets to be analyzed.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="liquorice.LIQUORICE_summary.zscore_to_control">
<span class="sig-prename descclassname"><span class="pre">liquorice.LIQUORICE_summary.</span></span><span class="sig-name descname"><span class="pre">zscore_to_control</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">row</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">control_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">col</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/LIQUORICE_summary.html#zscore_to_control"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.LIQUORICE_summary.zscore_to_control" title="Permalink to this definition">¶</a></dt>
<dd><p>For a row in a dataframe, find control values for the same regionset and return the zscore of the sample compared to
these controls, as measured by metric “col”.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>row</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">Series</span></code>) – A pd.Series with the columns ‘region-set’, ‘is control’, and <strong>col</strong>, corresponding to a single sample.</p></li>
<li><p><strong>control_df</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code>) – A <cite>pandas.DataFrame</cite> with at least the columns ‘region-set’, and <strong>col</strong>. Should only
contain data of control samples.</p></li>
<li><p><strong>col</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Calculate z-score by comparing this metric between the sample and controls.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>z-score, rounded to 2 decimals</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="subpackage-liquorice-utils">
<h2>Subpackage: liquorice.utils<a class="headerlink" href="#subpackage-liquorice-utils" title="Permalink to this headline">¶</a></h2>
<div class="section" id="module-liquorice.utils.AggregateAcrossRegions">
<span id="liquorice-utils-aggregateacrossregions-module"></span><h3>liquorice.utils.AggregateAcrossRegions module<a class="headerlink" href="#module-liquorice.utils.AggregateAcrossRegions" title="Permalink to this headline">¶</a></h3>
<dl class="py function">
<dt class="sig sig-object py" id="liquorice.utils.AggregateAcrossRegions.aggregate_across_regions">
<span class="sig-prename descclassname"><span class="pre">liquorice.utils.AggregateAcrossRegions.</span></span><span class="sig-name descname"><span class="pre">aggregate_across_regions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">column_of_interest</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/utils/AggregateAcrossRegions.html#aggregate_across_regions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.utils.AggregateAcrossRegions.aggregate_across_regions" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code>) – A <code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>, containing the columns “bin nr.” and <strong>column_of_interest</strong></p></li>
<li><p><strong>column_of_interest</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Column name for which the mean per bin nr. should be returned</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">Series</span></code></p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A <code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.Series</span></code> with mean values of <strong>column_of_interest</strong> per bin, aggregated across all
regions in the <code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>.</p>
</dd>
</dl>
<p>Example:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">mean_corrected_coverage_leftmost_bin</span> <span class="o">=</span> <span class="n">aggregate_across_regions</span><span class="p">(</span><span class="n">df</span><span class="p">,</span><span class="s2">&quot;corrected coverage&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
</pre></div>
</div>
</dd></dl>

</div>
<div class="section" id="module-liquorice.utils.BiasFactorWeights">
<span id="liquorice-utils-biasfactorweights-module"></span><h3>liquorice.utils.BiasFactorWeights module<a class="headerlink" href="#module-liquorice.utils.BiasFactorWeights" title="Permalink to this headline">¶</a></h3>
<dl class="py function">
<dt class="sig sig-object py" id="liquorice.utils.BiasFactorWeights.get_GC_weights_binsize1">
<span class="sig-prename descclassname"><span class="pre">liquorice.utils.BiasFactorWeights.</span></span><span class="sig-name descname"><span class="pre">get_GC_weights_binsize1</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fragments</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/utils/BiasFactorWeights.html#get_GC_weights_binsize1"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.utils.BiasFactorWeights.get_GC_weights_binsize1" title="Permalink to this definition">¶</a></dt>
<dd><p>For a given fragment size distribution, returns a list corresponding to the GC weight vector for a bin of size 1</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>fragments</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>]) – A list of fragment lengths that is representative of the sample’s global fragment length
distribution</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A list corresponding to the GC weight vector for a bin of size 1. It’s length is <cite>max(fragments)*2</cite>, the
weight corresponding to the bin itself is at index <cite>max(fragments)</cite>. Values represent influence of a nucleotide
at a given position on the bin’s coverage.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="liquorice.utils.BiasFactorWeights.get_GC_weights_binwide">
<span class="sig-prename descclassname"><span class="pre">liquorice.utils.BiasFactorWeights.</span></span><span class="sig-name descname"><span class="pre">get_GC_weights_binwide</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">binsize</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fragments</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dont_slide</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/utils/BiasFactorWeights.html#get_GC_weights_binwide"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.utils.BiasFactorWeights.get_GC_weights_binwide" title="Permalink to this definition">¶</a></dt>
<dd><p>For a given binsize, calculate the influence of a nucleotide at a given position on the bin’s coverage</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>binsize</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Size of the bin</p></li>
<li><p><strong>fragments</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>]) – A list of fragment lengths that is representative of the samples global fragment length
distribution</p></li>
<li><p><strong>dont_slide</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – Set to False to return weights of 1 in the bin area and 0 outside of it</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A list of length <cite>(2*max(fragments)+binsize)</cite>, the weight corresponding to the first position within the
bin itself is at index <cite>max(fragments)</cite>. Values represent influence of a nucleotide at a given position on the
bin’s coverage.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="liquorice.utils.BiasFactorWeights.get_dinuc_weights_binwide">
<span class="sig-prename descclassname"><span class="pre">liquorice.utils.BiasFactorWeights.</span></span><span class="sig-name descname"><span class="pre">get_dinuc_weights_binwide</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">GC_weights</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/utils/BiasFactorWeights.html#get_dinuc_weights_binwide"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.utils.BiasFactorWeights.get_dinuc_weights_binwide" title="Permalink to this definition">¶</a></dt>
<dd><p>For a given binsize, calculate the influence of a given dinucleotide starting at a given position on the bin’s
coverage. This simply averages the GC weight of the position of interest and the following position. For the last
position, a weight of 0 is returned.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>GC_weights</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]) – result of <a class="reference internal" href="#liquorice.utils.BiasFactorWeights.get_GC_weights_binwide" title="liquorice.utils.BiasFactorWeights.get_GC_weights_binwide"><code class="xref py py-func docutils literal notranslate"><span class="pre">get_GC_weights_binwide()</span></code></a> for the appropriate binsize and fragments of interest.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A list of length <cite>(2*max(fragments)+binsize)</cite>, the weight corresponding to the first position within the
bin itself is at index <cite>max(fragments)</cite>. Values represent influence of a dinucleotide starting at a given
position on the  bin’s coverage.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="liquorice.utils.BiasFactorWeights.get_mapp_weights_binwide">
<span class="sig-prename descclassname"><span class="pre">liquorice.utils.BiasFactorWeights.</span></span><span class="sig-name descname"><span class="pre">get_mapp_weights_binwide</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">binsize</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fragments</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dont_slide</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/utils/BiasFactorWeights.html#get_mapp_weights_binwide"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.utils.BiasFactorWeights.get_mapp_weights_binwide" title="Permalink to this definition">¶</a></dt>
<dd><p>For a given binsize, calculate the influence a fragment starting/ending at a given position on the bin’s coverage</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>binsize</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Size of the bin</p></li>
<li><p><strong>fragments</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>]) – A list of fragment lengths that is representative of the samples global fragment length 
distribution</p></li>
<li><p><strong>dont_slide</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – Set to False to return weights of 1 in the bin area and 0 outside of it</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]]</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>Two lists of length <cite>(2*max(fragments)+binsize)</cite>. Values represent influence a fragment
starting/ending at a given position on the bin’s coverage. First list is for forward mappability
(fragments starting at the position), second list is for reverse mappability (fragments ending at the position)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="liquorice.utils.BiasFactorWeights.get_nucleotide_dicts">
<span class="sig-prename descclassname"><span class="pre">liquorice.utils.BiasFactorWeights.</span></span><span class="sig-name descname"><span class="pre">get_nucleotide_dicts</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/utils/BiasFactorWeights.html#get_nucleotide_dicts"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.utils.BiasFactorWeights.get_nucleotide_dicts" title="Permalink to this definition">¶</a></dt>
<dd><p>Prepare dictionaries with di-and trinucleotides, as well as a dictionary for forward/reverse complement translation.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>]]</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Three dictionaries: All dinucleotides excluding reverse complements, all trinucleotides excluding reverse
complements, and a dictonary that allows translation of reverse complements to their forward complement counterparts
(in this order). Keys are all 0.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="liquorice.utils.BiasFactorWeights.get_trinuc_weights_binwide">
<span class="sig-prename descclassname"><span class="pre">liquorice.utils.BiasFactorWeights.</span></span><span class="sig-name descname"><span class="pre">get_trinuc_weights_binwide</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">GC_weights</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/utils/BiasFactorWeights.html#get_trinuc_weights_binwide"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.utils.BiasFactorWeights.get_trinuc_weights_binwide" title="Permalink to this definition">¶</a></dt>
<dd><p>For a given binsize, calculate the influence of a given trinucleotide starting at a given position on the bin’s
coverage. This simply averages the GC weight of the position of interest and the following 2 positions. For the last
two positions, a weight of 0 is returned.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>GC_weights</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]) – result of get_GC_weights_binwide for the binsize and fragments of interest.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A list of length <cite>(2*max(fragments)+binsize)</cite>, the weight corresponding to the first position within the
bin itself is at index <cite>max(fragments)</cite>. Values represent influence of a trinucleotide starting at a given
position on the  bin’s coverage.</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-liquorice.utils.BiasModel">
<span id="liquorice-utils-biasmodel-module"></span><h3>liquorice.utils.BiasModel module<a class="headerlink" href="#module-liquorice.utils.BiasModel" title="Permalink to this headline">¶</a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="liquorice.utils.BiasModel.BiasModel">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">liquorice.utils.BiasModel.</span></span><span class="sig-name descname"><span class="pre">BiasModel</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">training_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df_to_correct</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">biasmodel_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'trained_biasmodel.joblib'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">features</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'all'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nr_of_bins_for_training_and_testing</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sklearn_model</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_jobs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filename_performance_metrics</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'biasmodel_performance_metrics.csv'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filename_feature_importances</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_binsize_as_feature</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/utils/BiasModel.html#BiasModel"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.utils.BiasModel.BiasModel" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>An object that can be used to train a machine learning model that predicts coverage based on bias
factors. The performance of the trained model can be tested, and it can be used to correct coverage values
by regressing out the influence of the bias factors.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>training_df</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code>]) – <cite>pandas.DataFrame</cite> used for training (and optionally for performance assessment) of the
model. Must contain the column <em>coverage</em> and all columns specified under <strong>features</strong>. Can be None if
<a class="reference internal" href="#liquorice.utils.BiasModel.BiasModel.train_biasmodel" title="liquorice.utils.BiasModel.BiasModel.train_biasmodel"><code class="xref py py-func docutils literal notranslate"><span class="pre">train_biasmodel()</span></code></a> won’t be called. Ignored in :func`train_biasmodel_2fold_CV_and_predict`.</p></li>
<li><p><strong>df_to_correct</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code>]) – <cite>pandas.DataFrame</cite> for which coverage should be corrected by the trained model. Must
contain the column <em>coverage</em> and all columns specified under <strong>features</strong>. Can be None if
<a class="reference internal" href="#liquorice.utils.BiasModel.BiasModel.get_table_with_corrected_coverage_using_trained_biasmodel" title="liquorice.utils.BiasModel.BiasModel.get_table_with_corrected_coverage_using_trained_biasmodel"><code class="xref py py-func docutils literal notranslate"><span class="pre">get_table_with_corrected_coverage_using_trained_biasmodel()</span></code></a> won’t be called.</p></li>
<li><p><strong>biasmodel_path</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Path to which the trained biasmodel should be saved to and/or loaded from. Must be a
.joblib file. Can be None if <a class="reference internal" href="#liquorice.utils.BiasModel.BiasModel.get_table_with_corrected_coverage_using_trained_biasmodel" title="liquorice.utils.BiasModel.BiasModel.get_table_with_corrected_coverage_using_trained_biasmodel"><code class="xref py py-func docutils literal notranslate"><span class="pre">get_table_with_corrected_coverage_using_trained_biasmodel()</span></code></a> won’t be
called, in that case, no biasmodel will be saved.</p></li>
<li><p><strong>features</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]]) – A list of bias factors that should be used as features for the machine learning model. Default
“all” sets all bias-factors as features: forward,reverse, and max mappability, di/trinucleotide factors and
GC-content. Bin size is not included by default, it can be added as a feature via <cite>use_binsize_as_feature</cite>.</p></li>
<li><p><strong>nr_of_bins_for_training_and_testing</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>]) – Subset the training_df to this many bins. Can speed up the
computation time, but using too few bins will make the model less precise. Can be None, then all bins will be
used.</p></li>
<li><p><strong>sklearn_model</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<a class="reference internal" href="#liquorice.utils.BiasModel.SklearnStyleRegressor" title="liquorice.utils.BiasModel.SklearnStyleRegressor"><code class="xref py py-class docutils literal notranslate"><span class="pre">SklearnStyleRegressor</span></code></a>]) – A regressor that implements to functions .fit() and .predict() (e.g. from <cite>sklearn</cite>).
Default of None means using sklearn.ensemble.RandomForestRegressor with prior standard-scaling, and with
default settings.</p></li>
<li><p><strong>n_jobs</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – How many jobs to run in parallel when training the model</p></li>
<li><p><strong>filename_performance_metrics</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]) – If <strong>test_fraction</strong> or <strong>cross_validate_k</strong> is set, save a .csv containing
the performance metrics (r2, MSE) to this path.</p></li>
<li><p><strong>filename_feature_importances</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]) – If set, save a .csv file containing the feature importances inferred from
the trained model to this path.</p></li>
<li><p><strong>use_binsize_as_feature</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – If True, include “bin size” as a feature for the model.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="liquorice.utils.BiasModel.BiasModel.get_table_with_corrected_coverage_using_trained_biasmodel">
<span class="sig-name descname"><span class="pre">get_table_with_corrected_coverage_using_trained_biasmodel</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/utils/BiasModel.html#BiasModel.get_table_with_corrected_coverage_using_trained_biasmodel"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.utils.BiasModel.BiasModel.get_table_with_corrected_coverage_using_trained_biasmodel" title="Permalink to this definition">¶</a></dt>
<dd><p>Predicts coverage based on <code class="xref py py-attr docutils literal notranslate"><span class="pre">features</span></code> in the DataFrame <code class="xref py py-attr docutils literal notranslate"><span class="pre">df_to_correct</span></code> and the biasmodel
under <code class="xref py py-attr docutils literal notranslate"><span class="pre">biasmodel_path</span></code>. Subtracts this prediction from the observed coverage to regress out the effect
of the bias-factors (i.e. <code class="xref py py-attr docutils literal notranslate"><span class="pre">features</span></code>) on the coverage.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code></p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Returns <code class="xref py py-attr docutils literal notranslate"><span class="pre">df_to_correct</span></code> with an additional column “corrected coverage”.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="liquorice.utils.BiasModel.BiasModel.train_biasmodel">
<span class="sig-name descname"><span class="pre">train_biasmodel</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/utils/BiasModel.html#BiasModel.train_biasmodel"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.utils.BiasModel.BiasModel.train_biasmodel" title="Permalink to this definition">¶</a></dt>
<dd><p>Train a machine learning model that predicts the values of the ‘coverage’ column based on the given features. If
<code class="xref py py-attr docutils literal notranslate"><span class="pre">test_fraction</span></code> is set, this fraction of the <code class="xref py py-attr docutils literal notranslate"><span class="pre">training_df</span></code> is set aside to evaluate R^2 and MSE
of the model. Prior to training and a potential train/test split, the <code class="xref py py-attr docutils literal notranslate"><span class="pre">training_df</span></code> is subsetted to
<code class="xref py py-attr docutils literal notranslate"><span class="pre">nr_of_bins_for_training_and_testing</span></code> if it is not <em>None</em>. Writes writes a <em>.joblib</em>
file containing the biasmodel to <code class="xref py py-attr docutils literal notranslate"><span class="pre">biasmodel_path</span></code> (unless it is <cite>None</cite>)</p>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="liquorice.utils.BiasModel.BiasModel.train_biasmodel_2fold_CV_and_predict">
<span class="sig-name descname"><span class="pre">train_biasmodel_2fold_CV_and_predict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">exclude_these_bin_nrs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/utils/BiasModel.html#BiasModel.train_biasmodel_2fold_CV_and_predict"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.utils.BiasModel.BiasModel.train_biasmodel_2fold_CV_and_predict" title="Permalink to this definition">¶</a></dt>
<dd><p>Train a machine learning model that predicts the values of the ‘coverage’ column based on the given features.
Will use each half of the <code class="xref py py-attr docutils literal notranslate"><span class="pre">df_to_correct</span></code> to train the model for predictions of the other half.
Ignores <code class="xref py py-attr docutils literal notranslate"><span class="pre">nr_of_bins_for_training_and_testing</span></code>, <code class="xref py py-attr docutils literal notranslate"><span class="pre">cross_validate_k</span></code> and writes returns the performance</p>
<blockquote>
<div><p>metrics and returns the dataframe with predictions.
Important: does not use :attr`training_df`.</p>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="liquorice.utils.BiasModel.SklearnStyleRegressor">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">liquorice.utils.BiasModel.</span></span><span class="sig-name descname"><span class="pre">SklearnStyleRegressor</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwds</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/utils/BiasModel.html#SklearnStyleRegressor"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.utils.BiasModel.SklearnStyleRegressor" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">typing_extensions.Protocol</span></code></p>
<p>Introduced here for typing purposes only.</p>
<dl class="py method">
<dt class="sig sig-object py" id="liquorice.utils.BiasModel.SklearnStyleRegressor.fit">
<span class="sig-name descname"><span class="pre">fit</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample_weight</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/utils/BiasModel.html#SklearnStyleRegressor.fit"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.utils.BiasModel.SklearnStyleRegressor.fit" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="liquorice.utils.BiasModel.SklearnStyleRegressor.predict">
<span class="sig-name descname"><span class="pre">predict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/utils/BiasModel.html#SklearnStyleRegressor.predict"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.utils.BiasModel.SklearnStyleRegressor.predict" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="liquorice.utils.BiasModel.SklearnStyleRegressor.score">
<span class="sig-name descname"><span class="pre">score</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/utils/BiasModel.html#SklearnStyleRegressor.score"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.utils.BiasModel.SklearnStyleRegressor.score" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="liquorice.utils.BiasModel.SklearnStyleRegressor.set_params">
<span class="sig-name descname"><span class="pre">set_params</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/utils/BiasModel.html#SklearnStyleRegressor.set_params"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.utils.BiasModel.SklearnStyleRegressor.set_params" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

</div>
<div class="section" id="module-liquorice.utils.BinTableBiasFactors">
<span id="liquorice-utils-bintablebiasfactors-module"></span><h3>liquorice.utils.BinTableBiasFactors module<a class="headerlink" href="#module-liquorice.utils.BinTableBiasFactors" title="Permalink to this headline">¶</a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="liquorice.utils.BinTableBiasFactors.BiasFactorHandler">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">liquorice.utils.BinTableBiasFactors.</span></span><span class="sig-name descname"><span class="pre">BiasFactorHandler</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">binsize</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fragments</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">readlength</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_cores</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">skip_these_biasfactors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/utils/BinTableBiasFactors.html#BiasFactorHandler"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.utils.BinTableBiasFactors.BiasFactorHandler" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Object used for calculation of per-bin bias factors. Typically, after creation of the object a user would call
its method <code class="xref py py-func docutils literal notranslate"><span class="pre">get_table_with_bias_factors()</span></code> on it, and save the returned <cite>DataFrame</cite> for subsequent analysis
and correction of associations of bias factors with coverage.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>binsize</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Size of the bins. Higher values to reduce noise, lower values increase spatial
resolution.</p></li>
<li><p><strong>fragments</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>]) – A list containing fragment lengths that are representative of the sample’s global fragment
size distribution. Typically a few hundred fragments will suffice here.</p></li>
<li><p><strong>readlength</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Average length of reads in the .bam file.</p></li>
<li><p><strong>df</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code>) – <cite>pandas.DataFrame</cite> with one row per bin, containing columns “chromosome”, “start”, “end”,
“bin nr.”, “coverage”, “sequence”, and “mappability”. Suitable input is the output of the
<code class="xref py py-func docutils literal notranslate"><span class="pre">get_complete_table()</span></code> method of the <a class="reference internal" href="#module-liquorice.utils.CoverageAndSequenceTablePreparation" title="liquorice.utils.CoverageAndSequenceTablePreparation"><code class="xref py py-attr docutils literal notranslate"><span class="pre">liquorice.utils.CoverageAndSequenceTablePreparation</span></code></a> class
object.</p></li>
<li><p><strong>n_cores</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Max number of cores to use for calculations.</p></li>
<li><p><strong>skip_these_biasfactors</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]) – Do not calculate these bias factors. Only these entries are allowed:
[“di and trinucleotides and GC content”,”mappability”, “di and trinucleotides”]</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="liquorice.utils.BinTableBiasFactors.BiasFactorHandler.get_GC_and_di_and_trinuc_weights">
<span class="sig-name descname"><span class="pre">get_GC_and_di_and_trinuc_weights</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequence</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/utils/BinTableBiasFactors.html#BiasFactorHandler.get_GC_and_di_and_trinuc_weights"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.utils.BinTableBiasFactors.BiasFactorHandler.get_GC_and_di_and_trinuc_weights" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate bias factors for GC-content as well as bias factors for all di- and trinucleotides
(reverse and forward complements merged into single factors) for a given sequence. Factors are scaled between
0 and 1, 1 is highest (e.g. GC content: 0.. no G or C, 0.461… average GC content, 1… only G and C).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>sequence</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Genomic sequence of the bin extended by <em>max(</em> <code class="xref py py-attr docutils literal notranslate"><span class="pre">fragments</span></code> <em>)</em> in both directions,
such that its length matches <code class="xref py py-attr docutils literal notranslate"><span class="pre">GC_weights</span></code>.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A dictionary with entries corresponding to the bias factors for GC-content as well as bias factors
for all di- and trinucleotides (reverse and forward complements merged into single factors).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="liquorice.utils.BinTableBiasFactors.BiasFactorHandler.get_GC_bias_factor">
<span class="sig-name descname"><span class="pre">get_GC_bias_factor</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequence</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/utils/BinTableBiasFactors.html#BiasFactorHandler.get_GC_bias_factor"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.utils.BinTableBiasFactors.BiasFactorHandler.get_GC_bias_factor" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a number in the range 0-1 that represent’s the bin’s overall GC bias. Factors are scaled between
0 and 1, 1 is highest (e.g. GC content: 0.. no G or C, 0.461… average GC content, 1… only G and C).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>sequence</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Genomic sequence of the bin extended by <em>max(</em> <code class="xref py py-attr docutils literal notranslate"><span class="pre">fragments</span></code> <em>)</em> in both directions,
such that its length matches <code class="xref py py-attr docutils literal notranslate"><span class="pre">GC_weights</span></code>.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="liquorice.utils.BinTableBiasFactors.BiasFactorHandler.get_fwd_mappability_bias_factor">
<span class="sig-name descname"><span class="pre">get_fwd_mappability_bias_factor</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mappability</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/utils/BinTableBiasFactors.html#BiasFactorHandler.get_fwd_mappability_bias_factor"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.utils.BinTableBiasFactors.BiasFactorHandler.get_fwd_mappability_bias_factor" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a number representing the bin’s mappability for fragments on the forward strand.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>mappability</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]) – A vector of values between 0 and 1, representing the mappbility of positions in and around
the bin of interest. Must correspond to the coordinates &lt;Bin start coord&gt; - <em>max(</em> <code class="xref py py-attr docutils literal notranslate"><span class="pre">fragments</span></code> <em>)</em> -
<code class="xref py py-attr docutils literal notranslate"><span class="pre">readlength</span></code> to &lt;Bin end coord&gt; + <em>max(</em> <code class="xref py py-attr docutils literal notranslate"><span class="pre">fragments</span></code> <em>)</em> + <code class="xref py py-attr docutils literal notranslate"><span class="pre">readlength</span></code></p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A number in range 0-1 that represent’s the bin’s overall forward mappability. 1 … highest</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="liquorice.utils.BinTableBiasFactors.BiasFactorHandler.get_rev_mappability_bias_factor">
<span class="sig-name descname"><span class="pre">get_rev_mappability_bias_factor</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mappability</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/utils/BinTableBiasFactors.html#BiasFactorHandler.get_rev_mappability_bias_factor"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.utils.BinTableBiasFactors.BiasFactorHandler.get_rev_mappability_bias_factor" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a number representing the bin’s mappability for fragments on the reverse strand.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>mappability</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]) – A vector of values between 0 and 1, representing the mappbility of positions in and around
the bin of interest. Must correspond to the coordinates &lt;Bin start coord&gt; - <em>max(</em> <code class="xref py py-attr docutils literal notranslate"><span class="pre">fragments</span></code> <em>)</em> -
<code class="xref py py-attr docutils literal notranslate"><span class="pre">readlength</span></code> to &lt;Bin end coord&gt; + <em>max(</em> <code class="xref py py-attr docutils literal notranslate"><span class="pre">fragments</span></code> <em>)</em> + <code class="xref py py-attr docutils literal notranslate"><span class="pre">readlength</span></code></p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A number in range 0-1 that represent’s the bin’s overall reverse mappability. 1 … highest</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="liquorice.utils.BinTableBiasFactors.BiasFactorHandler.get_table_with_bias_factors">
<span class="sig-name descname"><span class="pre">get_table_with_bias_factors</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/utils/BinTableBiasFactors.html#BiasFactorHandler.get_table_with_bias_factors"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.utils.BinTableBiasFactors.BiasFactorHandler.get_table_with_bias_factors" title="Permalink to this definition">¶</a></dt>
<dd><p>Main method to retrieve bias factor information.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code></p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><p>A <cite>pandas.DataFrame</cite> with one row per bin, containing columns of the input DataFrame <code class="xref py py-attr docutils literal notranslate"><span class="pre">df</span></code>
(“chromosome”, “start”, “end”, “bin nr.”, “coverage”, “sequence”) as well as newly added
columns for bias
factors for mappability, GC-content, and  all di- and trinucleotides (with reverse and forward complements
merged into single factors). Bias factors are scaled between 0 and 1. Higher values correspond to higher
nucleotide content, mappability etc. Example: An average bin is expected to have a bias factor of 0.461
for humans (i.e. genomic GC content).
Note that the input columns “mappability” and “sequence” will not be part of the returned dataframe</p>
<blockquote>
<div><p>in order to reduce the memory usage of this function.</p>
</div></blockquote>
</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-liquorice.utils.CorrectForCNAs">
<span id="liquorice-utils-correctforcnas-module"></span><h3>liquorice.utils.CorrectForCNAs module<a class="headerlink" href="#module-liquorice.utils.CorrectForCNAs" title="Permalink to this headline">¶</a></h3>
<dl class="py function">
<dt class="sig sig-object py" id="liquorice.utils.CorrectForCNAs.correct_coverage_per_bin_for_cnas">
<span class="sig-prename descclassname"><span class="pre">liquorice.utils.CorrectForCNAs.</span></span><span class="sig-name descname"><span class="pre">correct_coverage_per_bin_for_cnas</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cna_seg_filepath</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_cores</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/utils/CorrectForCNAs.html#correct_coverage_per_bin_for_cnas"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.utils.CorrectForCNAs.correct_coverage_per_bin_for_cnas" title="Permalink to this definition">¶</a></dt>
<dd><p>Extracts the log2(observed/expected) read depth ratio for the corresponding segment for each bin, and corrects each
bin’s coverage accordingly.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code>) – A <cite>pandas.DataFrame</cite> with at least the following columns: “chromosome”, “start”,”end”,”coverage”</p></li>
<li><p><strong>cna_seg_filepath</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – A .seg file that should be used to correct the coverage by the values specified in
this file.  File must be tab-separated, with column names as first line. The second, third and fourth
column must be chromosome, start, and end of the segment, and the last column must be the log2-ratio of
observed/expected read depth.</p></li>
<li><p><strong>n_cores</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – How many cores to use to parallelize the operation.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code></p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A <cite>pandas.DataFrame</cite> similar to the input <em>df</em>, with the values in the column
“coverage” changed according to the CNA correction factor of each bin, and an additional column
“CNA-uncorrected coverage” that contains the original coverage values.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="liquorice.utils.CorrectForCNAs.get_CNV_corrected_coverage_for_bin">
<span class="sig-prename descclassname"><span class="pre">liquorice.utils.CorrectForCNAs.</span></span><span class="sig-name descname"><span class="pre">get_CNV_corrected_coverage_for_bin</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bin_chrom</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bin_start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bin_end</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bin_coverage</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seg_df</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/utils/CorrectForCNAs.html#get_CNV_corrected_coverage_for_bin"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.utils.CorrectForCNAs.get_CNV_corrected_coverage_for_bin" title="Permalink to this definition">¶</a></dt>
<dd><p>For a given bin, returns the coverage corrected for copy number aberrations.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>bin_chrom</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Chromosome of the bin.</p></li>
<li><p><strong>bin_start</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Genomic start position of the bin.</p></li>
<li><p><strong>bin_end</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Genomic end position of the bin.</p></li>
<li><p><strong>bin_coverage</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – Bin’s coverage before CNA correction</p></li>
<li><p><strong>seg_df</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code>) – A <cite>pandas.DataFrame</cite> of a .seg file, with the following columns: “chromosome”, “start”,”end”,
“1/log2_correction_factor”.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A float indicating the CNV-corrected coverage for the bin.</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-liquorice.utils.CoverageAndSequenceTablePreparation">
<span id="liquorice-utils-coverageandsequencetablepreparation-module"></span><h3>liquorice.utils.CoverageAndSequenceTablePreparation module<a class="headerlink" href="#module-liquorice.utils.CoverageAndSequenceTablePreparation" title="Permalink to this headline">¶</a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="liquorice.utils.CoverageAndSequenceTablePreparation.CoverageAndSequenceTablePreparation">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">liquorice.utils.CoverageAndSequenceTablePreparation.</span></span><span class="sig-name descname"><span class="pre">CoverageAndSequenceTablePreparation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bam_filepath</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bins_bed_filepath</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">refgenome_filepath</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">refgenome_chromsizes_filepath</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">refgenome_mappability_bigwig_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">readlength</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">longest_fraglen</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mean_seq_depth</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_cores</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_mapq</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">20</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">skip_these_steps</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">additional_crpb_kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/utils/CoverageAndSequenceTablePreparation.html#CoverageAndSequenceTablePreparation"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.utils.CoverageAndSequenceTablePreparation.CoverageAndSequenceTablePreparation" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Object used to set up a <cite>pandas.DataFrame</cite> containing basic information about each bin: Coverage (normalized for
overall sequencing depth), genomic sequence &amp; mappability (both incl. surrounding regions; used for bias
correction). Usually, a user would want to call <code class="xref py py-func docutils literal notranslate"><span class="pre">get_complete_table()</span></code> on this object, and save the
resulting <cite>pandas.DataFrame</cite> for subsequent bias-correction.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>bam_filepath</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Path to the .bam file containing the mapped reads for the sample. Does not need to be
duplicate-filtered, this is done by the function that calculates the coverage.</p></li>
<li><p><strong>bins_bed_filepath</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Path to the .bed file containing chromosome, start, end, and bin nr. of every bin to be
analyzed (in that order). Output of
<code class="xref py py-func docutils literal notranslate"><span class="pre">liquorice.utils.RegionFilteringAndBinning.Binning.write_bin_bedfile()</span></code> is a suitable input here.</p></li>
<li><p><strong>refgenome_filepath</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Path to the reference genome .fa(.gz) file. Must have a .fai index in the same dirname.</p></li>
<li><p><strong>refgenome_chromsizes_filepath</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Path to a tab-delimited file containing the chromosome sizes for the
reference genome. The first column must be the chromosome name, the second column its size.
Can be the .fa.fai file associated to the reference genome.</p></li>
<li><p><strong>refgenome_mappability_bigwig_path</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Path to a .bigWig file containing (forward) mappability values scaled
between 0 and 1.</p></li>
<li><p><strong>readlength</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – (Average) length of the reads in the .bam file.</p></li>
<li><p><strong>longest_fraglen</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Length by which sequencing and mappability information is extended beyond a bin’s
borders. Typically, this should be the longest sampled fragment length.</p></li>
<li><p><strong>min_mapq</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Minimum mapping quality for a fragment to still be counted when calculating the coverage.</p></li>
<li><p><strong>n_cores</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Number of cores to be used by deeptool’s <cite>CountReadsPerBin</cite>, which is used for coverage
calculations. Set to higher values to speed up the process.</p></li>
<li><p><strong>mean_seq_depth</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – A float that quantifies the global coverage/sequencing depth. Coverages per bin will be
normalized (i.e. divided) by this value.</p></li>
<li><p><strong>skip_these_steps</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]) – A list containing the steps which should be skipped. May contain: “coverage”,”sequence”,
“mappability”. Default []: Don’t skip any steps.</p></li>
<li><p><strong>**additional_crpb_kwargs</strong> – <p>Additional keywords to be used for deeptool’s <cite>CountReadsPerBin</cite>. Can be used
to override the default settings: <em>“ignoreDuplicates”:True,  “samFlag_exclude”:256,  “extendReads”:True</em></p>
</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="liquorice.utils.CoverageAndSequenceTablePreparation.CoverageAndSequenceTablePreparation.get_complete_table">
<span class="sig-name descname"><span class="pre">get_complete_table</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/utils/CoverageAndSequenceTablePreparation.html#CoverageAndSequenceTablePreparation.get_complete_table"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.utils.CoverageAndSequenceTablePreparation.CoverageAndSequenceTablePreparation.get_complete_table" title="Permalink to this definition">¶</a></dt>
<dd><p>Main method to retrieve a <cite>pandas.DateFrame</cite> that can be used to calculate bias factors on.
Calls <code class="xref py py-func docutils literal notranslate"><span class="pre">read_bed',</span> <span class="pre">:func:()</span></code>.get_coverage_per_bin_mean_over_per_basepair_values_chunked`,
<code class="xref py py-func docutils literal notranslate"><span class="pre">get_sequence_per_bin()</span></code>, and
<code class="xref py py-func docutils literal notranslate"><span class="pre">get_mappability_per_bin',</span> <span class="pre">constructs</span> <span class="pre">:attr:()</span></code>.df`, a <cite>pandas.DataFrame</cite>, from the results, and
returns it.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code></p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A <cite>pandas.DataFrame</cite> where rows correspond to the genomic bins in <code class="xref py py-attr docutils literal notranslate"><span class="pre">bins_bed_filepath</span></code>.
This <cite>DataFrame</cite> has the following columns: “chromosome”,”start”,”end”,”bin nr.”,”coverage”,”sequence”,
“mappability”.
Entries in “sequence” and “mappability” columns are extended by <code class="xref py py-attr docutils literal notranslate"><span class="pre">longest_fraglen</span></code>  (and downstream
additionally by <code class="xref py py-attr docutils literal notranslate"><span class="pre">readlength</span></code> for “mappability”).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="liquorice.utils.CoverageAndSequenceTablePreparation.CoverageAndSequenceTablePreparation.get_coverage_per_bin">
<span class="sig-name descname"><span class="pre">get_coverage_per_bin</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/utils/CoverageAndSequenceTablePreparation.html#CoverageAndSequenceTablePreparation.get_coverage_per_bin"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.utils.CoverageAndSequenceTablePreparation.CoverageAndSequenceTablePreparation.get_coverage_per_bin" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates the normalized coverage per bin, based on <code class="xref py py-attr docutils literal notranslate"><span class="pre">bins_bed_filepath</span></code> and <code class="xref py py-attr docutils literal notranslate"><span class="pre">bam_filepath</span></code>.
Uses deeptool’s <cite>countReadsPerBin</cite> for calculating the coverage in every bin in <code class="xref py py-attr docutils literal notranslate"><span class="pre">df</span></code>.
The obtained coverage value is then divided by <code class="xref py py-attr docutils literal notranslate"><span class="pre">mean_seq_depth</span></code>.
This function is faster than <a class="reference internal" href="#liquorice.utils.CoverageAndSequenceTablePreparation.CoverageAndSequenceTablePreparation.get_coverage_per_bin_mean_over_per_basepair_values_chunked" title="liquorice.utils.CoverageAndSequenceTablePreparation.CoverageAndSequenceTablePreparation.get_coverage_per_bin_mean_over_per_basepair_values_chunked"><code class="xref py py-func docutils literal notranslate"><span class="pre">get_coverage_per_bin_mean_over_per_basepair_values_chunked()</span></code></a>, but provides
less accurate results: Rather than the mean over the coverage of each base-pair in the bin, the total number of
reads mapping to the bin (even partially mapping reads) are reported (after normalization for
<cite>:attr:.mean_seq_depth</cite>).</p>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">Series</span></code></p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A <cite>pandas.Series</cite> with the coverage per bin, normalized for total sequencing coverage.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="liquorice.utils.CoverageAndSequenceTablePreparation.CoverageAndSequenceTablePreparation.get_coverage_per_bin_mean_over_per_basepair_values_chunked">
<span class="sig-name descname"><span class="pre">get_coverage_per_bin_mean_over_per_basepair_values_chunked</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/utils/CoverageAndSequenceTablePreparation.html#CoverageAndSequenceTablePreparation.get_coverage_per_bin_mean_over_per_basepair_values_chunked"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.utils.CoverageAndSequenceTablePreparation.CoverageAndSequenceTablePreparation.get_coverage_per_bin_mean_over_per_basepair_values_chunked" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates the normalized coverage per bin by averaging the per-base coverage over all positions in the bin and
then deviding the result by <code class="xref py py-attr docutils literal notranslate"><span class="pre">mean_seq_depth</span></code>.
Requires the following columns in <code class="xref py py-attr docutils literal notranslate"><span class="pre">df</span></code>: “chromosome”,”start”,”end”,”bin nr.”,”region nr.”,”bin size”.
Uses deeptool’s <cite>countReadsPerBin</cite> for calculating the coverage in every bin in <code class="xref py py-attr docutils literal notranslate"><span class="pre">df</span></code>.
The obtained coverage value is then divided by <code class="xref py py-attr docutils literal notranslate"><span class="pre">mean_seq_depth</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">Series</span></code></p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A <cite>pandas.Series</cite> with the coverage per bin, normalized for total sequencing coverage.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="liquorice.utils.CoverageAndSequenceTablePreparation.CoverageAndSequenceTablePreparation.get_mappability_for_bin">
<span class="sig-name descname"><span class="pre">get_mappability_for_bin</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">row</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mapp_bw</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/utils/CoverageAndSequenceTablePreparation.html#CoverageAndSequenceTablePreparation.get_mappability_for_bin"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.utils.CoverageAndSequenceTablePreparation.CoverageAndSequenceTablePreparation.get_mappability_for_bin" title="Permalink to this definition">¶</a></dt>
<dd><p>For a single bin, get a list of (forward) mappability values in and around the bin based on its genomic
coordinates and a <cite>pyBigWig.bigWig</cite> file that contains the genome-wide mappability.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>row</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">Series</span></code>) – <cite>pandas.Series</cite> containing columns “chromosome”,”start”, and “end”.</p></li>
<li><p><strong>mapp_bw</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bigWigFile</span></code>) – <cite>pyBigWig.pyBigWig</cite> object with mappability info (i.e. result of <cite>pyBigWig.open()</cite>)</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>Mappability for the bin at base resolultion.
Extended downstream by (<code class="xref py py-attr docutils literal notranslate"><span class="pre">longest_fraglen</span></code> + <code class="xref py py-attr docutils literal notranslate"><span class="pre">readlength</span></code>) and upstream by <code class="xref py py-attr docutils literal notranslate"><span class="pre">longest_fraglen</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="liquorice.utils.CoverageAndSequenceTablePreparation.CoverageAndSequenceTablePreparation.get_mappability_per_bin">
<span class="sig-name descname"><span class="pre">get_mappability_per_bin</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/utils/CoverageAndSequenceTablePreparation.html#CoverageAndSequenceTablePreparation.get_mappability_per_bin"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.utils.CoverageAndSequenceTablePreparation.CoverageAndSequenceTablePreparation.get_mappability_per_bin" title="Permalink to this definition">¶</a></dt>
<dd><p>Opens a <cite>.bigWig</cite> file using <cite>pyBigWig</cite> and calls <code class="xref py py-func docutils literal notranslate"><span class="pre">get_mappability_for_bin()</span></code> for every bin.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">Series</span></code></p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A <cite>pandas.Series</cite> with mappability information per bin. Extended downstream by
(<code class="xref py py-attr docutils literal notranslate"><span class="pre">longest_fraglen</span></code> + <code class="xref py py-attr docutils literal notranslate"><span class="pre">readlength</span></code>) and upstream by <code class="xref py py-attr docutils literal notranslate"><span class="pre">longest_fraglen</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="liquorice.utils.CoverageAndSequenceTablePreparation.CoverageAndSequenceTablePreparation.get_sequence_per_bin">
<span class="sig-name descname"><span class="pre">get_sequence_per_bin</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/utils/CoverageAndSequenceTablePreparation.html#CoverageAndSequenceTablePreparation.get_sequence_per_bin"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.utils.CoverageAndSequenceTablePreparation.CoverageAndSequenceTablePreparation.get_sequence_per_bin" title="Permalink to this definition">¶</a></dt>
<dd><p>Get a list of genomic sequences, one per entry in the .bed file attr:<cite>bins_bed_filepath</cite>.
Sequences are extended by <code class="xref py py-attr docutils literal notranslate"><span class="pre">longest_fraglen</span></code></p>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A list of the genomic sequences of every bin in the attr:<cite>bins_bed_filepath</cite>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="liquorice.utils.CoverageAndSequenceTablePreparation.CoverageAndSequenceTablePreparation.read_bed">
<span class="sig-name descname"><span class="pre">read_bed</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/utils/CoverageAndSequenceTablePreparation.html#CoverageAndSequenceTablePreparation.read_bed"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.utils.CoverageAndSequenceTablePreparation.CoverageAndSequenceTablePreparation.read_bed" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads in a .bed file and return corresponding <cite>pandas.DataFrame</cite>. 4th column in the .bed file is interpreted
as bin nr. Layout of the .bed file MUST be [“chromosome”,”start”,”end”,”bin nr.”].</p>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code></p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A <cite>pandas.DataFrame</cite> with the columns [“chromosome”,”start”,”end”,”bin nr.”].</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="liquorice.utils.CoverageAndSequenceTablePreparation.parallelize_get_coverage_of_region">
<span class="sig-prename descclassname"><span class="pre">liquorice.utils.CoverageAndSequenceTablePreparation.</span></span><span class="sig-name descname"><span class="pre">parallelize_get_coverage_of_region</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_cores</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bam_filepath</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_mapq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">additional_crpb_kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/utils/CoverageAndSequenceTablePreparation.html#parallelize_get_coverage_of_region"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.utils.CoverageAndSequenceTablePreparation.parallelize_get_coverage_of_region" title="Permalink to this definition">¶</a></dt>
<dd><p>Splits the provided <cite>pandas.DataFrame</cite> into chunks of the size :param`n_cores`, and
runs :func`run_get_coverage_of_region_per_chunk` in parallel on every chunk.
From: <a class="reference external" href="https://towardsdatascience.com/make-your-own-super-pandas-using-multiproc-1c04f41944a1">https://towardsdatascience.com/make-your-own-super-pandas-using-multiproc-1c04f41944a1</a></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> – Full <cite>pandas.DataFrame</cite> with the columns “chromosome”,”region start”,”region end”.</p></li>
<li><p><strong>n_cores</strong> – Run this many processes in parallel.</p></li>
<li><p><strong>bam_filepath</strong> – Path to the .bam file containing the mapped reads for the sample. Does not need to be
duplicate-filtered, this is done by the function that calculates the coverage.</p></li>
<li><p><strong>min_mapq</strong> – Minimum mapping quality for a fragment to still be counted when calculating the coverage.</p></li>
<li><p><strong>additional_crpb_kwargs</strong> – Additional keywords to be used for deeptool’s <cite>CountReadsPerBin</cite>. Can be used
to override the default settings: <em>“ignoreDuplicates”:True,  “samFlag_exclude”:256,  “extendReads”:True</em></p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Full <cite>pandas.DataFrame</cite> with the columns “chromosome”,”region start”,”region end” and
“region coverage array”.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="liquorice.utils.CoverageAndSequenceTablePreparation.run_get_coverage_of_region_per_chunk">
<span class="sig-prename descclassname"><span class="pre">liquorice.utils.CoverageAndSequenceTablePreparation.</span></span><span class="sig-name descname"><span class="pre">run_get_coverage_of_region_per_chunk</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dataframe_chunk</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bam_filepath</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_mapq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_cores</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">additional_crpb_kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/utils/CoverageAndSequenceTablePreparation.html#run_get_coverage_of_region_per_chunk"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.utils.CoverageAndSequenceTablePreparation.run_get_coverage_of_region_per_chunk" title="Permalink to this definition">¶</a></dt>
<dd><p>Uses :func`deeptools.countReadsPerBin.get_coverage_of_region()` to calculate the coverage in every bin
in <a href="#id3"><span class="problematic" id="id4">:param:`dataframe_chunk`</span></a>. Adds the result as a column “region coverage array” and returns
param:<cite>dataframe_chunk</cite>.
This function can receive only picklable arguments, and can therefore be called multiple times in parallel.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dataframe_chunk</strong> – A <cite>pandas.DataFrame</cite> with the columns “chromosome”,”region start”,”region end”.
Usually, this would be only a chunk of the full DataFrame.</p></li>
<li><p><strong>bam_filepath</strong> – Path to the .bam file containing the mapped reads for the sample. Does not need to be
duplicate-filtered, this is done by the function that calculates the coverage.</p></li>
<li><p><strong>min_mapq</strong> – Minimum mapping quality for a fragment to still be counted when calculating the coverage.</p></li>
<li><p><strong>n_cores</strong> – Number of cores used by <cite>deeptools.countReadsPerBin</cite>. Should probably be 1 if the function is called
in parallel</p></li>
<li><p><strong>additional_crpb_kwargs</strong> – Additional keywords to be used for deeptool’s <cite>CountReadsPerBin</cite>. Can be used
to override the default settings: <em>“ignoreDuplicates”:True,  “samFlag_exclude”:256,  “extendReads”:True</em></p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A <cite>pandas.DataFrame</cite> with the columns “chromosome”,”region start”,”region end” and “region coverage array”.</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-liquorice.utils.FitGaussians">
<span id="liquorice-utils-fitgaussians-module"></span><h3>liquorice.utils.FitGaussians module<a class="headerlink" href="#module-liquorice.utils.FitGaussians" title="Permalink to this headline">¶</a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="liquorice.utils.FitGaussians.FitGaussians">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">liquorice.utils.FitGaussians.</span></span><span class="sig-name descname"><span class="pre">FitGaussians</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">unfitted_y_values</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extend_to</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">binsize</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">avg_centersize</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">samplename</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">regionset_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/utils/FitGaussians.html#FitGaussians"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.utils.FitGaussians.FitGaussians" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>An object that can be used for quantification of coverage drops by fitting gaussian functions and an intercept
to the coverage values.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>unfitted_y_values</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]) – A pandas <cite>Series</cite> of coverage values, one value per bin</p>
</dd>
</dl>
<p>(and sorted by increasing bin nr.). A suitable input is the result of
<code class="xref py py-func docutils literal notranslate"><span class="pre">liquorice.utils.AggregateAcrossrRegions.aggregate_across_regions()</span></code>.
:type extend_to: <code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>
:param extend_to: <cite>extend_to</cite> setting that was used when calculating the y values. Needed to calculate relative</p>
<blockquote>
<div><p>bin positions.</p>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>binsize</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – <cite>binsize</cite> setting that was used when calculating the y values. Needed to calculate relative
bin positions.</p></li>
<li><p><strong>samplename</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]) – If not <cite>None</cite>, add this in a column “sample” to the result <cite>DataFrame</cite> when calling
<code class="xref py py-func docutils literal notranslate"><span class="pre">fit_gaussian_models()</span></code></p></li>
<li><p><strong>regionset_name</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]) – If not <cite>None</cite>, add this in a column “region-set” to the result <cite>DataFrame</cite> when calling
<code class="xref py py-func docutils literal notranslate"><span class="pre">fit_gaussian_models()</span></code></p></li>
<li><p><strong>avg_centersize</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – Average size (i.e. width in bp) of the regions of interest (before extension by extend_to).</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="liquorice.utils.FitGaussians.FitGaussians.fit_gaussian_models">
<span class="sig-name descname"><span class="pre">fit_gaussian_models</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">g1_min_sigma</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">20</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">g1_max_sigma</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">200</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">g2_min_sigma</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">200</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">g2_max_sigma</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">g3_min_sigma</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">g3_max_sigma</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">40000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'leastsq'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/utils/FitGaussians.html#FitGaussians.fit_gaussian_models"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.utils.FitGaussians.FitGaussians.fit_gaussian_models" title="Permalink to this definition">¶</a></dt>
<dd><p>Fits three gaussian functions to <code class="xref py py-attr docutils literal notranslate"><span class="pre">unfitted_y_values</span></code> (usually aggregated coverage values), based on given
limits. Sets <code class="xref py py-attr docutils literal notranslate"><span class="pre">x_values</span></code>, <code class="xref py py-attr docutils literal notranslate"><span class="pre">g1_y_values</span></code>, <code class="xref py py-attr docutils literal notranslate"><span class="pre">g2_y_values</span></code>, <code class="xref py py-attr docutils literal notranslate"><span class="pre">g3_y_values</span></code>,
<code class="xref py py-attr docutils literal notranslate"><span class="pre">intercept_y_values</span></code> and <code class="xref py py-attr docutils literal notranslate"><span class="pre">combined_model_y_values</span></code> for the object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>g1_min_sigma</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]) – Min σ value for the smallest gaussian</p></li>
<li><p><strong>g1_max_sigma</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]) – Max σ value for the smallest gaussian</p></li>
<li><p><strong>g2_min_sigma</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]) – Min σ value for the middle gaussian</p></li>
<li><p><strong>g2_max_sigma</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]) – Max σ value for the middle gaussian</p></li>
<li><p><strong>g3_min_sigma</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]) – Min σ value for the widest gaussian</p></li>
<li><p><strong>g3_max_sigma</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]) – Max σ value for the widest gaussian</p></li>
<li><p><strong>method</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – A method for fitting the combined model, used as parameter in lmfit’s .fit() function.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code></p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A <cite>pandas.DataFrame</cite> containing several summary metrics based on the fitted model: The amplitude and σ
values for each of the three gaussians, the total depth of the fitted model at the central bin, the Baysian
Information Criterion (measures model fit), and the area under the curve of the fitted model,
calculated as the area between the fitted intercept and the fitted model.
If <code class="xref py py-attr docutils literal notranslate"><span class="pre">samplename</span></code> or <code class="xref py py-attr docutils literal notranslate"><span class="pre">regionset_name</span></code> are not <cite>None</cite>, columns “sample” or “region-set” are
added, respectively.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-liquorice.utils.GlobalFragmentSize">
<span id="liquorice-utils-globalfragmentsize-module"></span><h3>liquorice.utils.GlobalFragmentSize module<a class="headerlink" href="#module-liquorice.utils.GlobalFragmentSize" title="Permalink to this headline">¶</a></h3>
<dl class="py function">
<dt class="sig sig-object py" id="liquorice.utils.GlobalFragmentSize.getFragmentLength_worker">
<span class="sig-prename descclassname"><span class="pre">liquorice.utils.GlobalFragmentSize.</span></span><span class="sig-name descname"><span class="pre">getFragmentLength_worker</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">chrom</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bamFile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">distanceBetweenBins</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/utils/GlobalFragmentSize.html#getFragmentLength_worker"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.utils.GlobalFragmentSize.getFragmentLength_worker" title="Permalink to this definition">¶</a></dt>
<dd><p>This is a function from deeptools, modified from <a class="reference external" href="https://github.com/deeptools/deepTools/blob/ac42d29c298c026aa0c53c9db2553087ebc86b97/deeptools/getFragmentAndReadSize.py#L14">https://github.com/deeptools/deepTools/blob/ac42d29c298c026aa0c53c9db2553087ebc86b97/deeptools/getFragmentAndReadSize.py#L14</a>
Queries the reads at the given region for the distance between
reads and the read length. As opposed to the original version, does not disregard pairs that are flagged as
“not properly paired”, as this behaviour excludes fragments from the two-nucleosome peak of the cfDNA-distribution.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>chrom</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – chromosome name</p></li>
<li><p><strong>start</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – region start</p></li>
<li><p><strong>end</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – region end</p></li>
<li><p><strong>bamFile</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – BAM file name</p></li>
<li><p><strong>distanceBetweenBins</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The number of bases at the end of each bin to ignore</p></li>
</ul>
</dd>
</dl>
<dl class="simple">
<dt>:return an np.array, where first column is fragment length, the</dt><dd><p>second is for read length</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">array</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="liquorice.utils.GlobalFragmentSize.getFragmentLength_wrapper">
<span class="sig-prename descclassname"><span class="pre">liquorice.utils.GlobalFragmentSize.</span></span><span class="sig-name descname"><span class="pre">getFragmentLength_wrapper</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/utils/GlobalFragmentSize.html#getFragmentLength_wrapper"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.utils.GlobalFragmentSize.getFragmentLength_wrapper" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="liquorice.utils.GlobalFragmentSize.get_list_of_fragment_lengths_and_avg_readlength">
<span class="sig-prename descclassname"><span class="pre">liquorice.utils.GlobalFragmentSize.</span></span><span class="sig-name descname"><span class="pre">get_list_of_fragment_lengths_and_avg_readlength</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bam_filepath</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_cores</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">upper_limit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">800</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/utils/GlobalFragmentSize.html#get_list_of_fragment_lengths_and_avg_readlength"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.utils.GlobalFragmentSize.get_list_of_fragment_lengths_and_avg_readlength" title="Permalink to this definition">¶</a></dt>
<dd><p>Sample fragments from the given .bam file to obtain a representative distribution of fragment lenghts.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>bam_filepath</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Path to the .bam file for which fragments should be sampled.</p></li>
<li><p><strong>n_cores</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Number of cores to use by <code class="xref py py-func docutils literal notranslate"><span class="pre">deeptools.getFragmentAndReadSize.get_read_and_fragment_length()</span></code>.</p></li>
<li><p><strong>n</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Number of randomly sampled fragment lengths to generate</p></li>
<li><p><strong>upper_limit</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Fragment lengths exceeding this limit will be excluded. Rarely, fragment size are wrongly
inferred and therefore huge. Sampling one of those incorrect lengths would unnecessarily blow up the sequence
and mapping information stored for each bin. As a default, 500 is used as a reasonable upper limit of relevant
fragment lengths for cfDNA.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>], <code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>]</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A tuple, consisting of a list of the randomly samples fragment lengths, and an integer value of the
average read length.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="liquorice.utils.GlobalFragmentSize.get_read_and_fragment_length">
<span class="sig-prename descclassname"><span class="pre">liquorice.utils.GlobalFragmentSize.</span></span><span class="sig-name descname"><span class="pre">get_read_and_fragment_length</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bamFile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_lengths</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">blackListFileName</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">binSize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">50000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">distanceBetweenBins</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1000000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">numberOfProcessors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/utils/GlobalFragmentSize.html#get_read_and_fragment_length"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.utils.GlobalFragmentSize.get_read_and_fragment_length" title="Permalink to this definition">¶</a></dt>
<dd><p>This is a function from deeptools. It was included in LIQUORICE’s source code to allow it to (indirectly) call the
modified version of :func`getFragmentLength_worker`.
Estimates the fragment length and read length through sampling</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>bamFile</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – BAM file name</p></li>
<li><p><strong>return_lengths</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – </p></li>
<li><p><strong>numberOfProcessors</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>]) – </p></li>
<li><p><strong>verbose</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – </p></li>
<li><p><strong>binSize</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – </p></li>
<li><p><strong>distanceBetweenBins</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – </p></li>
</ul>
</dd>
</dl>
<dl class="simple">
<dt>.:return A tuple of two dictionaries, one for the fragment length and the other</dt><dd><p>for the read length. The dictionaries summarise the mean, median etc. values</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code>]</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-liquorice.utils.MeanSequencingDepth">
<span id="liquorice-utils-meansequencingdepth-module"></span><h3>liquorice.utils.MeanSequencingDepth module<a class="headerlink" href="#module-liquorice.utils.MeanSequencingDepth" title="Permalink to this headline">¶</a></h3>
<dl class="py function">
<dt class="sig sig-object py" id="liquorice.utils.MeanSequencingDepth.sample_bam_to_get_sequencing_depth">
<span class="sig-prename descclassname"><span class="pre">liquorice.utils.MeanSequencingDepth.</span></span><span class="sig-name descname"><span class="pre">sample_bam_to_get_sequencing_depth</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bam_filepath</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_sites_to_sample</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_cores</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_mapq</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">20</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chromosomes_list</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">additional_crpb_kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/utils/MeanSequencingDepth.html#sample_bam_to_get_sequencing_depth"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.utils.MeanSequencingDepth.sample_bam_to_get_sequencing_depth" title="Permalink to this definition">¶</a></dt>
<dd><p>Estimates the overall sequencing coverage of the .bam file by sampling
sites that are regularily distributed across the genome (using <code class="xref py py-func docutils literal notranslate"><span class="pre">deeptools.countReadsPerBin.CountReadsPerBin()</span></code>
in a slightly modified version that allows fragments without the SAM flag is_proper_pair.).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>bam_filepath</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Path to the .bam file for which mean coverage should be calculated.</p></li>
<li><p><strong>n_sites_to_sample</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Number of sites (length 1) to sample.</p></li>
<li><p><strong>n_cores</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Number of cores to be used by <code class="xref py py-func docutils literal notranslate"><span class="pre">deeptools.countReadsPerBin.CountReadsPerBin()</span></code>.</p></li>
<li><p><strong>min_mapq</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – <cite>minMappingQuality</cite> setting for <cite>deeptools.countReadsPerBin.CountReadsPerBin</cite>.</p></li>
<li><p><strong>chromosomes_list</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]]) – A list of chromosomes that should be analyzed. Regions on chromosomes that are not in
this list will be excluded from analysis. Default None means [“chr1”, …, “chr22”].</p></li>
<li><p><strong>additional_crpb_kwargs</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code>) – Use to override the default parameters for
<code class="xref py py-func docutils literal notranslate"><span class="pre">deeptools.countReadsPerBin.CountReadsPerBin()</span></code>. If not specified otherwise in <cite>additional_crpb_kwargs</cite>,
the following attributes are passed to <cite>CountReadsPerBin</cite>: <cite>ignoreDuplicates=True</cite>, <cite>samFlag_exclude=256</cite>,
<cite>extendReads=True</cite>.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The average coverage determined from the sampled regions.</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-liquorice.utils.Plotting">
<span id="liquorice-utils-plotting-module"></span><h3>liquorice.utils.Plotting module<a class="headerlink" href="#module-liquorice.utils.Plotting" title="Permalink to this headline">¶</a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="liquorice.utils.Plotting.Plotting">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">liquorice.utils.Plotting.</span></span><span class="sig-name descname"><span class="pre">Plotting</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">samplename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">out_dir</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/utils/Plotting.html#Plotting"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.utils.Plotting.Plotting" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>An object used to generate plots of the aggragated coverage, differences between corrected and uncorrected
coverage, as well as correlations between bias-factors and coverage.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>samplename</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Name of the sample as it should appear in the plots</p></li>
<li><p><strong>out_dir</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Path to the directory to which plots should be saved.</p></li>
</ul>
</dd>
</dl>
<dl class="py attribute">
<dt class="sig sig-object py" id="liquorice.utils.Plotting.Plotting.none_or_list_of_float">
<span class="sig-name descname"><span class="pre">none_or_list_of_float</span></span><a class="headerlink" href="#liquorice.utils.Plotting.Plotting.none_or_list_of_float" title="Permalink to this definition">¶</a></dt>
<dd><p>alias of <code class="xref py py-obj docutils literal notranslate"><span class="pre">Union</span></code>[<code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]]</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="liquorice.utils.Plotting.Plotting.none_or_list_of_int">
<span class="sig-name descname"><span class="pre">none_or_list_of_int</span></span><a class="headerlink" href="#liquorice.utils.Plotting.Plotting.none_or_list_of_int" title="Permalink to this definition">¶</a></dt>
<dd><p>alias of <code class="xref py py-obj docutils literal notranslate"><span class="pre">Union</span></code>[<code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]]</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="liquorice.utils.Plotting.Plotting.none_or_list_of_str">
<span class="sig-name descname"><span class="pre">none_or_list_of_str</span></span><a class="headerlink" href="#liquorice.utils.Plotting.Plotting.none_or_list_of_str" title="Permalink to this definition">¶</a></dt>
<dd><p>alias of <code class="xref py py-obj docutils literal notranslate"><span class="pre">Union</span></code>[<code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]]</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="liquorice.utils.Plotting.Plotting.plot_corrected_vs_uncorrected_coverage">
<span class="sig-name descname"><span class="pre">plot_corrected_vs_uncorrected_coverage</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x_values</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">uncorrected_y_values</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">corrected_y_values</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">uncorrected_y_intercept</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">corrected_y_intercept</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filename</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'auto'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_figure</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/utils/Plotting.html#Plotting.plot_corrected_vs_uncorrected_coverage"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.utils.Plotting.Plotting.plot_corrected_vs_uncorrected_coverage" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots the aggregated coverages, both corrected and uncorrected.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x_values</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>]) – List of x coordinates - the coordinates of the bins’ centers relative to the center of the
regions of interest.</p></li>
<li><p><strong>uncorrected_y_values</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]) – List of aggregated, uncorrected coverage values.</p></li>
<li><p><strong>corrected_y_values</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]) – List of aggregated, uncorrected coverage values.</p></li>
<li><p><strong>uncorrected_y_intercept</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – Intercept value that should be subtracted from every value in
<strong>uncorrected_y_values</strong></p></li>
<li><p><strong>corrected_y_intercept</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – Intercept value that should be subtracted from every value in
<strong>corrected_y_values</strong></p></li>
<li><p><strong>filename</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]) – Filename to which the figure should be saved. “auto” sets this to
” <code class="xref py py-attr docutils literal notranslate"><span class="pre">out_dir</span></code>/corrected_vs_uncorrected_coverage.pdf “. Set to <cite>None</cite> to avoid saving as
file. File extension determines in which format the plot is saved.</p></li>
<li><p><strong>return_figure</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – If True, return a <cite>matplotlib figure</cite> that can be altered, plotted, or saved.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">Figure</span></code>]</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A <cite>matplotlib figure</cite> if <strong>return_figure</strong> is True, nothing otherwise.</p>
</dd>
</dl>
<p>Example:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">plotting</span><span class="o">.</span><span class="n">plot_corrected_vs_uncorrected_coverage</span><span class="p">(</span>
    <span class="n">x_values</span><span class="o">=</span><span class="n">fit_gaussians_obj</span><span class="o">.</span><span class="n">x_values</span><span class="p">,</span>
    <span class="n">uncorrected_y_values</span><span class="o">=</span><span class="n">fit_gaussians_obj_uncorrected</span><span class="o">.</span><span class="n">unfitted_y_values</span><span class="p">,</span>
    <span class="n">corrected_y_values</span><span class="o">=</span><span class="n">fit_gaussians_obj</span><span class="o">.</span><span class="n">unfitted_y_values</span><span class="p">,</span>
    <span class="n">uncorrected_y_intercept</span><span class="o">=</span><span class="n">fit_gaussians_obj_uncorrected</span><span class="o">.</span><span class="n">intercept_y_values</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
    <span class="n">corrected_y_intercept</span><span class="o">=</span><span class="n">fit_gaussians_obj</span><span class="o">.</span><span class="n">intercept_y_values</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="liquorice.utils.Plotting.Plotting.plot_correlations_biasfactors_coverage_bin_nr">
<span class="sig-name descname"><span class="pre">plot_correlations_biasfactors_coverage_bin_nr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x_varnames_list</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y_varnames_list</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bins</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(50,</span> <span class="pre">50)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_figure</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filename</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'bias_plots.pdf'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">percentile_cutoffs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(0.1,</span> <span class="pre">0.99,</span> <span class="pre">0.1,</span> <span class="pre">0.99)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fit_regression</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/utils/Plotting.html#Plotting.plot_correlations_biasfactors_coverage_bin_nr"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.utils.Plotting.Plotting.plot_correlations_biasfactors_coverage_bin_nr" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot the correlation between coverage and a bias factor as a heatmap, and fit a trendline.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code>) – <cite>pandas.DataFrame</cite> that contains the variables in <strong>x_varnames_list</strong> and  <strong>y_varnames_list</strong>.</p></li>
<li><p><strong>x_varnames_list</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]]) – List of the columns that should be plotted as independent variables. Default None uses
all columns in the <strong>df</strong> except for [“chromosome”,”start”,”end”,”sequence”,”mappability”].</p></li>
<li><p><strong>y_varnames_list</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]]) – List of the columns that should be plotted as dependent variables. Default None uses
“coverage” and “bin nr.”, plus “corrected coverage” if this column is present in the <strong>df</strong>.</p></li>
<li><p><strong>bins</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>]) – Tuple specifying the number of bins to be used for the x and y axis</p></li>
<li><p><strong>return_figure</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – If True, return a <cite>matplotlib figure</cite> that can be altered, plotted, or saved.</p></li>
<li><p><strong>fit_regression</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – If True, fit a qadratic polynomial and plot the regression function and R^2.</p></li>
<li><p><strong>percentile_cutoffs</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>]) – Tuple containing the min and max percentile of values that should be displayed on the
x and y axes: (x_min,x_max,y_min,y_max). This setting is ignored for “bin nr.”, for this column, all values
are shown. Percentiles must be in the interval [0,1].</p></li>
<li><p><strong>filename</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Filename to which the figure should be saved. File extension determines in which format the
plot is saved.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">Figure</span></code>]</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A <cite>matplotlib figure</cite> if <strong>return_figure</strong> is True, nothing otherwise.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="liquorice.utils.Plotting.Plotting.plot_coverage_bias_correlation">
<span class="sig-name descname"><span class="pre">plot_coverage_bias_correlation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">biasfactor_column</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">coverage_column</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">percent</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xmin</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xmax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ymin</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ymax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filename</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'auto'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_figure</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/utils/Plotting.html#Plotting.plot_coverage_bias_correlation"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.utils.Plotting.Plotting.plot_coverage_bias_correlation" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot the correlation between coverage and a bias factor as a heatmap, and fit a trendline.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code>) – <cite>pandas.DataFrame</cite> that contains <strong>biasfactor_column</strong> and  <strong>coverage_column</strong>.</p></li>
<li><p><strong>biasfactor_column</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Name of the column that contains data for the bias-factor of interest</p></li>
<li><p><strong>coverage_column</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Name of the column that contains the coverage data</p></li>
<li><p><strong>percent</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – Whether the output plot should multiply the x values by the factor 100</p></li>
<li><p><strong>xmin</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]) – min x coordinate to show in plot</p></li>
<li><p><strong>xmax</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]) – max y coordinate to show in plot</p></li>
<li><p><strong>ymin</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]) – min x coordinate to show in plot</p></li>
<li><p><strong>ymax</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]) – max y coordinate to show in plot</p></li>
<li><p><strong>filename</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]) – Filename to which the figure should be saved. “auto” sets this to
” <code class="xref py py-attr docutils literal notranslate"><span class="pre">out_dir</span></code>/<strong>biasfactor_column**__vs__**coverage_column</strong>.pdf ” (with spaces in column names
replaced by ‘_’). Set to <cite>None</cite> to avoid saving as
file. File extension determines in which format the plot is saved.</p></li>
<li><p><strong>return_figure</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – If True, return a <cite>matplotlib figure</cite> that can be altered, plotted, or saved.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">Figure</span></code>]</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A <cite>matplotlib figure</cite> if <strong>return_figure</strong> is True, nothing otherwise.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="liquorice.utils.Plotting.Plotting.plot_fitted_model">
<span class="sig-name descname"><span class="pre">plot_fitted_model</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fit_gaussians_obj</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x_values</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unfitted_y_values</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">g1_y_values</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">g2_y_values</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">g3_y_values</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">intercept_y_values</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">combined_model_y_values</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unfitted_y_values_label</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Corrected,</span> <span class="pre">aggregated</span> <span class="pre">coverage'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filename</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'auto'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_figure</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/utils/Plotting.html#Plotting.plot_fitted_model"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.utils.Plotting.Plotting.plot_fitted_model" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots the fitted model, i.e. the raw data and lines for G1, G2, G3, the intercept, and the combined model.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>fit_gaussians_obj</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<a class="reference internal" href="#liquorice.utils.FitGaussians.FitGaussians" title="liquorice.utils.FitGaussians.FitGaussians"><code class="xref py py-class docutils literal notranslate"><span class="pre">FitGaussians</span></code></a>]) – object of class FitGaussians, on which fit_gaussian_models() has been called. This
contains all nessesary information for plotting, so all other parameters except <strong>filename</strong> are ignored if
this is set.</p></li>
<li><p><strong>x_values</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]]) – List of x coordinates - the coordinates of the bins’ centers relative to the center of the
region of interest. Required if fit_gaussians_obj is not set, ignored otherwise.</p></li>
<li><p><strong>unfitted_y_values</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]]) – List of y coordinates of the unfitted coverage data.
Required if fit_gaussians_obj is not set, ignored otherwise.</p></li>
<li><p><strong>g1_y_values</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]]) – List of y coordinates for the first, smallest gaussian.
Required if fit_gaussians_obj is not set, ignored otherwise.</p></li>
<li><p><strong>g2_y_values</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]]) – List of y coordinates for the second, middle gaussian.
Required if fit_gaussians_obj is not set, ignored otherwise.</p></li>
<li><p><strong>g3_y_values</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]]) – List of y coordinates for the first, largest gaussian.
Required if fit_gaussians_obj is not set, ignored otherwise.</p></li>
<li><p><strong>intercept_y_values</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]]) – List of y coordinates for the intercept.
Required if fit_gaussians_obj is not set, ignored otherwise.</p></li>
<li><p><strong>combined_model_y_values</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]]) – List of y coordinates for the fitted, combined model.
Required if fit_gaussians_obj is not set, ignored otherwise.</p></li>
<li><p><strong>unfitted_y_values_label</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]]) – Label for the unfitted_y_values. By default it is assumed that the corrected,
aggregated, unfitted coverage is given.</p></li>
<li><p><strong>filename</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]) – Filename to which the figure should be saved. “auto” sets this to
” <code class="xref py py-attr docutils literal notranslate"><span class="pre">out_dir</span></code>/fitted_gaussians.pdf “. Set to <cite>None</cite> to avoid saving as
file. File extension determines in which format the plot is saved.</p></li>
<li><p><strong>return_figure</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – If True, return a <cite>matplotlib figure</cite> that can be altered, plotted, or saved.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">Figure</span></code>]</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A <cite>matplotlib figure</cite> if <strong>return_figure</strong> is True, nothing otherwise.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="liquorice.utils.Plotting.polyfit">
<span class="sig-prename descclassname"><span class="pre">liquorice.utils.Plotting.</span></span><span class="sig-name descname"><span class="pre">polyfit</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">degree</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/utils/Plotting.html#polyfit"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.utils.Plotting.polyfit" title="Permalink to this definition">¶</a></dt>
<dd><p>fit a trendline to data, see <a class="reference external" href="https://stackoverflow.com/questions/893657/how-do-i-calculate-r-squared-using-python-and-numpy">https://stackoverflow.com/questions/893657/how-do-i-calculate-r-squared-using-python-and-numpy</a></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]) – x values</p></li>
<li><p><strong>y</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]) – y values</p></li>
<li><p><strong>degree</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – 1 for linear, 2 for quadratic, etc.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A Dictionary, containing lists with the keys ‘polynomial’ (polynomial coefficients) and ‘determination’
(The coefficient of determination).</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-liquorice.utils.RegionFilteringAndBinning">
<span id="liquorice-utils-regionfilteringandbinning-module"></span><h3>liquorice.utils.RegionFilteringAndBinning module<a class="headerlink" href="#module-liquorice.utils.RegionFilteringAndBinning" title="Permalink to this headline">¶</a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="liquorice.utils.RegionFilteringAndBinning.RegionFilteringAndBinning">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">liquorice.utils.RegionFilteringAndBinning.</span></span><span class="sig-name descname"><span class="pre">RegionFilteringAndBinning</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bed_filepath</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">binsize</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extend_to</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">refgenome_chromsizes_filepath</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chromosomes_list</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['chr1',</span> <span class="pre">'chr2',</span> <span class="pre">'chr3',</span> <span class="pre">'chr4',</span> <span class="pre">'chr5',</span> <span class="pre">'chr6',</span> <span class="pre">'chr7',</span> <span class="pre">'chr8',</span> <span class="pre">'chr9',</span> <span class="pre">'chr10',</span> <span class="pre">'chr11',</span> <span class="pre">'chr12',</span> <span class="pre">'chr13',</span> <span class="pre">'chr14',</span> <span class="pre">'chr15',</span> <span class="pre">'chr16',</span> <span class="pre">'chr17',</span> <span class="pre">'chr18',</span> <span class="pre">'chr19',</span> <span class="pre">'chr20',</span> <span class="pre">'chr21',</span> <span class="pre">'chr22']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">blacklist_bed_filepath</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/utils/RegionFilteringAndBinning.html#RegionFilteringAndBinning"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.utils.RegionFilteringAndBinning.RegionFilteringAndBinning" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>A Binning object, which can be used to create a .bed file with the coordinates and bin numbers of every
bin. The method <code class="xref py py-func docutils literal notranslate"><span class="pre">write_bin_bedfile()</span></code> of this object can be used to create output.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>bed_filepath</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – path to a .bed file containing regions that should be extended and split into bins.
This could be e.g. a list of DNAse hypersensitivity sites or enhancer peaks.</p></li>
<li><p><strong>binsize</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Size of the bins. Use higher values to reduce noise, lower values to increase spatial
resolution.</p></li>
<li><p><strong>extend_to</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The regions will be extended by this value in both directions. Outmost bins will have their
center at <em>&lt;center of the region&gt;*+-*&lt;extend_to&gt;</em>.</p></li>
<li><p><strong>refgenome_chromsizes_filepath</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Path to a tab-delimited file containing the chromosome sizes for the
reference genome. The first column must be the chromosome name, the second column its size.
Can be the .fa.fai file associated to the reference genome. Must include all chromosomes given in
chromosomes_list.</p></li>
<li><p><strong>chromosomes_list</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]) – A list of chromosomes that should be analyzed. Regions on chromosomes that are not in
this list will be excluded from analysis. Default is [“chr1”, …, “chr22”].</p></li>
<li><p><strong>blacklist_bed_filepath</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]) – Optional: .bed file of a black list, such as the one from
<a href="#id5"><span class="problematic" id="id6">`</span></a>here &lt;<a class="reference external" href="https://github.com/Boyle-Lab/Blacklist/blob/master/lists/hg38-blacklist.v2.bed.gz">https://github.com/Boyle-Lab/Blacklist/blob/master/lists/hg38-blacklist.v2.bed.gz</a>&gt; <a href="#id7"><span class="problematic" id="id8">`</span></a>_ for hg38 
(unzip first). Regions that overlap any of the regions in this blacklist after extension by <strong>extend_to</strong> 
will be excluded from further analysis.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="liquorice.utils.RegionFilteringAndBinning.RegionFilteringAndBinning.filter_regions">
<span class="sig-name descname"><span class="pre">filter_regions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/utils/RegionFilteringAndBinning.html#RegionFilteringAndBinning.filter_regions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.utils.RegionFilteringAndBinning.RegionFilteringAndBinning.filter_regions" title="Permalink to this definition">¶</a></dt>
<dd><p>Remove regions that do not pass filters.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>df</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code>) – A <cite>pandas.DataFrame</cite> of a .bed file (with columns “chromosome”, “start”, “end”)</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code></p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>Filtered <cite>DataFrame</cite>, with regions excluded if (after extension by <code class="xref py py-attr docutils literal notranslate"><span class="pre">extend_by</span></code>) they :
i) are not on standard chromosomes (<code class="xref py py-attr docutils literal notranslate"><span class="pre">chromosomes_list</span></code>), ii) fall within
<code class="xref py py-attr docutils literal notranslate"><span class="pre">blacklist_bed_filepath</span></code>
(if defined), iii) extend beyond chromosome ends, or iv) if their start coordinate is larger than their end
coordinate.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="liquorice.utils.RegionFilteringAndBinning.RegionFilteringAndBinning.get_binstarts">
<span class="sig-name descname"><span class="pre">get_binstarts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">center</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/utils/RegionFilteringAndBinning.html#RegionFilteringAndBinning.get_binstarts"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.utils.RegionFilteringAndBinning.RegionFilteringAndBinning.get_binstarts" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets the bin start coordinates for a given center</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>center</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Coordinate of the center of the region of interest</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>]</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A list with start coordinates, one per bin. Length / number of created bins
depends on <code class="xref py py-attr docutils literal notranslate"><span class="pre">extend_to</span></code> and <code class="xref py py-attr docutils literal notranslate"><span class="pre">binsize</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="liquorice.utils.RegionFilteringAndBinning.RegionFilteringAndBinning.get_binstarts_percentile_split_central_roi">
<span class="sig-name descname"><span class="pre">get_binstarts_percentile_split_central_roi</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/utils/RegionFilteringAndBinning.html#RegionFilteringAndBinning.get_binstarts_percentile_split_central_roi"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.utils.RegionFilteringAndBinning.RegionFilteringAndBinning.get_binstarts_percentile_split_central_roi" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets the bin start coordinates for a given core region, splitting the core region into 5 bins with a length of
10,15,50,15,and 10 % of the core region length, respectively. The other bins, outside the core region,
have a length of <code class="xref py py-attr docutils literal notranslate"><span class="pre">binsize'.</span>
<span class="pre">The</span> <span class="pre">most</span> <span class="pre">upstream</span> <span class="pre">bin</span> <span class="pre">starts</span> <span class="pre">:attr:</span></code>.extend_to` bp upstream of the core region start, and the most downstream
bin starts (<code class="xref py py-attr docutils literal notranslate"><span class="pre">extend_to</span></code>-:attr:<a href="#id9"><span class="problematic" id="id10">`</span></a>.binsize`+&lt;length of the core region&gt;) bp downstream of the core region
end.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>start</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Coordinate of the start of the region of interest</p></li>
<li><p><strong>end</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Coordinate of the end of the region of interest</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>]</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A list with start coordinates, one per bin. Length / number of created bins
depends on <code class="xref py py-attr docutils literal notranslate"><span class="pre">extend_to</span></code> and <code class="xref py py-attr docutils literal notranslate"><span class="pre">binsize</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="liquorice.utils.RegionFilteringAndBinning.RegionFilteringAndBinning.is_within_chromsome_borders">
<span class="sig-name descname"><span class="pre">is_within_chromsome_borders</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">chrom</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chromsize_dict</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/utils/RegionFilteringAndBinning.html#RegionFilteringAndBinning.is_within_chromsome_borders"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.utils.RegionFilteringAndBinning.RegionFilteringAndBinning.is_within_chromsome_borders" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if a region is within its chromosome’s borders.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>chrom</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Chromosome name</p></li>
<li><p><strong>start</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Start coordinate</p></li>
<li><p><strong>end</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – End coordinate</p></li>
<li><p><strong>chromsize_dict</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>]) – Dictionary with chromosome names as keys and lengths as values. MUST contain <strong>chrom</strong> 
as a key, otherwise <cite>sys.exit()</cite> is called.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>True if the region is within the borders of the chromosome, False otherwise</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="liquorice.utils.RegionFilteringAndBinning.RegionFilteringAndBinning.write_bin_bedfile">
<span class="sig-name descname"><span class="pre">write_bin_bedfile</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">out_bedfile_path_bins</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">out_bedfile_path_regions_that_passed_filter</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/utils/RegionFilteringAndBinning.html#RegionFilteringAndBinning.write_bin_bedfile"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.utils.RegionFilteringAndBinning.RegionFilteringAndBinning.write_bin_bedfile" title="Permalink to this definition">¶</a></dt>
<dd><p>Splits every region in the <code class="xref py py-attr docutils literal notranslate"><span class="pre">bins_bed_filepath</span></code> file into bins, such that the central bin’s center is at
the center of the region, and such that the outermost bins each extend <code class="xref py py-attr docutils literal notranslate"><span class="pre">binsize</span></code>/2 over the edges of
(<em>&lt;the region’s center&gt;</em> - <code class="xref py py-attr docutils literal notranslate"><span class="pre">extend_to</span></code>) or (<em>&lt;the region’s center&gt;</em> + <code class="xref py py-attr docutils literal notranslate"><span class="pre">extend_to</span></code>),
respectively. Also assigns a bin nr. to every bin, starting with the most downstream bin.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>out_bedfile_path_bins</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Path to which the output .bed file with bin coordinates should be written to.</p></li>
<li><p><strong>out_bedfile_path_regions_that_passed_filter</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]) – If not <cite>None</cite>, write the regions that passed all filters and
that are the basis of the bins to this path (should end with .bed).</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="liquorice.utils.RegionFilteringAndBinning.RegionFilteringAndBinning.write_bin_bedfile_percentile_split_central_roi">
<span class="sig-name descname"><span class="pre">write_bin_bedfile_percentile_split_central_roi</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">out_bedfile_path_bins</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">out_bedfile_path_regions_that_passed_filter</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/utils/RegionFilteringAndBinning.html#RegionFilteringAndBinning.write_bin_bedfile_percentile_split_central_roi"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.utils.RegionFilteringAndBinning.RegionFilteringAndBinning.write_bin_bedfile_percentile_split_central_roi" title="Permalink to this definition">¶</a></dt>
<dd><p>Splits every region in the <code class="xref py py-attr docutils literal notranslate"><span class="pre">bins_bed_filepath</span></code> file into bins. The core region (given in
<code class="xref py py-attr docutils literal notranslate"><span class="pre">bed_filepath</span></code>) is splitted into 5 bins with a length of 10,15,50,15,and 10 % of the core region length,
respectively. The other bins, outside the core region, have a length of <code class="xref py py-attr docutils literal notranslate"><span class="pre">binsize'.</span>
<span class="pre">The</span> <span class="pre">most</span> <span class="pre">upstream</span> <span class="pre">bin</span> <span class="pre">starts</span> <span class="pre">:attr:</span></code>.extend_to` bp upstream of the core region start, and the most downstream
bin ends <code class="xref py py-attr docutils literal notranslate"><span class="pre">extend_to</span></code> bp downstream of the core region end. Also assigns a bin nr. to every bin,
starting with the most downstream bin.
:type out_bedfile_path_bins: <code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>
:param out_bedfile_path_bins: Path to which the output .bed file with bin coordinates should be written to.
:type out_bedfile_path_regions_that_passed_filter: <code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]
:param out_bedfile_path_regions_that_passed_filter: If not <cite>None</cite>, write the regions that passed all filters and</p>
<blockquote>
<div><p>that are the basis of the bins to this path (should end with .bed).</p>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-liquorice.utils.Workflows">
<span id="liquorice-utils-workflows-module"></span><span id="workflows-anchor"></span><h3>liquorice.utils.Workflows module<a class="headerlink" href="#module-liquorice.utils.Workflows" title="Permalink to this headline">¶</a></h3>
<dl class="py function">
<dt class="sig sig-object py" id="liquorice.utils.Workflows.add_biasfactors_percentile_split">
<span class="sig-prename descclassname"><span class="pre">liquorice.utils.Workflows.</span></span><span class="sig-name descname"><span class="pre">add_biasfactors_percentile_split</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">avg_readlength</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">liq_table</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_cores</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sampled_fragment_lengths</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">skip_these_biasfactors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/utils/Workflows.html#add_biasfactors_percentile_split"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.utils.Workflows.add_biasfactors_percentile_split" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a table with added bias-factors, taking into account different bin sizes. Also adds a “bin size” column.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>avg_readlength</strong> – Average length of reads in the .bam file.</p></li>
<li><p><strong>liq_table</strong> – pandas.DataFrame` with one row per bin, containing columns “chromosome”, “start”, “end”,
“bin nr.”, “coverage”, “sequence”, and “mappability”. Suitable input is the output of the
<code class="xref py py-func docutils literal notranslate"><span class="pre">get_complete_table()</span></code> method of the <a class="reference internal" href="#module-liquorice.utils.CoverageAndSequenceTablePreparation" title="liquorice.utils.CoverageAndSequenceTablePreparation"><code class="xref py py-attr docutils literal notranslate"><span class="pre">liquorice.utils.CoverageAndSequenceTablePreparation</span></code></a> class
object.</p></li>
<li><p><strong>n_cores</strong> – Max number of cores to use for calculations.</p></li>
<li><p><strong>sampled_fragment_lengths</strong> – A list containing fragment lengths that are representative of the sample’s global
fragment size distribution. Typically a few hundred fragments will suffice here.</p></li>
<li><p><strong>skip_these_biasfactors</strong> – Do not calculate these bias factors. Only these entries are allowed:
[“di and trinucleotides and GC content”,”mappability”, “di and trinucleotides”]</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A <cite>pandas.DataFrame</cite> with added bias-factors, taking into account different bin sizes, and a
“bin size” column.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="liquorice.utils.Workflows.run_liquorice_on_regionset_with_pretrained_biasmodel">
<span class="sig-prename descclassname"><span class="pre">liquorice.utils.Workflows.</span></span><span class="sig-name descname"><span class="pre">run_liquorice_on_regionset_with_pretrained_biasmodel</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">samplename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">regionset_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bam_filepath</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bed_filepath</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">biasmodel_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">refgenome_filepath</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">refgenome_chromsizes_filepath</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">refgenome_mappability_bigwig_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">blacklist_bed_filepath</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sampled_fragment_lengths</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">avg_readlength</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cna_seg_filepath</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mean_seq_depth</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_cores</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">binsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">500</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extend_to</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">20000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_default_fixed_sigma_values</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_biasfactor_table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">no_chr_prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">percentile_split_core_rois</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_this_roi_biasfactortable</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/utils/Workflows.html#run_liquorice_on_regionset_with_pretrained_biasmodel"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.utils.Workflows.run_liquorice_on_regionset_with_pretrained_biasmodel" title="Permalink to this definition">¶</a></dt>
<dd><p>Run the complete LIQUORICE workflow on a given region-set, using a pre-trained bias model. Main steps of this
workflow include: Filtering regions in the input .bed and splitting remaining regions into bins; calculating
sequence, coverage, and mappability for every bin; calculating bias factors for every bin, using the pre-trained
model and the inferred bias-factors to correct the coverage; aggregating information across regions and fitting
gaussian functions and an intercept to the corrected, aggregated coverage data. Also creates plots and result
tables.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>samplename</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Name of the sample (to be used in plots and output tables).</p></li>
<li><p><strong>regionset_name</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Name of the region-set (to be used in plots and output tables).</p></li>
<li><p><strong>bam_filepath</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Path to the .bam file containing the mapped reads for the sample. Does not need to be
duplicate-filtered, this is done by the function that calculates the coverage.</p></li>
<li><p><strong>bed_filepath</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – path to a .bed file containing regions-of-interest that should be extended and split into bins.
This could be e.g. a list of DNAse hypersensitivity sites or enhancer peaks.</p></li>
<li><p><strong>biasmodel_path</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Path to a trained biasmodel.</p></li>
<li><p><strong>refgenome_filepath</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Path to the reference genome .fa(.gz) file. Must have a .fai index in the same dirname.</p></li>
<li><p><strong>refgenome_chromsizes_filepath</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Path to a tab-delimited file containing the chromosome sizes for the
reference genome. The first column must be the chromosome name, the second column its size.
Can be the .fa.fai file associated to the reference genome.</p></li>
<li><p><strong>refgenome_mappability_bigwig_path</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Path to a .bigWig file containing (forward) mappability values scaled
between 0 and 1.</p></li>
<li><p><strong>blacklist_bed_filepath</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]) – .bed file of a black list, such as the one from
<a href="#id11"><span class="problematic" id="id12">`</span></a>here &lt;<a class="reference external" href="https://github.com/Boyle-Lab/Blacklist/blob/master/lists/hg38-blacklist.v2.bed.gz">https://github.com/Boyle-Lab/Blacklist/blob/master/lists/hg38-blacklist.v2.bed.gz</a>&gt; <a href="#id13"><span class="problematic" id="id14">`</span></a>_ for hg38
(unzip first). Regions that overlap any of the regions in this blacklist after extension by <strong>extend_to</strong>
will be excluded from further analysis. Set to None to skip this step.</p></li>
<li><p><strong>sampled_fragment_lengths</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>]) – A list containing fragment lengths that are representative of the sample’s global
fragment size distribution. Typically a few hundred fragments will suffice here.</p></li>
<li><p><strong>avg_readlength</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – (Average) length of the reads in the .bam file.</p></li>
<li><p><strong>cna_seg_filepath</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]) – If specified, use this .seg file to correct the coverage by the values specified in
this file prior to correcting coverage with the bias model. Use this if you want to normalize out the effects of
copy number aberrations (CNAs) on the coverage. File must be tab-separated, with column names as first line.
The second, third and fourth column must be chromosome, start, and end of the segment, and the last column must
be the log2-ratio of observed/expected read depth.</p></li>
<li><p><strong>mean_seq_depth</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – A float that quantifies the global coverage/sequencing depth. Coverages per bin will be
normalized (i.e. divided) by this value.</p></li>
<li><p><strong>binsize</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Size of the bins. Use higher values to reduce noise, lower values to increase spatial
resolution. Using the same binsize for the biasmodel generation as for the region-set-of-interest is probably
preferable.</p></li>
<li><p><strong>extend_to</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The regions will be extended by this value in both directions. Outmost bins will have their
center at * &lt;center of the region&gt; * +- * &lt;extend_to&gt; <a href="#id15"><span class="problematic" id="id16">*</span></a>.</p></li>
<li><p><strong>use_default_fixed_sigma_values</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – If <cite>True</cite>, use the following contraints for the sigma values of the small,
medium, and large Gaussian, repectively: 149-150 bp, 757-758 bp, and 6078-6079 bp.</p></li>
<li><p><strong>n_cores</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Maximum number of cores to use during steps that allow multiprocessing/multithreading.</p></li>
<li><p><strong>save_biasfactor_table</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – If <cite>True</cite>, save a table of coverage and biasfactor values per bin as
“coverage_and_biasfactors_per_bin.csv”.</p></li>
<li><p><strong>no_chr_prefix</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – If True, set the list of allowed chromosomes to [str(i) for i in range(1,23)] instead of
[“chr”+str(i) for i in range(1,23)]</p></li>
<li><p><strong>percentile_split_core_rois</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – If set, split the central region into 5 bins of variable size instead of always
using a fixed binsize. <cite>extend_to</cite> should not be set to 0 if this is used.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="liquorice.utils.Workflows.run_liquorice_train_biasmodel_on_same_regions">
<span class="sig-prename descclassname"><span class="pre">liquorice.utils.Workflows.</span></span><span class="sig-name descname"><span class="pre">run_liquorice_train_biasmodel_on_same_regions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">samplename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">regionset_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bam_filepath</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bed_filepath</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">refgenome_filepath</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">refgenome_chromsizes_filepath</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">refgenome_mappability_bigwig_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">blacklist_bed_filepath</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sampled_fragment_lengths</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">avg_readlength</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cna_seg_filepath</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mean_seq_depth</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_cores</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">binsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">500</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extend_to</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">20000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">biasmodel_output_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'trained_biasmodel.joblib'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nr_of_bins_for_training_and_testing</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">skip_central_n_bins_for_training</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_training_table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_default_fixed_sigma_values</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_biasfactor_table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">no_chr_prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">percentile_split_core_rois</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_cross_validated_predictions</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_this_roi_biasfactortable</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">speed_mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/utils/Workflows.html#run_liquorice_train_biasmodel_on_same_regions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.utils.Workflows.run_liquorice_train_biasmodel_on_same_regions" title="Permalink to this definition">¶</a></dt>
<dd><p>Run the complete LIQUORICE workflow on a given region-set, and train the bias-model on data from the same region-
set. Main steps of this workflow include: Filtering regions in the input .bed and splitting remaining regions into
bins; calculating
sequence, coverage, and mappability for every bin; calculating bias factors for every bin, training a bias model,
using the trained
model and the inferred bias-factors to correct the coverage; aggregating information across regions and fitting
gaussian functions and an intercept to the corrected, aggregated coverage data. Also creates plots and result
tables.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>samplename</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Name of the sample (to be used in plots and output tables).</p></li>
<li><p><strong>regionset_name</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Name of the region-set (to be used in plots and output tables).</p></li>
<li><p><strong>bam_filepath</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Path to the .bam file containing the mapped reads for the sample. Does not need to be
duplicate-filtered, this is done by the function that calculates the coverage.</p></li>
<li><p><strong>bed_filepath</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – path to a .bed file containing regions-of-interest that should be extended and split into bins.
This could be e.g. a list of DNAse hypersensitivity sites or enhancer peaks.</p></li>
<li><p><strong>refgenome_filepath</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Path to the reference genome .fa(.gz) file. Must have a .fai index in the same dirname.</p></li>
<li><p><strong>refgenome_chromsizes_filepath</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Path to a tab-delimited file containing the chromosome sizes for the
reference genome. The first column must be the chromosome name, the second column its size.
Can be the .fa.fai file associated to the reference genome.</p></li>
<li><p><strong>refgenome_mappability_bigwig_path</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Path to a .bigWig file containing (forward) mappability values scaled
between 0 and 1.</p></li>
<li><p><strong>blacklist_bed_filepath</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]) – .bed file of a black list, such as the one from
<a href="#id17"><span class="problematic" id="id18">`</span></a>here &lt;<a class="reference external" href="https://github.com/Boyle-Lab/Blacklist/blob/master/lists/hg38-blacklist.v2.bed.gz">https://github.com/Boyle-Lab/Blacklist/blob/master/lists/hg38-blacklist.v2.bed.gz</a>&gt; <a href="#id19"><span class="problematic" id="id20">`</span></a>_ for hg38
(unzip first). Regions that overlap any of the regions in this blacklist after extension by <strong>extend_to</strong>
will be excluded from further analysis. Set to None to skip this step.</p></li>
<li><p><strong>sampled_fragment_lengths</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>]) – A list containing fragment lengths that are representative of the sample’s global
fragment size distribution. Typically a few hundred fragments will suffice here.</p></li>
<li><p><strong>avg_readlength</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – (Average) length of the reads in the .bam file.</p></li>
<li><p><strong>cna_seg_filepath</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]) – If specified, use this .seg file to correct the coverage by the values specified in
this file prior to correcting coverage with the bias model. Use this if you want to normalize out the effects of
copy number aberrations (CNAs) on the coverage. File must be tab-separated, with column names as first line.
The second, third and fourth column must be chromosome, start, and end of the segment, and the last column must
be the log2-ratio of observed/expected read depth.</p></li>
<li><p><strong>mean_seq_depth</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – A float that quantifies the global coverage/sequencing depth. Coverages per bin will be
normalized (i.e. divided) by this value.</p></li>
<li><p><strong>binsize</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Size of the bins. Use higher values to reduce noise, lower values to increase spatial
resolution. Using the same binsize for the biasmodel generation as for the region-set-of-interest is probably
preferable.</p></li>
<li><p><strong>extend_to</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The regions will be extended by this value in both directions. Outmost bins will have their
center at * &lt;center of the region&gt; * +- * &lt;extend_to&gt; <a href="#id21"><span class="problematic" id="id22">*</span></a>.</p></li>
<li><p><strong>use_default_fixed_sigma_values</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – If <cite>True</cite>, use the following contraints for the sigma values of the small,
medium, and large Gaussian, repectively: 149-150 bp, 757-758 bp, and 6078-6079 bp.</p></li>
<li><p><strong>n_cores</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Maximum number of cores to use during steps that allow multiprocessing/multithreading.</p></li>
<li><p><strong>save_biasfactor_table</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – If <cite>True</cite>, save a table of coverage and biasfactor values per bin as
“coverage_and_biasfactors_per_bin.csv”.</p></li>
<li><p><strong>no_chr_prefix</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – If True, set the list of allowed chromosomes to [str(i) for i in range(1,23)] instead of
[“chr”+str(i) for i in range(1,23)]</p></li>
<li><p><strong>biasmodel_output_path</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Path to which the trained biasmodel should be saved to. Must have a
.joblib extension.</p></li>
<li><p><strong>nr_of_bins_for_training_and_testing</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>]) – Subset the training_df to this many bins. Can speed up the
computation time of the model training, but using too few bins will make the model less precise. To speed up
computations, we would recommend decreasing the number of regions in the .bed file rather than altering this
parameter, as this is more efficient.</p></li>
<li><p><strong>save_training_table</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – If <cite>True</cite>, save the table that was used to train the biasmodel (coverage and biasfactors
per bin) as “training_table.csv”.</p></li>
<li><p><strong>skip_central_n_bins_for_training</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The n most central bins will not be used for training the bias model.</p></li>
<li><p><strong>no_chr_prefix</strong> – If True, set the list of allowed chromosomes to [str(i) for i in range(1,23)] instead of
[“chr”+str(i) for i in range(1,23)]</p></li>
<li><p><strong>percentile_split_core_rois</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – If set, split the central region into 5 bins of variable size instead of always
using a fixed binsize. <cite>extend_to</cite> should not be set to 0 if this is used.</p></li>
<li><p><strong>use_cross_validated_predictions</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – Instead of training on the full dataset, train twice on half of the dataset
and predict the other half. Ignores nr_of_bins_for_training_and_testing</p></li>
<li><p><strong>use_this_roi_biasfactortable</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]) – If set use the specified biasfactor table and only train/apply the biasmodel.</p></li>
<li><p><strong>speed_mode</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – Only perform GC correction, don’t correct using mappability or di/trinucleotides. Setting this
flag makes LIQUORICE considerably faster, but may lead to less accurate results.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="liquorice.utils.Workflows.train_biasmodel_for_sample">
<span class="sig-prename descclassname"><span class="pre">liquorice.utils.Workflows.</span></span><span class="sig-name descname"><span class="pre">train_biasmodel_for_sample</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">samplename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bam_filepath</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bed_filepath</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">refgenome_filepath</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">refgenome_chromsizes_filepath</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">refgenome_mappability_bigwig_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">blacklist_bed_filepath</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sampled_fragment_lengths</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">avg_readlength</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mean_seq_depth</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cna_seg_filepath</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_cores</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extend_to</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">binsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">500</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">biasmodel_output_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'trained_biasmodel.joblib'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nr_of_bins_for_training_and_testing</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_training_table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">no_chr_prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">percentile_split_core_rois</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/liquorice/utils/Workflows.html#train_biasmodel_for_sample"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#liquorice.utils.Workflows.train_biasmodel_for_sample" title="Permalink to this definition">¶</a></dt>
<dd><p>Go through all steps of LIQUORICE up to the biasmodel training. The resulting biasmodel under
<strong>biasmodel_output_path</strong> can then be used when LIQUORICE is run for region-sets of interest.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>samplename</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Name of the sample (to be used in plots and output tables).</p></li>
<li><p><strong>bam_filepath</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Path to the .bam file containing the mapped reads for the sample. Does not need to be
duplicate-filtered, this is done by the function that calculates the coverage.</p></li>
<li><p><strong>bed_filepath</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – path to a .bed file containing regions that should be used to build the biasmodel.
A .bed file that contains many random regions across the genome is recommended.</p></li>
<li><p><strong>refgenome_filepath</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Path to the reference genome .fa(.gz) file. Must have a .fai index in the same dirname.</p></li>
<li><p><strong>refgenome_chromsizes_filepath</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Path to a tab-delimited file containing the chromosome sizes for the
reference genome. The first column must be the chromosome name, the second column its size.
Can be the .fa.fai file associated to the reference genome.</p></li>
<li><p><strong>refgenome_mappability_bigwig_path</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Path to a .bigWig file containing (forward) mappability values scaled
between 0 and 1.</p></li>
<li><p><strong>blacklist_bed_filepath</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]) – .bed file of a black list, such as the one from
<a href="#id23"><span class="problematic" id="id24">`</span></a>here &lt;<a class="reference external" href="https://github.com/Boyle-Lab/Blacklist/blob/master/lists/hg38-blacklist.v2.bed.gz">https://github.com/Boyle-Lab/Blacklist/blob/master/lists/hg38-blacklist.v2.bed.gz</a>&gt; <a href="#id25"><span class="problematic" id="id26">`</span></a>_ for hg38
(unzip first). Regions that overlap any of the regions in this blacklist after extension by <strong>extend_to</strong>
will be excluded from further analysis. Set to None to skip this step.</p></li>
<li><p><strong>sampled_fragment_lengths</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>]) – A list containing fragment lengths that are representative of the sample’s global
fragment size distribution. Typically a few hundred fragments will suffice here.</p></li>
<li><p><strong>avg_readlength</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – (Average) length of the reads in the .bam file.</p></li>
<li><p><strong>mean_seq_depth</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – A float that quantifies the global coverage/sequencing depth. Coverages per bin will be
normalized (i.e. divided) by this value.</p></li>
<li><p><strong>cna_seg_filepath</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]) – If specified, use this .seg file to correct the coverage by the values specified in
this file prior to model training. Use this if you want to normalize out the effects of copy number aberrations
(CNAs) on the coverage. File must be tab-separated, with column names as first line. The second, third and fourth
column must be chromosome, start, and end of the segment, and the last column must be the log2-ratio of
observed/expected read depth.</p></li>
<li><p><strong>binsize</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Size of the bins. Use higher values to reduce noise, lower values to increase spatial
resolution. Using the same binsize for the biasmodel generation as for the region-set-of-interest is probably
preferable.</p></li>
<li><p><strong>extend_to</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The regions will be extended by this value in both directions. Outmost bins will have their
center at <em>&lt;center of the region&gt;*+-*&lt;extend_to&gt;</em>. Here, the default is 0, meaning only a single bin will be
generated per region in the .bed file. This speeds up the computation - for more precise biasmodels, we
recommend running for a larger set of regions rather than increasing the <strong>extend_to</strong> parameter.</p></li>
<li><p><strong>n_cores</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Maximum number of cores to use during steps that allow multiprocessing/multithreading.</p></li>
<li><p><strong>biasmodel_output_path</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Path to which the trained biasmodel should be saved to. Must have a
.joblib extension.</p></li>
<li><p><strong>nr_of_bins_for_training_and_testing</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>]) – Subset the training_df to this many bins. Can speed up the
computation time of the model training, but using too few bins will make the model less precise. To speed up
computations, we would recommend decreasing the number of regions in the .bed file rather than altering this
parameter, as this is more efficient.</p></li>
<li><p><strong>save_training_table</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – If <cite>True</cite>, save the table that was used to train the biasmodel (coverage and biasfactors
per bin) as “training_table.csv”.</p></li>
<li><p><strong>no_chr_prefix</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – If True, set the list of allowed chromosomes to [str(i) for i in range(1,23)] instead of
[“chr”+str(i) for i in range(1,23)]</p></li>
<li><p><strong>percentile_split_core_rois</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – If set, split the central region into 5 bins of variable size instead of always
using a fixed binsize. <cite>extend_to</cite> should not be set to 0 if this is used.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></p>
</dd>
</dl>
</dd></dl>

</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="liquorice_commandline_tool.html" class="btn btn-neutral float-right" title="The LIQUORICE command-line-tool" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="intro.html" class="btn btn-neutral float-left" title="Introduction" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2021, Peter Peneder.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>