

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>liquorice.utils.BinTableBiasFactors &mdash; LIQUORICE 0.5.0 documentation</title>
  

  
  <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../../" src="../../../_static/documentation_options.js"></script>
        <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
        <script src="../../../_static/jquery.js"></script>
        <script src="../../../_static/underscore.js"></script>
        <script src="../../../_static/doctools.js"></script>
    
    <script type="text/javascript" src="../../../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../../index.html">
          

          
            
            <img src="../../../_static/liquorice_logo_fitted_transparent.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../intro.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../intro.html#installation">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../intro.html#citation">Citation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../liquorice.html">The <cite>liquorice</cite> python package</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../liquorice_commandline_tool.html">The LIQUORICE command-line-tool</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../liquorice_commandline_tool.html#liquorice-s-summary-tool">LIQUORICEâ€™s summary tool</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../liquorice_commandline_tool.html#usage-parameters-and-examples">Usage, Parameters, and Examples</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">LIQUORICE</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="../../index.html">Module code</a> &raquo;</li>
        
      <li>liquorice.utils.BinTableBiasFactors</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for liquorice.utils.BinTableBiasFactors</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="n">sys</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="vm">__file__</span><span class="p">),</span> <span class="s1">&#39;../..&#39;</span><span class="p">)))</span>
<span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="p">[</span><span class="s2">&quot;PYTHONPATH&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="vm">__file__</span><span class="p">),</span> <span class="s1">&#39;../..&#39;</span><span class="p">))</span> <span class="c1"># required by Ray, which is</span>
<span class="c1"># used by modin</span>
<span class="kn">from</span> <span class="nn">liquorice.utils.BiasFactorWeights</span> <span class="kn">import</span> <span class="o">*</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">modin.pandas</span> <span class="k">as</span> <span class="nn">modinpd</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">import</span> <span class="nn">swifter</span>
<span class="kn">from</span> <span class="nn">datetime</span> <span class="kn">import</span> <span class="n">datetime</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">List</span><span class="p">,</span><span class="n">Dict</span>

<div class="viewcode-block" id="BiasFactorHandler"><a class="viewcode-back" href="../../../liquorice.html#liquorice.utils.BinTableBiasFactors.BiasFactorHandler">[docs]</a><span class="k">class</span> <span class="nc">BiasFactorHandler</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Object used for calculation of per-bin bias factors. Typically, after creation of the object a user would call</span>
<span class="sd">    its method :func:`.get_table_with_bias_factors` on it, and save the returned `DataFrame` for subsequent analysis</span>
<span class="sd">    and correction of associations of bias factors with coverage.</span>

<span class="sd">    :param binsize: Size of the bins. Higher values to reduce noise, lower values increase spatial</span>
<span class="sd">        resolution.</span>

<span class="sd">    :param fragments: A list containing fragment lengths that are representative of the sample&#39;s global fragment</span>
<span class="sd">        size distribution. Typically a few hundred fragments will suffice here.</span>
<span class="sd">    :param readlength: Average length of reads in the .bam file.</span>
<span class="sd">    :param df: `pandas.DataFrame` with one row per bin, containing columns &quot;chromosome&quot;, &quot;start&quot;, &quot;end&quot;,</span>
<span class="sd">        &quot;bin nr.&quot;, &quot;coverage&quot;, &quot;sequence&quot;, and &quot;mappability&quot;. Suitable input is the output of the</span>
<span class="sd">        :func:`get_complete_table` method of the :attr:`liquorice.utils.CoverageAndSequenceTablePreparation` class</span>
<span class="sd">        object.</span>
<span class="sd">    :param n_cores: Max number of cores to use for calculations.</span>
<span class="sd">    :param skip_these_biasfactors: Do not calculate these bias factors. Only these entries are allowed:</span>
<span class="sd">        [&quot;di and trinucleotides and GC content&quot;,&quot;mappability&quot;, &quot;di and trinucleotides&quot;]</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">binsize</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">fragments</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">int</span><span class="p">],</span> <span class="n">readlength</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span>
                 <span class="n">n_cores</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">skip_these_biasfactors</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[])</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">binsize</span><span class="o">=</span><span class="n">binsize</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">readlength</span><span class="o">=</span><span class="n">readlength</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">fragments</span><span class="o">=</span><span class="n">fragments</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">longest_frag</span><span class="o">=</span><span class="nb">max</span><span class="p">(</span><span class="n">fragments</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">=</span><span class="n">df</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">GC_weights</span><span class="o">=</span><span class="n">get_GC_weights_binwide</span><span class="p">(</span><span class="n">binsize</span><span class="o">=</span><span class="n">binsize</span><span class="p">,</span><span class="n">fragments</span><span class="o">=</span><span class="n">fragments</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">dinuc_weights</span><span class="o">=</span><span class="n">get_dinuc_weights_binwide</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">GC_weights</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">trinuc_weights</span><span class="o">=</span><span class="n">get_trinuc_weights_binwide</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">GC_weights</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">total_GC_weight</span><span class="o">=</span><span class="nb">sum</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">GC_weights</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">total_dinuc_weight</span><span class="o">=</span><span class="nb">sum</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">dinuc_weights</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">total_trinuc_weight</span><span class="o">=</span><span class="nb">sum</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">trinuc_weights</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">fwd_mapp_weights</span><span class="p">,</span><span class="bp">self</span><span class="o">.</span><span class="n">rev_mapp_weights</span><span class="o">=</span><span class="n">get_mapp_weights_binwide</span><span class="p">(</span><span class="n">binsize</span><span class="o">=</span><span class="n">binsize</span><span class="p">,</span><span class="n">fragments</span><span class="o">=</span><span class="n">fragments</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">total_fwd_mapp_weight</span><span class="p">,</span><span class="bp">self</span><span class="o">.</span><span class="n">total_rev_mapp_weight</span><span class="o">=</span><span class="nb">sum</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">fwd_mapp_weights</span><span class="p">),</span><span class="nb">sum</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">rev_mapp_weights</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">dinucdict</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">trinucdict</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">revcompdict</span> <span class="o">=</span> <span class="n">get_nucleotide_dicts</span><span class="p">()</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">n_cores</span><span class="o">=</span><span class="n">n_cores</span>

        <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">skip_these_biasfactors</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">item</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;di and trinucleotides and GC content&quot;</span><span class="p">,</span><span class="s2">&quot;mappability&quot;</span><span class="p">,</span> <span class="s2">&quot;di and trinucleotides&quot;</span><span class="p">]:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;skip_these_biasfactors may only contain the following items: &#39;</span>
                                 <span class="s1">&#39;&quot;di and trinucleotides and GC content&quot;,&#39;</span>
                                 <span class="s1">&#39;&quot;mappability&quot;, &quot;di and trinucleotides&quot;&#39;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">skip_these_biasfactors</span><span class="o">=</span><span class="n">skip_these_biasfactors</span>

<div class="viewcode-block" id="BiasFactorHandler.get_GC_and_di_and_trinuc_weights"><a class="viewcode-back" href="../../../liquorice.html#liquorice.utils.BinTableBiasFactors.BiasFactorHandler.get_GC_and_di_and_trinuc_weights">[docs]</a>    <span class="k">def</span> <span class="nf">get_GC_and_di_and_trinuc_weights</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">sequence</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span><span class="nb">float</span><span class="p">]:</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculate bias factors for GC-content as well as bias factors for all di- and trinucleotides</span>
<span class="sd">        (reverse and forward complements merged into single factors) for a given sequence. Factors are scaled between</span>
<span class="sd">        0 and 1, 1 is highest (e.g. GC content: 0.. no G or C, 0.461... average GC content, 1... only G and C).</span>

<span class="sd">        :param sequence: Genomic sequence of the bin extended by *max(* :attr:`.fragments` *)* in both directions,</span>
<span class="sd">            such that its length matches :attr:`GC_weights`.</span>
<span class="sd">        :return: A dictionary with entries corresponding to the bias factors for GC-content as well as bias factors</span>
<span class="sd">            for all di- and trinucleotides (reverse and forward complements merged into single factors).</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span><span class="o">!=</span><span class="nb">str</span><span class="p">:</span>  <span class="c1"># Required because swifter calls this function with a series as sequence once, and</span>
            <span class="c1"># and does not handle the RunTimeError below properly.</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;The input to this function must be a str, but was </span><span class="si">{</span><span class="nb">type</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span>
        <span class="n">GC_list</span><span class="o">=</span><span class="p">[]</span>
        <span class="n">dinuc_res_dict</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">dinucdict</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">trinuc_res_dict</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">trinucdict</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

        <span class="c1"># start_time=datetime.now()</span>
        <span class="k">for</span> <span class="n">pos</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">)):</span>
            <span class="n">subseq</span><span class="o">=</span><span class="n">sequence</span><span class="p">[</span><span class="n">pos</span><span class="p">:</span><span class="n">pos</span><span class="o">+</span><span class="mi">3</span><span class="p">]</span>
            <span class="n">dinuc_weight</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">dinuc_weights</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span>
            <span class="n">trinuc_weight</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">trinuc_weights</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span>

            <span class="c1">### GC part ###</span>
            <span class="n">nuc</span><span class="o">=</span><span class="n">subseq</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">nuc</span><span class="o">==</span><span class="s2">&quot;G&quot;</span> <span class="ow">or</span> <span class="n">nuc</span><span class="o">==</span><span class="s2">&quot;g&quot;</span> <span class="ow">or</span> <span class="n">nuc</span><span class="o">==</span><span class="s2">&quot;c&quot;</span> <span class="ow">or</span> <span class="n">nuc</span><span class="o">==</span><span class="s2">&quot;C&quot;</span><span class="p">:</span>
                <span class="n">GC_at_pos</span><span class="o">=</span><span class="mi">1</span>
            <span class="k">elif</span> <span class="n">nuc</span><span class="o">==</span><span class="s2">&quot;A&quot;</span> <span class="ow">or</span> <span class="n">nuc</span><span class="o">==</span><span class="s2">&quot;a&quot;</span> <span class="ow">or</span> <span class="n">nuc</span><span class="o">==</span><span class="s2">&quot;t&quot;</span> <span class="ow">or</span> <span class="n">nuc</span><span class="o">==</span><span class="s2">&quot;T&quot;</span><span class="p">:</span>
                <span class="n">GC_at_pos</span><span class="o">=</span><span class="mi">0</span>
            <span class="k">elif</span> <span class="n">nuc</span><span class="o">==</span><span class="s2">&quot;N&quot;</span> <span class="ow">or</span> <span class="n">nuc</span><span class="o">==</span><span class="s2">&quot;n&quot;</span><span class="p">:</span>
                <span class="n">GC_at_pos</span><span class="o">=</span><span class="mf">0.461</span> <span class="c1">#0.461 is the genome-wide GC content mean</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unexpected letter &#39;</span><span class="si">{</span><span class="n">nuc</span><span class="si">}</span><span class="s2">&#39; encountered in genomic sequence.&quot;</span><span class="p">)</span>
                <span class="k">raise</span> <span class="ne">RuntimeError</span>
            <span class="n">GC_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">GC_at_pos</span><span class="p">)</span>

            <span class="c1">### Dinucleotide part ###</span>
            <span class="n">dinuc</span><span class="o">=</span><span class="n">subseq</span><span class="p">[:</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
            <span class="c1"># try:</span>
            <span class="c1">#     weight=(GC_weights_subsetted[0]+GC_weights_subsetted[1])/2</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">dinuc_res_dict</span><span class="p">[</span><span class="n">dinuc</span><span class="p">]</span><span class="o">+=</span><span class="n">dinuc_weight</span>
            <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>  <span class="c1"># if the dinuc. is not a key in the dict, usually its reverse complement should be</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">dinuc_revcomp</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">revcompdict</span><span class="p">[</span><span class="n">dinuc</span><span class="p">]</span>
                    <span class="n">dinuc_res_dict</span><span class="p">[</span><span class="n">dinuc_revcomp</span><span class="p">]</span><span class="o">+=</span><span class="n">dinuc_weight</span>
                <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                    <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">dinuc</span><span class="p">)</span><span class="o">&lt;</span><span class="mi">2</span> <span class="ow">or</span> <span class="s2">&quot;N&quot;</span> <span class="ow">in</span> <span class="n">dinuc</span> <span class="c1"># if at last position or an &quot;N&quot; is contained</span>
                    <span class="k">pass</span>
            <span class="c1"># except IndexError:  # this happens if we are looking at the last position</span>
            <span class="c1">#     pass</span>

            <span class="c1">### Trinucleotide part ###</span>
            <span class="n">trinuc</span><span class="o">=</span><span class="n">subseq</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
            <span class="c1"># try:</span>
            <span class="c1">#     weight=(GC_weights_subsetted[0]+GC_weights_subsetted[1]+GC_weights_subsetted[2])/3</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">trinuc_res_dict</span><span class="p">[</span><span class="n">trinuc</span><span class="p">]</span><span class="o">+=</span><span class="n">trinuc_weight</span>
            <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>  <span class="c1"># if the trinuc. is not a key in the dict, usually its reverse complement should be</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">trinuc_revcomp</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">revcompdict</span><span class="p">[</span><span class="n">trinuc</span><span class="p">]</span>
                    <span class="n">trinuc_res_dict</span><span class="p">[</span><span class="n">trinuc_revcomp</span><span class="p">]</span><span class="o">+=</span><span class="n">trinuc_weight</span>
                <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                    <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">trinuc</span><span class="p">)</span><span class="o">&lt;</span><span class="mi">3</span> <span class="ow">or</span> <span class="s2">&quot;N&quot;</span> <span class="ow">in</span> <span class="n">trinuc</span>  <span class="c1"># if at (second-to-) last position or an &quot;N&quot; is contained</span>
                    <span class="k">pass</span>
            <span class="c1"># except IndexError:</span>
            <span class="c1">#     pass</span>

        <span class="c1">## Normalize, combine and return result</span>
        <span class="n">GC_bias_factor</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;GC content&quot;</span><span class="p">:</span><span class="nb">sum</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">GC_list</span><span class="p">)</span><span class="o">*</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">GC_weights</span><span class="p">))</span> <span class="o">/</span> <span class="bp">self</span><span class="o">.</span><span class="n">total_GC_weight</span><span class="p">}</span>
        <span class="n">dinuc_res_dict</span><span class="o">=</span><span class="p">{</span><span class="n">key</span><span class="p">:</span><span class="n">value</span><span class="o">/</span><span class="bp">self</span><span class="o">.</span><span class="n">total_GC_weight</span> <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">dinuc_res_dict</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span>
        <span class="n">trinuc_res_dict</span><span class="o">=</span><span class="p">{</span><span class="n">key</span><span class="p">:</span><span class="n">value</span><span class="o">/</span><span class="bp">self</span><span class="o">.</span><span class="n">total_GC_weight</span> <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">trinuc_res_dict</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span>

        <span class="n">res</span><span class="o">=</span><span class="n">trinuc_res_dict</span>
        <span class="n">res</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">dinuc_res_dict</span><span class="p">)</span>
        <span class="n">res</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">GC_bias_factor</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">res</span></div>



<div class="viewcode-block" id="BiasFactorHandler.get_GC_bias_factor"><a class="viewcode-back" href="../../../liquorice.html#liquorice.utils.BinTableBiasFactors.BiasFactorHandler.get_GC_bias_factor">[docs]</a>    <span class="k">def</span> <span class="nf">get_GC_bias_factor</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">sequence</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns a number in the range 0-1 that represent&#39;s the bin&#39;s overall GC bias. Factors are scaled between</span>
<span class="sd">        0 and 1, 1 is highest (e.g. GC content: 0.. no G or C, 0.461... average GC content, 1... only G and C).</span>

<span class="sd">        :param sequence: Genomic sequence of the bin extended by *max(* :attr:`.fragments` *)* in both directions,</span>
<span class="sd">            such that its length matches :attr:`GC_weights`.</span>
<span class="sd">        :return: A</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span><span class="o">!=</span><span class="nb">str</span><span class="p">:</span>  <span class="c1"># Required because swifter calls this function with a series as sequence once, and</span>
        <span class="c1"># and does not handle the RunTimeError below properly.</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;The input to this function must be a str, but was </span><span class="si">{</span><span class="nb">type</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span>
        <span class="n">GC_list</span><span class="o">=</span><span class="p">[]</span>

        <span class="c1"># start_time=datetime.now()</span>
        <span class="k">for</span> <span class="n">pos</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">)):</span>
            <span class="n">subseq</span><span class="o">=</span><span class="n">sequence</span><span class="p">[</span><span class="n">pos</span><span class="p">:</span><span class="n">pos</span><span class="o">+</span><span class="mi">3</span><span class="p">]</span>

            <span class="c1">### GC part ###</span>
            <span class="n">nuc</span><span class="o">=</span><span class="n">subseq</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">nuc</span><span class="o">==</span><span class="s2">&quot;G&quot;</span> <span class="ow">or</span> <span class="n">nuc</span><span class="o">==</span><span class="s2">&quot;g&quot;</span> <span class="ow">or</span> <span class="n">nuc</span><span class="o">==</span><span class="s2">&quot;c&quot;</span> <span class="ow">or</span> <span class="n">nuc</span><span class="o">==</span><span class="s2">&quot;C&quot;</span><span class="p">:</span>
                <span class="n">GC_at_pos</span><span class="o">=</span><span class="mi">1</span>
            <span class="k">elif</span> <span class="n">nuc</span><span class="o">==</span><span class="s2">&quot;A&quot;</span> <span class="ow">or</span> <span class="n">nuc</span><span class="o">==</span><span class="s2">&quot;a&quot;</span> <span class="ow">or</span> <span class="n">nuc</span><span class="o">==</span><span class="s2">&quot;t&quot;</span> <span class="ow">or</span> <span class="n">nuc</span><span class="o">==</span><span class="s2">&quot;T&quot;</span><span class="p">:</span>
                <span class="n">GC_at_pos</span><span class="o">=</span><span class="mi">0</span>
            <span class="k">elif</span> <span class="n">nuc</span><span class="o">==</span><span class="s2">&quot;N&quot;</span> <span class="ow">or</span> <span class="n">nuc</span><span class="o">==</span><span class="s2">&quot;n&quot;</span><span class="p">:</span>
                <span class="n">GC_at_pos</span><span class="o">=</span><span class="mf">0.461</span> <span class="c1">#0.461 is the genome-wide GC content mean</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unexpected letter &#39;</span><span class="si">{</span><span class="n">nuc</span><span class="si">}</span><span class="s2">&#39; encountered in genomic sequence.&quot;</span><span class="p">)</span>
                <span class="k">raise</span> <span class="ne">RuntimeError</span>
            <span class="n">GC_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">GC_at_pos</span><span class="p">)</span>

        <span class="n">GC_bias_factor</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;GC content&quot;</span><span class="p">:</span><span class="nb">sum</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">GC_list</span><span class="p">)</span><span class="o">*</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">GC_weights</span><span class="p">))</span> <span class="o">/</span> <span class="bp">self</span><span class="o">.</span><span class="n">total_GC_weight</span><span class="p">}</span>


        <span class="n">res</span><span class="o">=</span><span class="n">GC_bias_factor</span>
        <span class="k">return</span> <span class="n">res</span></div>


    <span class="c1"># if type(sequence)!=str:  # Required because swifter calls this function with a series as sequence once, and</span>
    <span class="c1">#                              # does not handle the RunTimeError below properly.</span>
    <span class="c1">#         raise TypeError(f&quot;The input to this function must be a str, but was {type(sequence)}.&quot;)</span>
    <span class="c1">#</span>
    <span class="c1">#     GC_list=[]</span>
    <span class="c1">#     for nuc in sequence:</span>
    <span class="c1">#         if nuc==&quot;G&quot; or nuc==&quot;g&quot; or nuc==&quot;c&quot; or nuc==&quot;C&quot;:</span>
    <span class="c1">#             GC_at_pos=1</span>
    <span class="c1">#         elif nuc==&quot;A&quot; or nuc==&quot;a&quot; or nuc==&quot;t&quot; or nuc==&quot;T&quot;:</span>
    <span class="c1">#             GC_at_pos=0</span>
    <span class="c1">#         elif nuc==&quot;N&quot; or nuc==&quot;n&quot;:</span>
    <span class="c1">#             GC_at_pos=0.461 #0.461 is the genome-wide GC content mean</span>
    <span class="c1">#         else:</span>
    <span class="c1">#             logging.error(f&quot;Unexpected letter &#39;{nuc}&#39; encountered in genomic sequence.&quot;)</span>
    <span class="c1">#             raise RuntimeError</span>
    <span class="c1">#         GC_list.append(GC_at_pos)</span>
    <span class="c1">#</span>
    <span class="c1">#     GC_bias_factor=sum(np.array(GC_list)*np.array(self.GC_weights)) / self.total_GC_weight</span>
    <span class="c1">#     return GC_bias_factor</span>

<div class="viewcode-block" id="BiasFactorHandler.get_fwd_mappability_bias_factor"><a class="viewcode-back" href="../../../liquorice.html#liquorice.utils.BinTableBiasFactors.BiasFactorHandler.get_fwd_mappability_bias_factor">[docs]</a>    <span class="k">def</span> <span class="nf">get_fwd_mappability_bias_factor</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">mappability</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">float</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns a number representing the bin&#39;s mappability for fragments on the forward strand.</span>

<span class="sd">        :param mappability: A vector of values between 0 and 1, representing the mappbility of positions in and around</span>
<span class="sd">            the bin of interest. Must correspond to the coordinates &lt;Bin start coord&gt; - *max(* :attr:`.fragments` *)* -</span>
<span class="sd">            :attr:`.readlength` to &lt;Bin end coord&gt; + *max(* :attr:`.fragments` *)* + :attr:`.readlength`</span>
<span class="sd">        :return: A number in range 0-1 that represent&#39;s the bin&#39;s overall forward mappability. 1 ... highest</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">mapp_fwd</span><span class="o">=</span><span class="n">mappability</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">readlength</span><span class="p">:]</span>
        <span class="n">fwd_mappbias_of_bin</span><span class="o">=</span><span class="nb">sum</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">mapp_fwd</span><span class="p">)</span><span class="o">*</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">fwd_mapp_weights</span><span class="p">))</span> <span class="o">/</span> <span class="bp">self</span><span class="o">.</span><span class="n">total_fwd_mapp_weight</span>
        <span class="k">return</span> <span class="n">fwd_mappbias_of_bin</span></div>

<div class="viewcode-block" id="BiasFactorHandler.get_rev_mappability_bias_factor"><a class="viewcode-back" href="../../../liquorice.html#liquorice.utils.BinTableBiasFactors.BiasFactorHandler.get_rev_mappability_bias_factor">[docs]</a>    <span class="k">def</span> <span class="nf">get_rev_mappability_bias_factor</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">mappability</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">float</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns a number representing the bin&#39;s mappability for fragments on the reverse strand.</span>

<span class="sd">        :param mappability: A vector of values between 0 and 1, representing the mappbility of positions in and around</span>
<span class="sd">            the bin of interest. Must correspond to the coordinates &lt;Bin start coord&gt; - *max(* :attr:`.fragments` *)* -</span>
<span class="sd">            :attr:`.readlength` to &lt;Bin end coord&gt; + *max(* :attr:`.fragments` *)* + :attr:`.readlength`</span>
<span class="sd">        :return: A number in range 0-1 that represent&#39;s the bin&#39;s overall reverse mappability. 1 ... highest</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">mapp_rev</span><span class="o">=</span><span class="n">mappability</span><span class="p">[:</span><span class="o">-</span><span class="bp">self</span><span class="o">.</span><span class="n">readlength</span><span class="p">]</span>
        <span class="n">rev_mappbias_of_bin</span><span class="o">=</span><span class="nb">sum</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">mapp_rev</span><span class="p">)</span><span class="o">*</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">rev_mapp_weights</span><span class="p">))</span> <span class="o">/</span> <span class="bp">self</span><span class="o">.</span><span class="n">total_rev_mapp_weight</span>
        <span class="k">return</span> <span class="n">rev_mappbias_of_bin</span></div>

<div class="viewcode-block" id="BiasFactorHandler.get_table_with_bias_factors"><a class="viewcode-back" href="../../../liquorice.html#liquorice.utils.BinTableBiasFactors.BiasFactorHandler.get_table_with_bias_factors">[docs]</a>    <span class="k">def</span> <span class="nf">get_table_with_bias_factors</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Main method to retrieve bias factor information.</span>

<span class="sd">        :return: A `pandas.DataFrame` with one row per bin, containing columns of the input DataFrame :attr:`df`</span>
<span class="sd">            (&quot;chromosome&quot;, &quot;start&quot;, &quot;end&quot;, &quot;bin nr.&quot;, &quot;coverage&quot;, &quot;sequence&quot;) as well as newly added</span>
<span class="sd">            columns for bias</span>
<span class="sd">            factors for mappability, GC-content, and  all di- and trinucleotides (with reverse and forward complements</span>
<span class="sd">            merged into single factors). Bias factors are scaled between 0 and 1. Higher values correspond to higher</span>
<span class="sd">            nucleotide content, mappability etc. Example: An average bin is expected to have a bias factor of 0.461</span>
<span class="sd">            for humans (i.e. genomic GC content).</span>
<span class="sd">            Note that the input columns &quot;mappability&quot; and &quot;sequence&quot; will not be part of the returned dataframe</span>
<span class="sd">             in order to reduce the memory usage of this function.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="s2">&quot;mappability&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">skip_these_biasfactors</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Adding forward mappability bias factors with up to </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">n_cores</span><span class="si">}</span><span class="s2"> cores ...&quot;</span><span class="p">)</span>
            <span class="c1"># alternative versions that turned out to be slower:</span>
            <span class="c1"># modin: 3:19 min with 25 cores</span>
            <span class="c1"># swifter and modin: 3:17 with 25 cores</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;forward mappability&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;mappability&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">swifter</span><span class="o">.</span><span class="n">progress_bar</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">set_npartitions</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">n_cores</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">get_fwd_mappability_bias_factor</span><span class="p">)</span>  <span class="c1"># fast: 1:27 min with 25 cores allowed</span>
            <span class="c1"># (still uses 1 core only)</span>

            <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Adding reverse mappability bias factors with up to </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">n_cores</span><span class="si">}</span><span class="s2"> cores ...&quot;</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;reverse mappability&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;mappability&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">swifter</span><span class="o">.</span><span class="n">progress_bar</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">set_npartitions</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">n_cores</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">get_rev_mappability_bias_factor</span><span class="p">)</span>  <span class="c1"># fast: 1:27 min with 25 cores allowed</span>
            <span class="c1"># (still uses 1 core only)</span>

            <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Adding maximum mappability bias factors ...&quot;</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;max mappability&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="p">[[</span><span class="s2">&quot;forward mappability&quot;</span><span class="p">,</span> <span class="s2">&quot;reverse mappability&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>  <span class="c1"># fast</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">drop</span><span class="p">([</span><span class="s2">&quot;mappability&quot;</span><span class="p">],</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Adding sequence-based bias factors with up to </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">n_cores</span><span class="si">}</span><span class="s2"> cores - this may take a while. &quot;</span>
                     <span class="sa">f</span><span class="s2">&quot;(The following &#39;UserWarning&#39; from modin can be ignored) ...&quot;</span><span class="p">)</span>
        <span class="c1"># alternative versions that turned out to be slower:</span>

        <span class="c1"># Loop: slow, takes probably about 10 min:</span>
        <span class="c1"># nuc_GC_factors_dicts = [self.get_GC_and_di_and_trinuc_weights(seq) for seq in self.df[&quot;sequence&quot;].values]</span>

        <span class="c1"># multiprocessing pool.map: slow. schedules 25 tasks, but htop shows low usage.</span>
        <span class="c1"># with mp.Pool(processes=self.n_cores) as pool:</span>
        <span class="c1">#    nuc_GC_factors_dicts = pool.map(self.get_GC_and_di_and_trinuc_weights, self.df[&quot;sequence&quot;].values)</span>

        <span class="c1"># modin Series.apply: 3:40 on 25 cores incl conversion to df and concat</span>
        <span class="c1">#nuc_GC_factors_dicts = modinpd.Series(self.df[&quot;sequence&quot;]).apply(self.get_GC_and_di_and_trinuc_weights)</span>

        <span class="k">if</span> <span class="s2">&quot;di and trinucleotides and GC content&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">skip_these_biasfactors</span> <span class="ow">and</span> <span class="s2">&quot;di and trinucleotides&quot;</span> <span class="ow">not</span> \
                <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">skip_these_biasfactors</span><span class="p">:</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">n_cores</span><span class="o">==</span><span class="mi">1</span><span class="p">:</span>
                <span class="n">nuc_and_gc_factors_dicts</span><span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;sequence&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">get_GC_and_di_and_trinuc_weights</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">nuc_and_gc_factors_dicts</span><span class="o">=</span> <span class="n">modinpd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;sequence&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">swifter</span><span class="o">.</span><span class="n">set_npartitions</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">n_cores</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
                     <span class="bp">self</span><span class="o">.</span><span class="n">get_GC_and_di_and_trinuc_weights</span><span class="p">)</span>  <span class="c1"># fast: 2:01 min on 25 cores incl. conversion to df and concat</span>
            <span class="c1"># nuc_and_gc_factors_dicts=self.df[&quot;sequence&quot;].apply(self.get_GC_and_di_and_trinuc_weights)</span>
            <span class="n">nuc_and_gc_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">nuc_and_gc_factors_dicts</span><span class="o">.</span><span class="n">values</span><span class="p">))</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="p">,</span><span class="n">nuc_and_gc_df</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">drop</span><span class="p">([</span><span class="s2">&quot;sequence&quot;</span><span class="p">],</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

        <span class="k">if</span> <span class="s2">&quot;di and trinucleotides and GC content&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">skip_these_biasfactors</span> <span class="ow">and</span> \
            <span class="s2">&quot;di and trinucleotides&quot;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">skip_these_biasfactors</span><span class="p">:</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">n_cores</span><span class="o">==</span><span class="mi">1</span><span class="p">:</span>
                <span class="n">gc_factors_dict</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;sequence&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">get_GC_bias_factor</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">gc_factors_dict</span><span class="o">=</span><span class="n">modinpd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;sequence&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">swifter</span><span class="o">.</span><span class="n">set_npartitions</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">n_cores</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">get_GC_bias_factor</span><span class="p">)</span>
            <span class="n">gc_factor_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">gc_factors_dict</span><span class="o">.</span><span class="n">values</span><span class="p">))</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="p">,</span><span class="n">gc_factor_df</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

                <span class="c1"># self.df[&quot;GC content&quot;]=modinpd.Series(self.df[&quot;sequence&quot;]).swifter.set_npartitions(self.n_cores).apply(</span>
                <span class="c1">#     self.get_GC_bias_factor)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">drop</span><span class="p">([</span><span class="s2">&quot;sequence&quot;</span><span class="p">],</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>


        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span></div></div>
</pre></div>

           </div>
           
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2021, Peter Peneder.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>