

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>liquorice.utils.CoverageAndSequenceTablePreparation &mdash; LIQUORICE 0.5.0 documentation</title>
  

  
  <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../../" src="../../../_static/documentation_options.js"></script>
        <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
        <script src="../../../_static/jquery.js"></script>
        <script src="../../../_static/underscore.js"></script>
        <script src="../../../_static/doctools.js"></script>
    
    <script type="text/javascript" src="../../../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../../index.html">
          

          
            
            <img src="../../../_static/liquorice_logo_fitted_transparent.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../intro.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../intro.html#installation">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../intro.html#citation">Citation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../liquorice.html">The <cite>liquorice</cite> python package</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../liquorice_commandline_tool.html">The LIQUORICE command-line-tool</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../liquorice_commandline_tool.html#liquorice-s-summary-tool">LIQUORICEâ€™s summary tool</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../liquorice_commandline_tool.html#usage-parameters-and-examples">Usage, Parameters, and Examples</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">LIQUORICE</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="../../index.html">Module code</a> &raquo;</li>
        
      <li>liquorice.utils.CoverageAndSequenceTablePreparation</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for liquorice.utils.CoverageAndSequenceTablePreparation</h1><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">liquorice.utils</span> <span class="kn">import</span> <span class="n">deeptoolsCountReadsPerBinDontCheckProperPairSAMflag</span> <span class="k">as</span> <span class="n">crpb</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">pybedtools</span>
<span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">getenv</span><span class="p">(</span><span class="s2">&quot;TMP&quot;</span><span class="p">)</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
    <span class="n">pybedtools</span><span class="o">.</span><span class="n">set_tempdir</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">getenv</span><span class="p">(</span><span class="s2">&quot;TMP&quot;</span><span class="p">))</span>
<span class="kn">import</span> <span class="nn">pyBigWig</span>
<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">typing</span>
<span class="kn">from</span> <span class="nn">multiprocessing</span> <span class="kn">import</span> <span class="n">Pool</span>
<span class="kn">import</span> <span class="nn">pysam</span> <span class="c1"># TODO</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="p">[</span><span class="s2">&quot;PYTHONPATH&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="vm">__file__</span><span class="p">),</span> <span class="s1">&#39;../..&#39;</span><span class="p">))</span> <span class="c1"># required by Ray, which is</span>
<span class="c1"># used by modin</span>
<span class="kn">import</span> <span class="nn">modin.pandas</span> <span class="k">as</span> <span class="nn">modinpd</span>
<span class="kn">import</span> <span class="nn">swifter</span>

<div class="viewcode-block" id="run_get_coverage_of_region_per_chunk"><a class="viewcode-back" href="../../../liquorice.html#liquorice.utils.CoverageAndSequenceTablePreparation.run_get_coverage_of_region_per_chunk">[docs]</a><span class="k">def</span> <span class="nf">run_get_coverage_of_region_per_chunk</span><span class="p">(</span><span class="n">dataframe_chunk</span><span class="p">,</span><span class="n">bam_filepath</span><span class="p">,</span><span class="n">min_mapq</span><span class="p">,</span><span class="n">n_cores</span><span class="p">,</span> <span class="n">additional_crpb_kwargs</span><span class="p">,):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Uses :func`deeptools.countReadsPerBin.get_coverage_of_region()` to calculate the coverage in every bin</span>
<span class="sd">    in :param:`dataframe_chunk`. Adds the result as a column &quot;region coverage array&quot; and returns</span>
<span class="sd">    param:`dataframe_chunk`.</span>
<span class="sd">    This function can receive only picklable arguments, and can therefore be called multiple times in parallel.</span>

<span class="sd">    :param dataframe_chunk: A `pandas.DataFrame` with the columns &quot;chromosome&quot;,&quot;region start&quot;,&quot;region end&quot;.</span>
<span class="sd">        Usually, this would be only a chunk of the full DataFrame.</span>
<span class="sd">    :param bam_filepath: Path to the .bam file containing the mapped reads for the sample. Does not need to be</span>
<span class="sd">        duplicate-filtered, this is done by the function that calculates the coverage.</span>
<span class="sd">    :param min_mapq: Minimum mapping quality for a fragment to still be counted when calculating the coverage.</span>
<span class="sd">    :param n_cores: Number of cores used by `deeptools.countReadsPerBin`. Should probably be 1 if the function is called</span>
<span class="sd">        in parallel</span>
<span class="sd">    :param additional_crpb_kwargs: Additional keywords to be used for deeptool&#39;s `CountReadsPerBin`. Can be used</span>
<span class="sd">        to override the default settings: *&quot;ignoreDuplicates&quot;:True,  &quot;samFlag_exclude&quot;:256,  &quot;extendReads&quot;:True*</span>
<span class="sd">    :return: A `pandas.DataFrame` with the columns &quot;chromosome&quot;,&quot;region start&quot;,&quot;region end&quot; and &quot;region coverage array&quot;.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">crpb_kwargs</span><span class="o">=</span><span class="p">{</span>
        <span class="s2">&quot;minMappingQuality&quot;</span><span class="p">:</span><span class="n">min_mapq</span><span class="p">,</span>
        <span class="s2">&quot;ignoreDuplicates&quot;</span><span class="p">:</span><span class="kc">True</span><span class="p">,</span>
        <span class="s2">&quot;numberOfProcessors&quot;</span><span class="p">:</span><span class="n">n_cores</span><span class="p">,</span>
        <span class="s2">&quot;samFlag_exclude&quot;</span><span class="p">:</span><span class="mi">256</span><span class="p">,</span>  <span class="c1"># not primary alignment</span>
        <span class="s2">&quot;extendReads&quot;</span><span class="p">:</span><span class="kc">True</span><span class="p">,</span> <span class="c1"># extend reads does not seem to fully work</span>
        <span class="s2">&quot;numberOfSamples&quot;</span><span class="p">:</span><span class="mi">0</span><span class="p">}</span> <span class="c1"># required for technical reasons</span>
    <span class="n">crpb_kwargs</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">additional_crpb_kwargs</span><span class="p">)</span>
    <span class="n">crpb_obj</span> <span class="o">=</span> <span class="n">crpb</span><span class="o">.</span><span class="n">CountReadsPerBin</span><span class="p">([</span><span class="n">bam_filepath</span><span class="p">],</span> <span class="o">**</span><span class="n">crpb_kwargs</span><span class="p">)</span>

    <span class="n">pysam_alignment_file</span><span class="o">=</span><span class="n">pysam_alignment_file</span><span class="o">=</span><span class="n">pysam</span><span class="o">.</span><span class="n">AlignmentFile</span><span class="p">(</span><span class="n">bam_filepath</span><span class="p">)</span>

    <span class="n">stepsize</span><span class="o">=</span><span class="mi">1</span>
    <span class="n">dataframe_chunk</span><span class="p">[</span><span class="s2">&quot;region coverage array&quot;</span><span class="p">]</span><span class="o">=</span><span class="n">dataframe_chunk</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
        <span class="k">lambda</span> <span class="n">row</span><span class="p">:</span> <span class="n">crpb_obj</span><span class="o">.</span><span class="n">get_coverage_of_region</span><span class="p">(</span>
            <span class="n">pysam_alignment_file</span><span class="p">,</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;chromosome&quot;</span><span class="p">],[(</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;region start&quot;</span><span class="p">],</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;region end&quot;</span><span class="p">],</span><span class="n">stepsize</span><span class="p">)]),</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

    <span class="n">pysam_alignment_file</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="k">return</span> <span class="n">dataframe_chunk</span></div>

<div class="viewcode-block" id="parallelize_get_coverage_of_region"><a class="viewcode-back" href="../../../liquorice.html#liquorice.utils.CoverageAndSequenceTablePreparation.parallelize_get_coverage_of_region">[docs]</a><span class="k">def</span> <span class="nf">parallelize_get_coverage_of_region</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">n_cores</span><span class="p">,</span> <span class="n">bam_filepath</span><span class="p">,</span><span class="n">min_mapq</span><span class="p">,</span> <span class="n">additional_crpb_kwargs</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Splits the provided `pandas.DataFrame` into chunks of the size :param`n_cores`, and</span>
<span class="sd">    runs :func`run_get_coverage_of_region_per_chunk` in parallel on every chunk.</span>
<span class="sd">    From: https://towardsdatascience.com/make-your-own-super-pandas-using-multiproc-1c04f41944a1</span>

<span class="sd">    :param df: Full `pandas.DataFrame` with the columns &quot;chromosome&quot;,&quot;region start&quot;,&quot;region end&quot;.</span>
<span class="sd">    :param n_cores: Run this many processes in parallel.</span>
<span class="sd">    :param bam_filepath: Path to the .bam file containing the mapped reads for the sample. Does not need to be</span>
<span class="sd">        duplicate-filtered, this is done by the function that calculates the coverage.</span>
<span class="sd">    :param min_mapq: Minimum mapping quality for a fragment to still be counted when calculating the coverage.</span>
<span class="sd">    :param additional_crpb_kwargs: Additional keywords to be used for deeptool&#39;s `CountReadsPerBin`. Can be used</span>
<span class="sd">        to override the default settings: *&quot;ignoreDuplicates&quot;:True,  &quot;samFlag_exclude&quot;:256,  &quot;extendReads&quot;:True*</span>
<span class="sd">    :return: Full `pandas.DataFrame` with the columns &quot;chromosome&quot;,&quot;region start&quot;,&quot;region end&quot; and</span>
<span class="sd">        &quot;region coverage array&quot;.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">n_cores</span><span class="o">=</span><span class="nb">min</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="n">n_cores</span><span class="p">)</span> <span class="c1"># cannot split a df in more pieces than it is long</span>
    <span class="n">list_of_splitted_dfs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array_split</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">n_cores</span><span class="p">)</span>
    <span class="n">pool</span> <span class="o">=</span> <span class="n">Pool</span><span class="p">(</span><span class="n">n_cores</span><span class="p">)</span>
    <span class="n">tasks</span><span class="o">=</span><span class="p">[(</span><span class="n">splitted_df</span><span class="p">,</span><span class="n">bam_filepath</span><span class="p">,</span><span class="n">min_mapq</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="n">additional_crpb_kwargs</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">splitted_df</span> <span class="ow">in</span> <span class="n">list_of_splitted_dfs</span><span class="p">]</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">pool</span><span class="o">.</span><span class="n">starmap</span><span class="p">(</span><span class="n">run_get_coverage_of_region_per_chunk</span><span class="p">,</span> <span class="n">tasks</span><span class="p">))</span>
    <span class="n">pool</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="n">pool</span><span class="o">.</span><span class="n">join</span><span class="p">()</span>
    <span class="k">return</span> <span class="n">df</span></div>

<div class="viewcode-block" id="CoverageAndSequenceTablePreparation"><a class="viewcode-back" href="../../../liquorice.html#liquorice.utils.CoverageAndSequenceTablePreparation.CoverageAndSequenceTablePreparation">[docs]</a><span class="k">class</span> <span class="nc">CoverageAndSequenceTablePreparation</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Object used to set up a `pandas.DataFrame` containing basic information about each bin: Coverage (normalized for</span>
<span class="sd">    overall sequencing depth), genomic sequence &amp; mappability (both incl. surrounding regions; used for bias</span>
<span class="sd">    correction). Usually, a user would want to call :func:`.get_complete_table` on this object, and save the</span>
<span class="sd">    resulting `pandas.DataFrame` for subsequent bias-correction.</span>

<span class="sd">    :param bam_filepath: Path to the .bam file containing the mapped reads for the sample. Does not need to be</span>
<span class="sd">        duplicate-filtered, this is done by the function that calculates the coverage.</span>
<span class="sd">    :param bins_bed_filepath: Path to the .bed file containing chromosome, start, end, and bin nr. of every bin to be</span>
<span class="sd">        analyzed (in that order). Output of</span>
<span class="sd">        :func:`liquorice.utils.RegionFilteringAndBinning.Binning.write_bin_bedfile` is a suitable input here.</span>
<span class="sd">    :param refgenome_filepath: Path to the reference genome .fa(.gz) file. Must have a .fai index in the same dirname.</span>
<span class="sd">    :param refgenome_chromsizes_filepath: Path to a tab-delimited file containing the chromosome sizes for the</span>
<span class="sd">        reference genome. The first column must be the chromosome name, the second column its size.</span>
<span class="sd">        Can be the .fa.fai file associated to the reference genome.</span>
<span class="sd">    :param refgenome_mappability_bigwig_path: Path to a .bigWig file containing (forward) mappability values scaled</span>
<span class="sd">        between 0 and 1.</span>
<span class="sd">    :param readlength: (Average) length of the reads in the .bam file.</span>
<span class="sd">    :param longest_fraglen: Length by which sequencing and mappability information is extended beyond a bin&#39;s</span>
<span class="sd">        borders. Typically, this should be the longest sampled fragment length.</span>
<span class="sd">    :param min_mapq: Minimum mapping quality for a fragment to still be counted when calculating the coverage.</span>
<span class="sd">    :param n_cores: Number of cores to be used by deeptool&#39;s `CountReadsPerBin`, which is used for coverage</span>
<span class="sd">        calculations. Set to higher values to speed up the process.</span>
<span class="sd">    :param mean_seq_depth: A float that quantifies the global coverage/sequencing depth. Coverages per bin will be</span>
<span class="sd">        normalized (i.e. divided) by this value.</span>
<span class="sd">    :param skip_these_steps: A list containing the steps which should be skipped. May contain: &quot;coverage&quot;,&quot;sequence&quot;,</span>
<span class="sd">        &quot;mappability&quot;. Default []: Don&#39;t skip any steps.</span>
<span class="sd">    :param **additional_crpb_kwargs: Additional keywords to be used for deeptool&#39;s `CountReadsPerBin`. Can be used</span>
<span class="sd">        to override the default settings: *&quot;ignoreDuplicates&quot;:True,  &quot;samFlag_exclude&quot;:256,  &quot;extendReads&quot;:True*</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">bam_filepath</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">bins_bed_filepath</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">refgenome_filepath</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
                 <span class="n">refgenome_chromsizes_filepath</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">refgenome_mappability_bigwig_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
                 <span class="n">readlength</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">longest_fraglen</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">mean_seq_depth</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span> <span class="n">n_cores</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
                 <span class="n">min_mapq</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">20</span><span class="p">,</span> <span class="n">skip_these_steps</span><span class="p">:</span> <span class="n">typing</span><span class="o">.</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span><span class="p">[],</span><span class="o">**</span><span class="n">additional_crpb_kwargs</span> <span class="p">:</span> <span class="nb">dict</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">bam_filepath</span><span class="o">=</span><span class="n">bam_filepath</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">bins_bed_filepath</span><span class="o">=</span><span class="n">bins_bed_filepath</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">refgenome_filepath</span><span class="o">=</span><span class="n">refgenome_filepath</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">refgenome_chromsizes_filepath</span><span class="o">=</span><span class="n">refgenome_chromsizes_filepath</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">refgenome_mappability_bigwig_path</span><span class="o">=</span><span class="n">refgenome_mappability_bigwig_path</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">readlength</span><span class="o">=</span><span class="n">readlength</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">longest_fraglen</span><span class="o">=</span><span class="n">longest_fraglen</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">min_mapq</span><span class="o">=</span><span class="n">min_mapq</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">n_cores</span><span class="o">=</span><span class="n">n_cores</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">additional_crpb_kwargs</span><span class="o">=</span><span class="n">additional_crpb_kwargs</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">=</span><span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">mean_seq_depth</span><span class="o">=</span><span class="n">mean_seq_depth</span>

        <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">skip_these_steps</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">item</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;coverage&quot;</span><span class="p">,</span><span class="s2">&quot;sequence&quot;</span><span class="p">,</span><span class="s2">&quot;mappability&quot;</span><span class="p">]:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;skip_these_steps may only contain the following items: &quot;coverage&quot;,&quot;sequence&quot;,&#39;</span>
                                 <span class="s1">&#39;&quot;mappability&quot;&#39;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">skip_these_steps</span><span class="o">=</span><span class="n">skip_these_steps</span>


<div class="viewcode-block" id="CoverageAndSequenceTablePreparation.read_bed"><a class="viewcode-back" href="../../../liquorice.html#liquorice.utils.CoverageAndSequenceTablePreparation.CoverageAndSequenceTablePreparation.read_bed">[docs]</a>    <span class="k">def</span> <span class="nf">read_bed</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Reads in a .bed file and return corresponding `pandas.DataFrame`. 4th column in the .bed file is interpreted</span>
<span class="sd">        as bin nr. Layout of the .bed file MUST be [&quot;chromosome&quot;,&quot;start&quot;,&quot;end&quot;,&quot;bin nr.&quot;].</span>

<span class="sd">        :return: A `pandas.DataFrame` with the columns [&quot;chromosome&quot;,&quot;start&quot;,&quot;end&quot;,&quot;bin nr.&quot;].</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">result_df</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">bins_bed_filepath</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
        <span class="n">result_df</span><span class="o">.</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;chromosome&quot;</span><span class="p">,</span><span class="s2">&quot;start&quot;</span><span class="p">,</span><span class="s2">&quot;end&quot;</span><span class="p">,</span><span class="s2">&quot;bin nr.&quot;</span><span class="p">]</span>
        <span class="n">result_df</span><span class="p">[</span><span class="s2">&quot;chromosome&quot;</span><span class="p">]</span><span class="o">=</span><span class="n">result_df</span><span class="p">[</span><span class="s2">&quot;chromosome&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s2">&quot;str&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">result_df</span></div>

<div class="viewcode-block" id="CoverageAndSequenceTablePreparation.get_coverage_per_bin"><a class="viewcode-back" href="../../../liquorice.html#liquorice.utils.CoverageAndSequenceTablePreparation.CoverageAndSequenceTablePreparation.get_coverage_per_bin">[docs]</a>    <span class="k">def</span> <span class="nf">get_coverage_per_bin</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculates the normalized coverage per bin, based on :attr:`.bins_bed_filepath` and :attr:`.bam_filepath`.</span>
<span class="sd">        Uses deeptool&#39;s `countReadsPerBin` for calculating the coverage in every bin in :attr:`.df`.</span>
<span class="sd">        The obtained coverage value is then divided by :attr:`.mean_seq_depth`.</span>
<span class="sd">        This function is faster than :func:`get_coverage_per_bin_mean_over_per_basepair_values_chunked`, but provides</span>
<span class="sd">        less accurate results: Rather than the mean over the coverage of each base-pair in the bin, the total number of</span>
<span class="sd">        reads mapping to the bin (even partially mapping reads) are reported (after normalization for</span>
<span class="sd">        `:attr:.mean_seq_depth`).</span>

<span class="sd">        :return: A `pandas.Series` with the coverage per bin, normalized for total sequencing coverage.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">crpb_kwargs</span><span class="o">=</span><span class="p">{</span>
            <span class="s2">&quot;minMappingQuality&quot;</span><span class="p">:</span><span class="bp">self</span><span class="o">.</span><span class="n">min_mapq</span><span class="p">,</span>
            <span class="s2">&quot;ignoreDuplicates&quot;</span><span class="p">:</span><span class="kc">True</span><span class="p">,</span>
            <span class="s2">&quot;numberOfProcessors&quot;</span><span class="p">:</span><span class="bp">self</span><span class="o">.</span><span class="n">n_cores</span><span class="p">,</span>
            <span class="s2">&quot;samFlag_exclude&quot;</span><span class="p">:</span><span class="mi">256</span><span class="p">,</span>  <span class="c1"># not primary alignment</span>
            <span class="s2">&quot;extendReads&quot;</span><span class="p">:</span><span class="kc">True</span><span class="p">}</span>
        <span class="n">crpb_kwargs</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">additional_crpb_kwargs</span><span class="p">)</span>

        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Calculating coverage at every bin defined in the bed file &#39;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">bins_bed_filepath</span><span class="si">}</span><span class="s2">&#39;, &quot;</span>
                     <span class="sa">f</span><span class="s2">&quot;using </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">n_cores</span><span class="si">}</span><span class="s2"> cores - this may take a while ...&quot;</span><span class="p">)</span>
        <span class="n">crpb_obj</span> <span class="o">=</span> <span class="n">crpb</span><span class="o">.</span><span class="n">CountReadsPerBin</span><span class="p">(</span>
            <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">bam_filepath</span><span class="p">],</span>
            <span class="n">bedFile</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">bins_bed_filepath</span><span class="p">,</span>
            <span class="n">out_file_for_raw_data</span><span class="o">=</span><span class="s2">&quot;readcount_per_bin.tsv&quot;</span><span class="p">,</span>
            <span class="o">**</span><span class="n">crpb_kwargs</span><span class="p">)</span>
        <span class="n">crpb_obj</span><span class="o">.</span><span class="n">run</span><span class="p">()</span>

        <span class="c1"># re-order, as multiprocessing messes up the order of readcount_per_bin.tsv</span>
        <span class="n">raw_coverage_df</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;readcount_per_bin.tsv&quot;</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span><span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
        <span class="n">raw_coverage_df</span><span class="o">.</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;chromosome&quot;</span><span class="p">,</span><span class="s2">&quot;start&quot;</span><span class="p">,</span><span class="s2">&quot;end&quot;</span><span class="p">,</span><span class="s2">&quot;raw coverage&quot;</span><span class="p">]</span>
        <span class="n">raw_coverage_df</span><span class="p">[</span><span class="s2">&quot;chromosome&quot;</span><span class="p">]</span><span class="o">=</span><span class="n">raw_coverage_df</span><span class="p">[</span><span class="s2">&quot;chromosome&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s2">&quot;str&quot;</span><span class="p">)</span>

        <span class="n">raw_coverage_df</span><span class="o">=</span><span class="n">raw_coverage_df</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span> <span class="c1"># It can happen that 2 close regions share some bins</span>
        <span class="n">result_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="p">,</span> <span class="n">raw_coverage_df</span><span class="p">,</span> <span class="n">left_on</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;chromosome&quot;</span><span class="p">,</span> <span class="s2">&quot;start&quot;</span><span class="p">,</span> <span class="s2">&quot;end&quot;</span><span class="p">],</span>
                             <span class="n">right_on</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;chromosome&quot;</span><span class="p">,</span><span class="s2">&quot;start&quot;</span><span class="p">,</span><span class="s2">&quot;end&quot;</span><span class="p">],</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">)</span>
        <span class="k">assert</span> <span class="n">result_df</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="k">assert</span> <span class="nb">all</span><span class="p">([</span><span class="n">x</span><span class="o">==</span><span class="n">y</span> <span class="k">for</span> <span class="n">x</span><span class="p">,</span><span class="n">y</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">result_df</span><span class="p">[</span><span class="s2">&quot;start&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;start&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">)])</span>
        <span class="k">except</span> <span class="ne">AssertionError</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s2">&quot;The following result and self.df starts are not the same!:</span><span class="se">\n</span><span class="s2"> &quot;</span>
                          <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="p">[(</span><span class="n">x</span><span class="p">,</span><span class="n">y</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span><span class="p">,</span><span class="n">y</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">result_df</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">)</span> <span class="k">if</span> <span class="n">x</span> <span class="o">!=</span> <span class="n">y</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">raise</span> <span class="ne">AssertionError</span>

        <span class="c1"># Normalize coverage by genome-wide average</span>
        <span class="n">result_df</span><span class="p">[</span><span class="s2">&quot;coverage&quot;</span><span class="p">]</span><span class="o">=</span><span class="n">result_df</span><span class="p">[</span><span class="s2">&quot;raw coverage&quot;</span><span class="p">]</span><span class="o">/</span><span class="bp">self</span><span class="o">.</span><span class="n">mean_seq_depth</span>

        <span class="c1"># clean up</span>
        <span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="s2">&quot;readcount_per_bin.tsv&quot;</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">result_df</span><span class="p">[</span><span class="s2">&quot;coverage&quot;</span><span class="p">]</span></div>


<div class="viewcode-block" id="CoverageAndSequenceTablePreparation.get_coverage_per_bin_mean_over_per_basepair_values_chunked"><a class="viewcode-back" href="../../../liquorice.html#liquorice.utils.CoverageAndSequenceTablePreparation.CoverageAndSequenceTablePreparation.get_coverage_per_bin_mean_over_per_basepair_values_chunked">[docs]</a>    <span class="k">def</span> <span class="nf">get_coverage_per_bin_mean_over_per_basepair_values_chunked</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculates the normalized coverage per bin by averaging the per-base coverage over all positions in the bin and</span>
<span class="sd">        then deviding the result by :attr:`.mean_seq_depth`.</span>
<span class="sd">        Requires the following columns in :attr:`df`: &quot;chromosome&quot;,&quot;start&quot;,&quot;end&quot;,&quot;bin nr.&quot;,&quot;region nr.&quot;,&quot;bin size&quot;.</span>
<span class="sd">        Uses deeptool&#39;s `countReadsPerBin` for calculating the coverage in every bin in :attr:`.df`.</span>
<span class="sd">        The obtained coverage value is then divided by :attr:`.mean_seq_depth`.</span>

<span class="sd">        :return: A `pandas.Series` with the coverage per bin, normalized for total sequencing coverage.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Calculating coverage at every bin defined in the bed file &#39;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">bins_bed_filepath</span><span class="si">}</span><span class="s2">&#39;, &quot;</span>
                     <span class="sa">f</span><span class="s2">&quot;using </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">n_cores</span><span class="si">}</span><span class="s2"> cores - this may take a while ...&quot;</span><span class="p">)</span>

        <span class="n">df_regions</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s2">&quot;region nr.&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">agg</span><span class="p">({</span><span class="s2">&quot;chromosome&quot;</span><span class="p">:</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span><span class="n">x</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="s2">&quot;start&quot;</span><span class="p">:</span><span class="s2">&quot;min&quot;</span><span class="p">,</span><span class="s2">&quot;end&quot;</span><span class="p">:</span><span class="s2">&quot;max&quot;</span><span class="p">,})</span>
        <span class="n">df_regions</span><span class="o">=</span><span class="n">df_regions</span><span class="o">.</span><span class="n">rename</span><span class="p">({</span><span class="s2">&quot;start&quot;</span><span class="p">:</span><span class="s2">&quot;region start&quot;</span><span class="p">,</span><span class="s2">&quot;end&quot;</span><span class="p">:</span><span class="s2">&quot;region end&quot;</span><span class="p">},</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">df_regions</span><span class="o">=</span><span class="n">parallelize_get_coverage_of_region</span><span class="p">(</span><span class="n">df_regions</span><span class="p">,</span><span class="bp">self</span><span class="o">.</span><span class="n">n_cores</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">bam_filepath</span><span class="p">,</span><span class="bp">self</span><span class="o">.</span><span class="n">min_mapq</span><span class="p">,</span>
                                                      <span class="bp">self</span><span class="o">.</span><span class="n">additional_crpb_kwargs</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">df_regions</span><span class="p">[[</span><span class="s2">&quot;region coverage array&quot;</span><span class="p">,</span><span class="s2">&quot;region start&quot;</span><span class="p">]],</span><span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="p">,</span><span class="n">left_index</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                         <span class="n">right_on</span><span class="o">=</span><span class="s2">&quot;region nr.&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;relative start&quot;</span><span class="p">]</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;start&quot;</span><span class="p">]</span><span class="o">-</span><span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;region start&quot;</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;relative end&quot;</span><span class="p">]</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;relative start&quot;</span><span class="p">]</span><span class="o">+</span><span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;bin size&quot;</span><span class="p">]</span>

        <span class="k">if</span> <span class="kc">True</span><span class="p">:</span> <span class="c1"># based on testing results, this version works more stably than the one below:</span>
            <span class="n">result</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
                <span class="k">lambda</span> <span class="n">row</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span>
                    <span class="n">row</span><span class="p">[</span><span class="s2">&quot;region coverage array&quot;</span><span class="p">][</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;relative start&quot;</span><span class="p">]:</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;relative end&quot;</span><span class="p">]]</span><span class="o">/</span><span class="bp">self</span><span class="o">.</span><span class="n">mean_seq_depth</span><span class="p">),</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">result</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">swifter</span><span class="o">.</span><span class="n">progress_bar</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">set_npartitions</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">n_cores</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
                <span class="k">lambda</span> <span class="n">row</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span>
                    <span class="n">row</span><span class="p">[</span><span class="s2">&quot;region coverage array&quot;</span><span class="p">][</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;relative start&quot;</span><span class="p">]:</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;relative end&quot;</span><span class="p">]]</span><span class="o">/</span><span class="bp">self</span><span class="o">.</span><span class="n">mean_seq_depth</span><span class="p">),</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">drop</span><span class="p">([</span><span class="s2">&quot;region coverage array&quot;</span><span class="p">,</span><span class="s2">&quot;region start&quot;</span><span class="p">,</span><span class="s2">&quot;relative start&quot;</span><span class="p">,</span><span class="s2">&quot;relative end&quot;</span><span class="p">,],</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

        <span class="c1">#result=result/self.mean_seq_depth # TODO what is better: This, or the inner mean in the row where result is</span>
        <span class="c1"># first created?</span>

        <span class="k">return</span> <span class="n">result</span></div>


<div class="viewcode-block" id="CoverageAndSequenceTablePreparation.get_sequence_per_bin"><a class="viewcode-back" href="../../../liquorice.html#liquorice.utils.CoverageAndSequenceTablePreparation.CoverageAndSequenceTablePreparation.get_sequence_per_bin">[docs]</a>    <span class="k">def</span> <span class="nf">get_sequence_per_bin</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">typing</span><span class="o">.</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Get a list of genomic sequences, one per entry in the .bed file attr:`bins_bed_filepath`.</span>
<span class="sd">        Sequences are extended by :attr:`longest_fraglen`</span>

<span class="sd">        :return: A list of the genomic sequences of every bin in the attr:`bins_bed_filepath`.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">bed</span> <span class="o">=</span> <span class="n">pybedtools</span><span class="o">.</span><span class="n">BedTool</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">bins_bed_filepath</span><span class="p">)</span>

        <span class="c1"># in case self.refgenome_chromsizes_filepath is a .fa.fai file, we need to convert it to a chrom.sizes file</span>
        <span class="c1"># for use in bedtools</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">refgenome_chromsizes_filepath</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.fai&quot;</span><span class="p">):</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;tmp.chrom.sizes&quot;</span><span class="p">,</span><span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">outfile</span><span class="p">:</span>
                <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">refgenome_chromsizes_filepath</span><span class="p">)</span> <span class="k">as</span> <span class="n">infile</span><span class="p">:</span>
                    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">infile</span><span class="p">:</span>
                        <span class="n">line</span><span class="o">=</span><span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
                        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">line</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">line</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span><span class="n">file</span><span class="o">=</span><span class="n">outfile</span><span class="p">)</span>
            <span class="n">bed</span> <span class="o">=</span> <span class="n">bed</span><span class="o">.</span><span class="n">slop</span><span class="p">(</span><span class="n">g</span><span class="o">=</span><span class="s2">&quot;tmp.chrom.sizes&quot;</span><span class="p">,</span><span class="n">b</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">longest_fraglen</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">bed</span> <span class="o">=</span> <span class="n">bed</span><span class="o">.</span><span class="n">slop</span><span class="p">(</span><span class="n">g</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">refgenome_chromsizes_filepath</span><span class="p">,</span><span class="n">b</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">longest_fraglen</span><span class="p">)</span>
        <span class="n">fasta_of_sequences</span> <span class="o">=</span> <span class="n">bed</span><span class="o">.</span><span class="n">sequence</span><span class="p">(</span><span class="n">fi</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">refgenome_filepath</span><span class="p">)</span><span class="o">.</span><span class="n">seqfn</span>

        <span class="n">sequence_list</span><span class="o">=</span><span class="p">[]</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">fasta_of_sequences</span><span class="p">)</span> <span class="k">as</span> <span class="n">infile</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">infile</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;&gt;&quot;</span><span class="p">):</span>
                    <span class="k">continue</span>
                <span class="n">sequence_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">line</span><span class="o">.</span><span class="n">rstrip</span><span class="p">())</span>

        <span class="c1">#  clean up if necessary</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">refgenome_chromsizes_filepath</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.fai&quot;</span><span class="p">):</span>
            <span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="s2">&quot;tmp.chrom.sizes&quot;</span><span class="p">)</span>
        <span class="k">del</span> <span class="n">bed</span>

        <span class="n">pybedtools</span><span class="o">.</span><span class="n">cleanup</span><span class="p">()</span>
        <span class="k">return</span> <span class="n">sequence_list</span></div>

<div class="viewcode-block" id="CoverageAndSequenceTablePreparation.get_mappability_for_bin"><a class="viewcode-back" href="../../../liquorice.html#liquorice.utils.CoverageAndSequenceTablePreparation.CoverageAndSequenceTablePreparation.get_mappability_for_bin">[docs]</a>    <span class="k">def</span> <span class="nf">get_mappability_for_bin</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">row</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">,</span> <span class="n">mapp_bw</span><span class="p">:</span> <span class="n">pyBigWig</span><span class="o">.</span><span class="n">pyBigWig</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">typing</span><span class="o">.</span><span class="n">List</span><span class="p">[</span><span class="nb">float</span><span class="p">]:</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        For a single bin, get a list of (forward) mappability values in and around the bin based on its genomic</span>
<span class="sd">        coordinates and a `pyBigWig.bigWig` file that contains the genome-wide mappability.</span>

<span class="sd">        :param row: `pandas.Series` containing columns &quot;chromosome&quot;,&quot;start&quot;, and &quot;end&quot;.</span>
<span class="sd">        :param mapp_bw: `pyBigWig.pyBigWig` object with mappability info (i.e. result of `pyBigWig.open()`)</span>
<span class="sd">        :return: Mappability for the bin at base resolultion.</span>
<span class="sd">            Extended downstream by (:attr:`longest_fraglen` + :attr:`.readlength`) and upstream by :attr:`longest_fraglen`.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">mapp_bw</span><span class="o">.</span><span class="n">values</span><span class="p">(</span>
            <span class="n">row</span><span class="p">[</span><span class="s2">&quot;chromosome&quot;</span><span class="p">],</span>
            <span class="n">row</span><span class="p">[</span><span class="s2">&quot;start&quot;</span><span class="p">]</span><span class="o">-</span><span class="bp">self</span><span class="o">.</span><span class="n">longest_fraglen</span><span class="o">-</span><span class="bp">self</span><span class="o">.</span><span class="n">readlength</span><span class="p">,</span>
            <span class="n">row</span><span class="p">[</span><span class="s2">&quot;end&quot;</span><span class="p">]</span><span class="o">+</span><span class="bp">self</span><span class="o">.</span><span class="n">longest_fraglen</span><span class="p">,</span> <span class="n">numpy</span><span class="o">=</span><span class="kc">True</span>
        <span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">float16</span><span class="p">)</span> <span class="c1">#float16 saves a lot of memory and should still allow enough precision for our purposes</span></div>
        <span class="c1"># for 6000 regions, binsize 500 and extend_to 20k, memory usage of the dataframe was reduced from 5GB to 3GB.</span>

<div class="viewcode-block" id="CoverageAndSequenceTablePreparation.get_mappability_per_bin"><a class="viewcode-back" href="../../../liquorice.html#liquorice.utils.CoverageAndSequenceTablePreparation.CoverageAndSequenceTablePreparation.get_mappability_per_bin">[docs]</a>    <span class="k">def</span> <span class="nf">get_mappability_per_bin</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Opens a `.bigWig` file using `pyBigWig` and calls :func:`.get_mappability_for_bin` for every bin.</span>

<span class="sd">        :return: A `pandas.Series` with mappability information per bin. Extended downstream by</span>
<span class="sd">            (:attr:`longest_fraglen` + :attr:`.readlength`) and upstream by :attr:`longest_fraglen`.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">mapp_bw</span> <span class="o">=</span> <span class="n">pyBigWig</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">refgenome_mappability_bigwig_path</span><span class="p">)</span>
        <span class="n">result</span><span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">row</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_mappability_for_bin</span><span class="p">(</span><span class="n">row</span><span class="p">,</span> <span class="n">mapp_bw</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">mapp_bw</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
        <span class="k">return</span> <span class="n">result</span></div>

<div class="viewcode-block" id="CoverageAndSequenceTablePreparation.get_complete_table"><a class="viewcode-back" href="../../../liquorice.html#liquorice.utils.CoverageAndSequenceTablePreparation.CoverageAndSequenceTablePreparation.get_complete_table">[docs]</a>    <span class="k">def</span> <span class="nf">get_complete_table</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Main method to retrieve a `pandas.DateFrame` that can be used to calculate bias factors on.</span>
<span class="sd">        Calls :func:`.read_bed&#39;, :func:`.get_coverage_per_bin_mean_over_per_basepair_values_chunked`,</span>
<span class="sd">        :func:`.get_sequence_per_bin`, and</span>
<span class="sd">        :func:`.get_mappability_per_bin&#39;, constructs :attr:`.df`, a `pandas.DataFrame`, from the results, and</span>
<span class="sd">        returns it.</span>

<span class="sd">        :return: A `pandas.DataFrame` where rows correspond to the genomic bins in :attr:`.bins_bed_filepath`.</span>
<span class="sd">            This `DataFrame` has the following columns: &quot;chromosome&quot;,&quot;start&quot;,&quot;end&quot;,&quot;bin nr.&quot;,&quot;coverage&quot;,&quot;sequence&quot;,</span>
<span class="sd">            &quot;mappability&quot;.</span>
<span class="sd">            Entries in &quot;sequence&quot; and &quot;mappability&quot; columns are extended by :attr:`longest_fraglen`  (and downstream</span>
<span class="sd">            additionally by :attr:`.readlength` for &quot;mappability&quot;).</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">read_bed</span><span class="p">()</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;bin size&quot;</span><span class="p">]</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;end&quot;</span><span class="p">]</span><span class="o">-</span><span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;start&quot;</span><span class="p">]</span> <span class="c1"># TODO remove if not required anymore</span>

        <span class="k">if</span> <span class="s2">&quot;coverage&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">skip_these_steps</span><span class="p">:</span>
           <span class="c1"># logging.info(&quot;Retrieving fragment coverage values ... &quot;)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;region nr.&quot;</span><span class="p">]</span><span class="o">=</span><span class="p">((</span><span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;bin nr.&quot;</span><span class="p">]</span><span class="o">-</span><span class="mi">1</span><span class="p">)</span><span class="o">!=</span><span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;bin nr.&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">shift</span><span class="p">())</span><span class="o">.</span><span class="n">cumsum</span><span class="p">()</span><span class="o">-</span><span class="mi">1</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;coverage&quot;</span><span class="p">]</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">get_coverage_per_bin_mean_over_per_basepair_values_chunked</span><span class="p">()</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">drop</span><span class="p">({</span><span class="s2">&quot;region nr.&quot;</span><span class="p">},</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span> <span class="c1"># clean up</span>

        <span class="k">if</span> <span class="s2">&quot;sequence&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">skip_these_steps</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Retrieving genomic sequences for every bin ... &quot;</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;sequence&quot;</span><span class="p">]</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">get_sequence_per_bin</span><span class="p">()</span>

        <span class="k">if</span> <span class="s2">&quot;mappability&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">skip_these_steps</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Retrieving mappability information for every bin ... &quot;</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;mappability&quot;</span><span class="p">]</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">get_mappability_per_bin</span><span class="p">()</span>

        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span></div></div>


</pre></div>

           </div>
           
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2021, Peter Peneder.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>