

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>liquorice.utils.MeanSequencingDepth &mdash; LIQUORICE 0.5.0 documentation</title>
  

  
  <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../../" src="../../../_static/documentation_options.js"></script>
        <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
        <script src="../../../_static/jquery.js"></script>
        <script src="../../../_static/underscore.js"></script>
        <script src="../../../_static/doctools.js"></script>
    
    <script type="text/javascript" src="../../../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../../index.html">
          

          
            
            <img src="../../../_static/liquorice_logo_fitted_transparent.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../intro.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../intro.html#installation">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../intro.html#citation">Citation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../liquorice.html">The <cite>liquorice</cite> python package</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../liquorice_commandline_tool.html">The LIQUORICE command-line-tool</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../liquorice_commandline_tool.html#liquorice-s-summary-tool">LIQUORICEâ€™s summary tool</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../liquorice_commandline_tool.html#usage-parameters-and-examples">Usage, Parameters, and Examples</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">LIQUORICE</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="../../index.html">Module code</a> &raquo;</li>
        
      <li>liquorice.utils.MeanSequencingDepth</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for liquorice.utils.MeanSequencingDepth</h1><div class="highlight"><pre>
<span></span><span class="c1">#import deeptools.countReadsPerBin as crpb</span>
<span class="kn">from</span> <span class="nn">liquorice.utils</span> <span class="kn">import</span> <span class="n">deeptoolsCountReadsPerBinDontCheckProperPairSAMflag</span> <span class="k">as</span> <span class="n">crpb</span>
<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">typing</span>
<span class="kn">import</span> <span class="nn">os</span>


<div class="viewcode-block" id="sample_bam_to_get_sequencing_depth"><a class="viewcode-back" href="../../../liquorice.html#liquorice.utils.MeanSequencingDepth.sample_bam_to_get_sequencing_depth">[docs]</a><span class="k">def</span> <span class="nf">sample_bam_to_get_sequencing_depth</span><span class="p">(</span><span class="n">bam_filepath</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">n_sites_to_sample</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">10000</span><span class="p">,</span> <span class="n">n_cores</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
                                       <span class="n">min_mapq</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">20</span><span class="p">,</span>
                                       <span class="n">chromosomes_list</span><span class="p">:</span> <span class="n">typing</span><span class="o">.</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
                                       <span class="o">**</span><span class="n">additional_crpb_kwargs</span><span class="p">:</span> <span class="nb">dict</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Estimates the overall sequencing coverage of the .bam file by sampling</span>
<span class="sd">    sites that are regularily distributed across the genome (using :func:`deeptools.countReadsPerBin.CountReadsPerBin`</span>
<span class="sd">    in a slightly modified version that allows fragments without the SAM flag is_proper_pair.).</span>


<span class="sd">    :param bam_filepath: Path to the .bam file for which mean coverage should be calculated.</span>
<span class="sd">    :param n_sites_to_sample: Number of sites (length 1) to sample.</span>
<span class="sd">    :param n_cores: Number of cores to be used by :func:`deeptools.countReadsPerBin.CountReadsPerBin`.</span>
<span class="sd">    :param min_mapq: `minMappingQuality` setting for `deeptools.countReadsPerBin.CountReadsPerBin`.</span>
<span class="sd">    :param chromosomes_list: A list of chromosomes that should be analyzed. Regions on chromosomes that are not in</span>
<span class="sd">        this list will be excluded from analysis. Default None means [&quot;chr1&quot;, ..., &quot;chr22&quot;].</span>
<span class="sd">    :param additional_crpb_kwargs: Use to override the default parameters for</span>
<span class="sd">        :func:`deeptools.countReadsPerBin.CountReadsPerBin`. If not specified otherwise in `additional_crpb_kwargs`,</span>
<span class="sd">        the following attributes are passed to `CountReadsPerBin`: `ignoreDuplicates=True`, `samFlag_exclude=256`,</span>
<span class="sd">        `extendReads=True`.</span>

<span class="sd">    :return: The average coverage determined from the sampled regions.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">if</span> <span class="n">chromosomes_list</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">chromosomes_list</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;chr&quot;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">23</span><span class="p">)]</span>

    <span class="n">crpb_kwargs</span><span class="o">=</span><span class="p">{</span>
        <span class="s2">&quot;minMappingQuality&quot;</span><span class="p">:</span><span class="n">min_mapq</span><span class="p">,</span>
        <span class="s2">&quot;ignoreDuplicates&quot;</span><span class="p">:</span><span class="kc">True</span><span class="p">,</span>
        <span class="s2">&quot;numberOfProcessors&quot;</span><span class="p">:</span><span class="n">n_cores</span><span class="p">,</span>
        <span class="s2">&quot;samFlag_exclude&quot;</span><span class="p">:</span><span class="mi">256</span><span class="p">,</span>  <span class="c1"># not primary alignment</span>
        <span class="s2">&quot;extendReads&quot;</span><span class="p">:</span><span class="kc">True</span><span class="p">}</span>
    <span class="n">crpb_kwargs</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">additional_crpb_kwargs</span><span class="p">)</span>

    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Calculating coverage at </span><span class="si">{</span><span class="n">n_sites_to_sample</span><span class="si">}</span><span class="s2"> regions&quot;</span>
                 <span class="sa">f</span><span class="s2">&quot; regularily distributed across the genome to determine overall coverage ...&quot;</span><span class="p">)</span>
    <span class="n">crpb_obj</span> <span class="o">=</span> <span class="n">crpb</span><span class="o">.</span><span class="n">CountReadsPerBin</span><span class="p">(</span>
        <span class="p">[</span><span class="n">bam_filepath</span><span class="p">],</span>
        <span class="n">binLength</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
        <span class="n">numberOfSamples</span><span class="o">=</span><span class="n">n_sites_to_sample</span><span class="p">,</span>
        <span class="n">out_file_for_raw_data</span><span class="o">=</span><span class="s2">&quot;sequencing_depth_at_sampled_regions.tsv&quot;</span><span class="p">,</span>
        <span class="o">**</span><span class="n">crpb_kwargs</span><span class="p">)</span>

    <span class="n">crpb_obj</span><span class="o">.</span><span class="n">run</span><span class="p">()</span>

    <span class="n">seqdepth_df</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;sequencing_depth_at_sampled_regions.tsv&quot;</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span><span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
    <span class="n">seqdepth_df</span><span class="o">.</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;chr&quot;</span><span class="p">,</span><span class="s2">&quot;start&quot;</span><span class="p">,</span><span class="s2">&quot;end&quot;</span><span class="p">,</span><span class="s2">&quot;coverage&quot;</span><span class="p">]</span>
    <span class="n">seqdepth_df</span><span class="p">[</span><span class="s2">&quot;chr&quot;</span><span class="p">]</span><span class="o">=</span><span class="n">seqdepth_df</span><span class="p">[</span><span class="s2">&quot;chr&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s2">&quot;str&quot;</span><span class="p">)</span>
    <span class="n">mean_seq_depth</span><span class="o">=</span><span class="n">seqdepth_df</span><span class="p">[</span><span class="n">seqdepth_df</span><span class="p">[</span><span class="s2">&quot;chr&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">chromosomes_list</span><span class="p">)][</span><span class="s2">&quot;coverage&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>
    <span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="s2">&quot;sequencing_depth_at_sampled_regions.tsv&quot;</span><span class="p">)</span>
    <span class="c1"># mean_seq_depth=cr.run().mean(axis=0)[0]</span>

    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Mean coverage at the </span><span class="si">{</span><span class="n">n_sites_to_sample</span><span class="si">}</span><span class="s2"> regions of length 1 is &quot;</span>
                 <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">mean_seq_depth</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">mean_seq_depth</span></div>
</pre></div>

           </div>
           
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2021, Peter Peneder.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>